{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "8ad7879a-06f6-4268-8e13-cdef86c43f24",
   "metadata": {},
   "source": [
    "<div style=\"max-width:1200px\"><img src=\"../_resources/mgnify_banner.png\" width=\"100%\"></div>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1bebce35-5332-44bc-9b9c-83ffd1237b1c",
   "metadata": {},
   "source": [
    "<img src=\"../_resources/mgnify_logo.png\" width=\"200px\">\n",
    "\n",
    "# Cross-Study analysis, using MGnifyR\n",
    "\n",
    "The [MGnify API](https://www.ebi.ac.uk/metagenomics/api/v1) returns data and relationships as JSON. \n",
    "[MGnifyR](https://github.com/beadyallen/MGnifyR) is a package to help you read MGnify data into your R analyses.\n",
    "\n",
    "**This example shows you how to perform an analysis across multiple Studies. It is an example of comparing taxonomic diversity from two places in different Studies.**\n",
    "\n",
    "You can also discover more about the \"API\" using the [Browsable API interface in your web browser](https://www.ebi.ac.uk/metagenomics/api/v1).\n",
    "\n",
    "This is an interactive code notebook (a Jupyter Notebook).\n",
    "To run this code, click into each cell and press the ▶ button in the top toolbar, or press `shift+enter`.\n",
    "\n",
    "---"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "40008ca6-98da-407a-ba43-5813e7b3dd0c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/markdown": [
       "# Help with MGnifyR\n",
       "\n",
       "MGnifyR is an R package that provides a convenient way for R users to access data from [the MGnify API](https://www.ebi.ac.uk/metagenomics/api/).\n",
       "\n",
       "Detailed help for each function is available in R using the standard `?function_name` command (i.e. typing `?mgnify_query` will bring up built-in help for the mgnify_query command). \n",
       "\n",
       "A vignette is available containing a reasonably verbose overview of the main functionality. \n",
       "This can be read either within R with the `vignette(\"MGnifyR\")` command, or [in the development repository](https://htmlpreview.github.io/?https://github.com/beadyallen/MGnifyR/blob/master/doc/MGnifyR.html)\n",
       "\n",
       "## MGnifyR Command cheat sheet\n",
       "\n",
       "The following list of key functions should give a starting point for finding relevent documentation.\n",
       "\n",
       "- `mgnify_client()` : Create the client object required for all other functions.\n",
       "- `mgnify_query()` : Search the whole MGnify database.\n",
       "- `mgnify_analyses_from_xxx()` : Convert xxx accessions to analyses accessions. xxx is either samples or studies.\n",
       "- `mgnify_get_analyses_metadata()` : Retrieve all study, sample and analysis metadata for given analyses.\n",
       "- `mgnify_get_analyses_phyloseq()` : Convert abundance, taxonomic, and sample metadata into a single phyloseq object.\n",
       "- `mgnify_get_analyses_results()` : Get functional annotation results for a set of analyses.\n",
       "- `mgnify_download()` : Download raw results files from MGnify.\n",
       "- `mgnify_retrieve_json()` : Low level API access helper function.\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "library(IRdisplay)\n",
    "display_markdown(file = '../_resources/mgnifyr_help.md')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8c8e8846-a74b-4100-93a5-05b3ed4305c5",
   "metadata": {},
   "source": [
    "Load libraries:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "803a6b89-7c5c-4d19-98fe-223a98112af8",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: permute\n",
      "\n",
      "Loading required package: lattice\n",
      "\n",
      "This is vegan 2.5-7\n",
      "\n",
      "Loading required package: mlr\n",
      "\n",
      "Loading required package: ParamHelpers\n",
      "\n",
      "Warning message: 'mlr' is in 'maintenance-only' mode since July 2019.\n",
      "Future development will only happen in 'mlr3'\n",
      "(<https://mlr3.mlr-org.com>). Due to the focus on 'mlr3' there might be\n",
      "uncaught bugs meanwhile in {mlr} - please consider switching.\n",
      "\n",
      "\n",
      "Attaching package: ‘BiocGenerics’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    anyDuplicated, append, as.data.frame, basename, cbind, colnames,\n",
      "    dirname, do.call, duplicated, eval, evalq, Filter, Find, get, grep,\n",
      "    grepl, intersect, is.unsorted, lapply, Map, mapply, match, mget,\n",
      "    order, paste, pmax, pmax.int, pmin, pmin.int, Position, rank,\n",
      "    rbind, Reduce, rownames, sapply, setdiff, sort, table, tapply,\n",
      "    union, unique, unsplit, which.max, which.min\n",
      "\n",
      "\n",
      "Welcome to Bioconductor\n",
      "\n",
      "    Vignettes contain introductory material; view with\n",
      "    'browseVignettes()'. To cite Bioconductor, see\n",
      "    'citation(\"Biobase\")', and for packages 'citation(\"pkgname\")'.\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘Biobase’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:phyloseq’:\n",
      "\n",
      "    sampleNames\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘limma’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:BiocGenerics’:\n",
      "\n",
      "    plotMA\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:ParamHelpers’:\n",
      "\n",
      "    isNumeric\n",
      "\n",
      "\n",
      "Loaded glmnet 4.1-2\n",
      "\n"
     ]
    }
   ],
   "source": [
    "library(vegan)\n",
    "library(ggplot2)\n",
    "library(phyloseq)\n",
    "library(MGnifyR)\n",
    "library(SIAMCAT)\n",
    "library(metagenomeSeq, quietly = T)\n",
    "\n",
    "mg <- mgnify_client(usecache = T, cache_dir = '/tmp/mgnify_cache')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "69c43ac2-1238-4528-947e-0eb0d8ad8e10",
   "metadata": {},
   "source": [
    "# Example: compare taxonomic abundances of two soil studies\n",
    "*This example is based on a [MGnify workshop exercise](https://beadyallen.github.io/MGnifyR/Exercises.html) created by [Ben Allen](https://github.com/beadyallen) (the author of MGnifyR).*"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d0164772-7177-4a63-991b-151866ddf928",
   "metadata": {},
   "source": [
    "## Fetch the MGnify Analyses accessions for each of two Studies\n",
    "(one with samples from Malaysia, one with samples from Panama)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "47a77e8c-af32-4cd3-b3be-af8a57c09d37",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "  |======================================================================| 100%\n",
      "  |======================================================================| 100%\n"
     ]
    }
   ],
   "source": [
    "panama <- mgnify_analyses_from_studies(mg, 'MGYS00003920')\n",
    "malaysia <- mgnify_analyses_from_studies(mg, 'MGYS00003918')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "49b01fc0-ec0a-4f66-afeb-7d4501579388",
   "metadata": {},
   "source": [
    "Join the Analyses accession lists"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "ff0060e4-8c37-4f97-832c-2c7b98a1c52e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "'There are 78 accessions between the studies'"
      ],
      "text/latex": [
       "'There are 78 accessions between the studies'"
      ],
      "text/markdown": [
       "'There are 78 accessions between the studies'"
      ],
      "text/plain": [
       "[1] \"There are 78 accessions between the studies\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "accessions <- c(panama, malaysia)\n",
    "sprintf('There are %d accessions between the studies', length(accessions))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "378f735e-9fb8-4f96-b8c2-a478c29078e9",
   "metadata": {},
   "source": [
    "Fetch metadata for all of the Analyses from the MGnify API"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "8b3a9b2d-5733-42aa-af96-1984379dcbf5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "  |======================================================================| 100%\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 6 × 59</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>analysis_experiment-type</th><th scope=col>analysis_accession</th><th scope=col>analysis_analysis-status</th><th scope=col>analysis_pipeline-version</th><th scope=col>analysis_complete-time</th><th scope=col>analysis_instrument-platform</th><th scope=col>analysis_instrument-model</th><th scope=col>analysis_Submitted nucleotide sequences</th><th scope=col>analysis_Nucleotide sequences after format-specific filtering</th><th scope=col>analysis_Nucleotide sequences after length filtering</th><th scope=col>⋯</th><th scope=col>sample_depth</th><th scope=col>sample_elevation</th><th scope=col>sample_pH</th><th scope=col>sample_geographic location (latitude)</th><th scope=col>sample_instrument model</th><th scope=col>sample_ENA checklist</th><th scope=col>sample_last update date</th><th scope=col>sample_acc_type</th><th scope=col>run_accession</th><th scope=col>biome_string</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>⋯</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>MGYA00329605</th><td>unknown</td><td>MGYA00329605</td><td>completed</td><td>4.1</td><td>2019-03-06T00:00:00</td><td>ILLUMINA</td><td>Illumina MiSeq</td><td>120  </td><td>120  </td><td>118  </td><td>⋯</td><td>0.0-0.2</td><td>53.0 </td><td>3.3</td><td>8.971447 </td><td>Illumina MiSeq</td><td>ERC000011</td><td>2017-02-18</td><td>samples</td><td>ERR1844693</td><td>root:Environmental:Terrestrial:Soil</td></tr>\n",
       "\t<tr><th scope=row>MGYA00329606</th><td>unknown</td><td>MGYA00329606</td><td>completed</td><td>4.1</td><td>2019-03-06T00:00:00</td><td>ILLUMINA</td><td>Illumina MiSeq</td><td>27607</td><td>27607</td><td>27233</td><td>⋯</td><td>0.0-0.2</td><td>120.0</td><td>6.4</td><td>9.1442795</td><td>Illumina MiSeq</td><td>ERC000011</td><td>2017-02-18</td><td>samples</td><td>ERR1844694</td><td>root:Environmental:Terrestrial:Soil</td></tr>\n",
       "\t<tr><th scope=row>MGYA00329607</th><td>unknown</td><td>MGYA00329607</td><td>completed</td><td>4.1</td><td>2019-03-06T00:00:00</td><td>ILLUMINA</td><td>Illumina MiSeq</td><td>25238</td><td>25238</td><td>24806</td><td>⋯</td><td>0.0-0.2</td><td>120.0</td><td>5.6</td><td>9.1442795</td><td>Illumina MiSeq</td><td>ERC000011</td><td>2017-02-18</td><td>samples</td><td>ERR1844695</td><td>root:Environmental:Terrestrial:Soil</td></tr>\n",
       "\t<tr><th scope=row>MGYA00329608</th><td>unknown</td><td>MGYA00329608</td><td>completed</td><td>4.1</td><td>2019-03-06T00:00:00</td><td>ILLUMINA</td><td>Illumina MiSeq</td><td>5    </td><td>5    </td><td>5    </td><td>⋯</td><td>0.0-0.2</td><td>53.0 </td><td>3.6</td><td>8.971447 </td><td>Illumina MiSeq</td><td>ERC000011</td><td>2017-02-18</td><td>samples</td><td>ERR1844696</td><td>root:Environmental:Terrestrial:Soil</td></tr>\n",
       "\t<tr><th scope=row>MGYA00329609</th><td>unknown</td><td>MGYA00329609</td><td>completed</td><td>4.1</td><td>2019-03-06T00:00:00</td><td>ILLUMINA</td><td>Illumina MiSeq</td><td>20805</td><td>20805</td><td>20464</td><td>⋯</td><td>0.0-0.2</td><td>155.0</td><td>6.9</td><td>9.080094 </td><td>Illumina MiSeq</td><td>ERC000011</td><td>2017-02-18</td><td>samples</td><td>ERR1844697</td><td>root:Environmental:Terrestrial:Soil</td></tr>\n",
       "\t<tr><th scope=row>MGYA00329610</th><td>unknown</td><td>MGYA00329610</td><td>completed</td><td>4.1</td><td>2019-03-06T00:00:00</td><td>ILLUMINA</td><td>Illumina MiSeq</td><td>21171</td><td>21171</td><td>20834</td><td>⋯</td><td>0.0-0.2</td><td>64.0 </td><td>6.3</td><td>8.908218 </td><td>Illumina MiSeq</td><td>ERC000011</td><td>2017-02-18</td><td>samples</td><td>ERR1844698</td><td>root:Environmental:Terrestrial:Soil</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 6 × 59\n",
       "\\begin{tabular}{r|lllllllllllllllllllll}\n",
       "  & analysis\\_experiment-type & analysis\\_accession & analysis\\_analysis-status & analysis\\_pipeline-version & analysis\\_complete-time & analysis\\_instrument-platform & analysis\\_instrument-model & analysis\\_Submitted nucleotide sequences & analysis\\_Nucleotide sequences after format-specific filtering & analysis\\_Nucleotide sequences after length filtering & ⋯ & sample\\_depth & sample\\_elevation & sample\\_pH & sample\\_geographic location (latitude) & sample\\_instrument model & sample\\_ENA checklist & sample\\_last update date & sample\\_acc\\_type & run\\_accession & biome\\_string\\\\\n",
       "  & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & ⋯ & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\tMGYA00329605 & unknown & MGYA00329605 & completed & 4.1 & 2019-03-06T00:00:00 & ILLUMINA & Illumina MiSeq & 120   & 120   & 118   & ⋯ & 0.0-0.2 & 53.0  & 3.3 & 8.971447  & Illumina MiSeq & ERC000011 & 2017-02-18 & samples & ERR1844693 & root:Environmental:Terrestrial:Soil\\\\\n",
       "\tMGYA00329606 & unknown & MGYA00329606 & completed & 4.1 & 2019-03-06T00:00:00 & ILLUMINA & Illumina MiSeq & 27607 & 27607 & 27233 & ⋯ & 0.0-0.2 & 120.0 & 6.4 & 9.1442795 & Illumina MiSeq & ERC000011 & 2017-02-18 & samples & ERR1844694 & root:Environmental:Terrestrial:Soil\\\\\n",
       "\tMGYA00329607 & unknown & MGYA00329607 & completed & 4.1 & 2019-03-06T00:00:00 & ILLUMINA & Illumina MiSeq & 25238 & 25238 & 24806 & ⋯ & 0.0-0.2 & 120.0 & 5.6 & 9.1442795 & Illumina MiSeq & ERC000011 & 2017-02-18 & samples & ERR1844695 & root:Environmental:Terrestrial:Soil\\\\\n",
       "\tMGYA00329608 & unknown & MGYA00329608 & completed & 4.1 & 2019-03-06T00:00:00 & ILLUMINA & Illumina MiSeq & 5     & 5     & 5     & ⋯ & 0.0-0.2 & 53.0  & 3.6 & 8.971447  & Illumina MiSeq & ERC000011 & 2017-02-18 & samples & ERR1844696 & root:Environmental:Terrestrial:Soil\\\\\n",
       "\tMGYA00329609 & unknown & MGYA00329609 & completed & 4.1 & 2019-03-06T00:00:00 & ILLUMINA & Illumina MiSeq & 20805 & 20805 & 20464 & ⋯ & 0.0-0.2 & 155.0 & 6.9 & 9.080094  & Illumina MiSeq & ERC000011 & 2017-02-18 & samples & ERR1844697 & root:Environmental:Terrestrial:Soil\\\\\n",
       "\tMGYA00329610 & unknown & MGYA00329610 & completed & 4.1 & 2019-03-06T00:00:00 & ILLUMINA & Illumina MiSeq & 21171 & 21171 & 20834 & ⋯ & 0.0-0.2 & 64.0  & 6.3 & 8.908218  & Illumina MiSeq & ERC000011 & 2017-02-18 & samples & ERR1844698 & root:Environmental:Terrestrial:Soil\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 6 × 59\n",
       "\n",
       "| <!--/--> | analysis_experiment-type &lt;chr&gt; | analysis_accession &lt;chr&gt; | analysis_analysis-status &lt;chr&gt; | analysis_pipeline-version &lt;chr&gt; | analysis_complete-time &lt;chr&gt; | analysis_instrument-platform &lt;chr&gt; | analysis_instrument-model &lt;chr&gt; | analysis_Submitted nucleotide sequences &lt;chr&gt; | analysis_Nucleotide sequences after format-specific filtering &lt;chr&gt; | analysis_Nucleotide sequences after length filtering &lt;chr&gt; | ⋯ ⋯ | sample_depth &lt;chr&gt; | sample_elevation &lt;chr&gt; | sample_pH &lt;chr&gt; | sample_geographic location (latitude) &lt;chr&gt; | sample_instrument model &lt;chr&gt; | sample_ENA checklist &lt;chr&gt; | sample_last update date &lt;chr&gt; | sample_acc_type &lt;chr&gt; | run_accession &lt;chr&gt; | biome_string &lt;chr&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| MGYA00329605 | unknown | MGYA00329605 | completed | 4.1 | 2019-03-06T00:00:00 | ILLUMINA | Illumina MiSeq | 120   | 120   | 118   | ⋯ | 0.0-0.2 | 53.0  | 3.3 | 8.971447  | Illumina MiSeq | ERC000011 | 2017-02-18 | samples | ERR1844693 | root:Environmental:Terrestrial:Soil |\n",
       "| MGYA00329606 | unknown | MGYA00329606 | completed | 4.1 | 2019-03-06T00:00:00 | ILLUMINA | Illumina MiSeq | 27607 | 27607 | 27233 | ⋯ | 0.0-0.2 | 120.0 | 6.4 | 9.1442795 | Illumina MiSeq | ERC000011 | 2017-02-18 | samples | ERR1844694 | root:Environmental:Terrestrial:Soil |\n",
       "| MGYA00329607 | unknown | MGYA00329607 | completed | 4.1 | 2019-03-06T00:00:00 | ILLUMINA | Illumina MiSeq | 25238 | 25238 | 24806 | ⋯ | 0.0-0.2 | 120.0 | 5.6 | 9.1442795 | Illumina MiSeq | ERC000011 | 2017-02-18 | samples | ERR1844695 | root:Environmental:Terrestrial:Soil |\n",
       "| MGYA00329608 | unknown | MGYA00329608 | completed | 4.1 | 2019-03-06T00:00:00 | ILLUMINA | Illumina MiSeq | 5     | 5     | 5     | ⋯ | 0.0-0.2 | 53.0  | 3.6 | 8.971447  | Illumina MiSeq | ERC000011 | 2017-02-18 | samples | ERR1844696 | root:Environmental:Terrestrial:Soil |\n",
       "| MGYA00329609 | unknown | MGYA00329609 | completed | 4.1 | 2019-03-06T00:00:00 | ILLUMINA | Illumina MiSeq | 20805 | 20805 | 20464 | ⋯ | 0.0-0.2 | 155.0 | 6.9 | 9.080094  | Illumina MiSeq | ERC000011 | 2017-02-18 | samples | ERR1844697 | root:Environmental:Terrestrial:Soil |\n",
       "| MGYA00329610 | unknown | MGYA00329610 | completed | 4.1 | 2019-03-06T00:00:00 | ILLUMINA | Illumina MiSeq | 21171 | 21171 | 20834 | ⋯ | 0.0-0.2 | 64.0  | 6.3 | 8.908218  | Illumina MiSeq | ERC000011 | 2017-02-18 | samples | ERR1844698 | root:Environmental:Terrestrial:Soil |\n",
       "\n"
      ],
      "text/plain": [
       "             analysis_experiment-type analysis_accession\n",
       "MGYA00329605 unknown                  MGYA00329605      \n",
       "MGYA00329606 unknown                  MGYA00329606      \n",
       "MGYA00329607 unknown                  MGYA00329607      \n",
       "MGYA00329608 unknown                  MGYA00329608      \n",
       "MGYA00329609 unknown                  MGYA00329609      \n",
       "MGYA00329610 unknown                  MGYA00329610      \n",
       "             analysis_analysis-status analysis_pipeline-version\n",
       "MGYA00329605 completed                4.1                      \n",
       "MGYA00329606 completed                4.1                      \n",
       "MGYA00329607 completed                4.1                      \n",
       "MGYA00329608 completed                4.1                      \n",
       "MGYA00329609 completed                4.1                      \n",
       "MGYA00329610 completed                4.1                      \n",
       "             analysis_complete-time analysis_instrument-platform\n",
       "MGYA00329605 2019-03-06T00:00:00    ILLUMINA                    \n",
       "MGYA00329606 2019-03-06T00:00:00    ILLUMINA                    \n",
       "MGYA00329607 2019-03-06T00:00:00    ILLUMINA                    \n",
       "MGYA00329608 2019-03-06T00:00:00    ILLUMINA                    \n",
       "MGYA00329609 2019-03-06T00:00:00    ILLUMINA                    \n",
       "MGYA00329610 2019-03-06T00:00:00    ILLUMINA                    \n",
       "             analysis_instrument-model analysis_Submitted nucleotide sequences\n",
       "MGYA00329605 Illumina MiSeq            120                                    \n",
       "MGYA00329606 Illumina MiSeq            27607                                  \n",
       "MGYA00329607 Illumina MiSeq            25238                                  \n",
       "MGYA00329608 Illumina MiSeq            5                                      \n",
       "MGYA00329609 Illumina MiSeq            20805                                  \n",
       "MGYA00329610 Illumina MiSeq            21171                                  \n",
       "             analysis_Nucleotide sequences after format-specific filtering\n",
       "MGYA00329605 120                                                          \n",
       "MGYA00329606 27607                                                        \n",
       "MGYA00329607 25238                                                        \n",
       "MGYA00329608 5                                                            \n",
       "MGYA00329609 20805                                                        \n",
       "MGYA00329610 21171                                                        \n",
       "             analysis_Nucleotide sequences after length filtering ⋯\n",
       "MGYA00329605 118                                                  ⋯\n",
       "MGYA00329606 27233                                                ⋯\n",
       "MGYA00329607 24806                                                ⋯\n",
       "MGYA00329608 5                                                    ⋯\n",
       "MGYA00329609 20464                                                ⋯\n",
       "MGYA00329610 20834                                                ⋯\n",
       "             sample_depth sample_elevation sample_pH\n",
       "MGYA00329605 0.0-0.2      53.0             3.3      \n",
       "MGYA00329606 0.0-0.2      120.0            6.4      \n",
       "MGYA00329607 0.0-0.2      120.0            5.6      \n",
       "MGYA00329608 0.0-0.2      53.0             3.6      \n",
       "MGYA00329609 0.0-0.2      155.0            6.9      \n",
       "MGYA00329610 0.0-0.2      64.0             6.3      \n",
       "             sample_geographic location (latitude) sample_instrument model\n",
       "MGYA00329605 8.971447                              Illumina MiSeq         \n",
       "MGYA00329606 9.1442795                             Illumina MiSeq         \n",
       "MGYA00329607 9.1442795                             Illumina MiSeq         \n",
       "MGYA00329608 8.971447                              Illumina MiSeq         \n",
       "MGYA00329609 9.080094                              Illumina MiSeq         \n",
       "MGYA00329610 8.908218                              Illumina MiSeq         \n",
       "             sample_ENA checklist sample_last update date sample_acc_type\n",
       "MGYA00329605 ERC000011            2017-02-18              samples        \n",
       "MGYA00329606 ERC000011            2017-02-18              samples        \n",
       "MGYA00329607 ERC000011            2017-02-18              samples        \n",
       "MGYA00329608 ERC000011            2017-02-18              samples        \n",
       "MGYA00329609 ERC000011            2017-02-18              samples        \n",
       "MGYA00329610 ERC000011            2017-02-18              samples        \n",
       "             run_accession biome_string                       \n",
       "MGYA00329605 ERR1844693    root:Environmental:Terrestrial:Soil\n",
       "MGYA00329606 ERR1844694    root:Environmental:Terrestrial:Soil\n",
       "MGYA00329607 ERR1844695    root:Environmental:Terrestrial:Soil\n",
       "MGYA00329608 ERR1844696    root:Environmental:Terrestrial:Soil\n",
       "MGYA00329609 ERR1844697    root:Environmental:Terrestrial:Soil\n",
       "MGYA00329610 ERR1844698    root:Environmental:Terrestrial:Soil"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "metadata <- mgnify_get_analyses_metadata(mg, accessions)\n",
    "head(metadata)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0c7acf9f-cdd6-4e20-ab1c-c09823050059",
   "metadata": {},
   "source": [
    "## Taxonomic analysis\n",
    "First, build a [phyloseq](https://joey711.github.io/phyloseq/) object of the Analyses metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "9ff639ea-6e49-46e4-ab89-5a360617ed90",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "  |======================================================================| 100%\n"
     ]
    }
   ],
   "source": [
    "ps <- mgnify_get_analyses_phyloseq(mg, metadata$analysis_accession)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fe03725b-e3f6-4a90-b416-5aaa49b1e2fb",
   "metadata": {},
   "source": [
    "#### Filter out low-abundance samples\n",
    "Filter out samples with low abundances, leaving just those in the \"normal\" distribution. This is particularly important given the normalisation approach we're about to use ([`phyloseq`'s slightly controversial `rarefy_even_depth`](https://www.rdocumentation.org/packages/phyloseq/versions/1.16.2/topics/rarefy_even_depth).)\n",
    "\n",
    "We make a histogram and note that samples with abundances $\\lt 10^3$ i.e. $\\lt 1000$ seem to be outliers."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "c28ff212-59ad-4944-8357-3941a64dbbee",
   "metadata": {},
   "outputs": [
    {
     "ename": "ERROR",
     "evalue": "Error in h(simpleError(msg, call)): error in evaluating the argument 'object' in selecting a method for function 'otu_table': object 'ps' not found\n",
     "output_type": "error",
     "traceback": [
      "Error in h(simpleError(msg, call)): error in evaluating the argument 'object' in selecting a method for function 'otu_table': object 'ps' not found\nTraceback:\n",
      "1. hist(log10(sample_sums(ps)), breaks = 50)",
      "2. sample_sums(ps)",
      "3. otu_table(x)",
      "4. .handleSimpleError(function (cond) \n . .Internal(C_tryCatchHelper(addr, 1L, cond)), \"object 'ps' not found\", \n .     base::quote(otu_table(x)))",
      "5. h(simpleError(msg, call))"
     ]
    }
   ],
   "source": [
    "hist(log10(sample_sums(ps)), breaks = 50)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "edf6f9ef-533f-4c26-bfae-91b3898ca8c0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAA0gAAANICAMAAADKOT/pAAADAFBMVEUAAAABAQECAgIDAwME\nBAQFBQUGBgYHBwcICAgJCQkKCgoLCwsMDAwNDQ0ODg4PDw8QEBARERESEhITExMUFBQVFRUW\nFhYXFxcYGBgZGRkaGhobGxscHBwdHR0eHh4fHx8gICAhISEiIiIjIyMkJCQlJSUmJiYnJyco\nKCgpKSkqKiorKyssLCwtLS0uLi4vLy8wMDAxMTEyMjIzMzM0NDQ1NTU2NjY3Nzc4ODg5OTk6\nOjo7Ozs8PDw9PT0+Pj4/Pz9AQEBBQUFCQkJDQ0NERERFRUVGRkZHR0dISEhJSUlKSkpLS0tM\nTExNTU1OTk5PT09QUFBRUVFSUlJTU1NUVFRVVVVWVlZXV1dYWFhZWVlaWlpbW1tcXFxdXV1e\nXl5fX19gYGBhYWFiYmJjY2NkZGRlZWVmZmZnZ2doaGhpaWlqampra2tsbGxtbW1ubm5vb29w\ncHBxcXFycnJzc3N0dHR1dXV2dnZ3d3d4eHh5eXl6enp7e3t8fHx9fX1+fn5/f3+AgICBgYGC\ngoKDg4OEhISFhYWGhoaHh4eIiIiJiYmKioqLi4uMjIyNjY2Ojo6Pj4+QkJCRkZGSkpKTk5OU\nlJSVlZWWlpaXl5eYmJiZmZmampqbm5ucnJydnZ2enp6fn5+goKChoaGioqKjo6OkpKSlpaWm\npqanp6eoqKipqamqqqqrq6usrKytra2urq6vr6+wsLCxsbGysrKzs7O0tLS1tbW2tra3t7e4\nuLi5ubm6urq7u7u8vLy9vb2+vr6/v7/AwMDBwcHCwsLDw8PExMTFxcXGxsbHx8fIyMjJycnK\nysrLy8vMzMzNzc3Ozs7Pz8/Q0NDR0dHS0tLT09PU1NTV1dXW1tbX19fY2NjZ2dna2trb29vc\n3Nzd3d3e3t7f39/g4ODh4eHi4uLj4+Pk5OTl5eXm5ubn5+fo6Ojp6enq6urr6+vs7Ozt7e3u\n7u7v7+/w8PDx8fHy8vLz8/P09PT19fX29vb39/f4+Pj5+fn6+vr7+/v8/Pz9/f3+/v7////i\nsF19AAAACXBIWXMAABJ0AAASdAHeZh94AAAgAElEQVR4nO3dCZgU1b2w8TPM9AzDMCAoi8ji\nek00CkEj0UjirhEUE42CmgtCghEVk08jrnFJxERvFuOKW9yu+xJ3I1H0Gre45CYasuBy44Lg\nAhoVkGXqq6Wn+1R1T9VM86+pU5z39zxOV0131zlV1jvdPdPMKAfAWlNZTwBYFxASIICQAAGE\nBAggJEAAIQECCAkQQEiAAEICBBASIICQAAGEBAggJEAAIQECCAkQQEiAAEICBBASIICQAAGE\nBAggJEAAIQECCAkQQEiAAEICBBASIICQAAGEBAggJEAAIQECCAkQQEiAAEICBBASIICQAAGE\nBAggJEAAIQECCAkQsM6FdJtSTZkNfuvons2bLhObyRSlJq7ljNI/HAKTTHKXUvWOc7BS3/XW\n2i/NkveQrlBKLQ0Wd1dq72xDekZ5Pi6udXYmL2+v1EbF5QXTRjT23/v+YOWPdar3O2s5pdQP\nh8QkkwQhLWxRPZ53ypdmWedCWvDLX15YcatF9epv3TCZI5Xq+8vrVhbXOncOr57dqEohPdXq\np6jO9dd2VOpHazul1EOSmGSSICTnFKV2crRLo6xzIVX1a9UtIX1dqSPLa506h193H456tIe0\nYjOlhk52T80eLzj+41th8dpOKe2QRCaZpBjSwgal/qhdGsWOkL7SPSG5Mzi+vNapc/h6pab9\nV3tINys1aInjfEOpb7trRyh10FpPKe2QRCaZpBiSM6H46miCga+S1rmQ2s+cVXN2H9AwYPvZ\n7znOuOD50nHuZz86Z0y/wsC9rl4d3OXq7Xr1H/fca+6Vyx3nKqW+umrmBgMdp+2mvQY0tO7w\na+9W7mfHOreNah5x6kpn/n7rtez5sj68vr0jg2EqXyPFDXr9Zo96+xCENEmpme7FA+4zxDbn\ns15K3R/Zk05M7UqltnJu+Uqf1j2eDE3i70du3tS6/QWrKo6gvvlTlRrjfa545lbfd/0O7ZOM\njBqec1ho9ysOTmT1N9v16rfvc78thuRetq7SLk2yroa0cpfiab3pK1pI/7tR8bM7vu/d4wf+\nctPF7gd37Ualtv25/7/ssOKtxrf5jxJfuLXOWzv6tQ28iwEflkcPba+jkGIH/b9lTjmkrZW6\n0r14273qNedZpeo+jOxJJ6Z2g1JDzvdv0vA7bRJ39Azut9vyyAEMbT4SUtUBQndon2Rk1PCc\nQ8K7Hz04kdXjgxufVgzpA3flee3SJOtqSJcq9bmbnnroQKW+5vz1bvdG//3Ea84S93/TJpf+\ndpb7HHuce5s/up8edcX1O/UO/j+599x4WGHUls697ouUS166yr3Vrf5nNxy6/zF93c1+fchx\nY9y7/LI0eHh7rzzhvt6Z+MQTa4rXFmcSP2hxH4KQWpS6x71Y7Z67v3cu8L7IR/akE1O7xV1r\nmnzTbHf7I1aWJvFas1I//MdzX1Pq5MgBDG0+ElLVAUJ3aJ9kZNTQbUIiux85OJHVZ73y77xn\nn4b2Y7WFUhfqlwZZF0IqK4c0RamfuxcrJx3z0zXOOyp4jXSWUn3edvwXJt5XtOlKred+4Vs2\nonTaqC3edBcuGjfOexq4v1L/GXz2W+5XdPei56vOZ1sqtV9p8Mj2qr9Gih+0uA9+SG1uQHO9\nBfesv8v5rlKHOdE9SZ6a98mpjv9oon5XmsQxSu3iXrzXW7VGHpJCm68MqXKA0B3aJxkZNXSb\nkMjuRw5OZNXd+oBP3OeJn2s/VgcrNV2/NMi6GtJMpYZft6h4o/aQRio1xVtf3U+psx1nq+BF\nvXN6+bS5Udv0sUrtFXz2j+450aTUJPeTJ7hfUEu3iGyvekjxgxb3wQ9pmTvUo95CH6Vucg4M\nXtWF9yR5at4n/9ddW9Wq1JmlSWym1KnLXV/1HutCQpuvElLFAKE7tE8yMmq1OQciux85OJHV\nzwd1Oj9uP1Yzit/amNEd3+LomnUhpGEjfD31kP7Uywtrs2l3eK9aiyG1NQRfKP0ffrhnhnuT\n2d7a7eXTJvifP3fCpk1+mLsHn/XeqzA0+OmO+yJqs/axo9urGlLCoMV9qHhEutvZTakznOie\nJE/N/WTBfxT4olJHlCbRo/zF5oLwAQxtvkpIFQOE7tA+ycio4TnrwrsfOTjRY9Vc/Inane3H\n6hR/v8uXBlkXQqr6Xbt5WwfnzcZPlUL6xL2Y4990D/fVutPWflY9VDpt6v2T4RL3ipbPj9yg\n/Wz1t+d+Tb/UvbhQCymyveohJQxa3IfgNVKr95TO/bLuBjXPcV+SnO9E9yR5au4n1/PvtrNS\nh+iTaHdG5Ajqm68Mqcq+63don2Rk1NBtdJHdjxycxGN1rlJf1i8Nss6G5LQ9dfY+7stk71m2\n9ogUnJtj/Kf2TcUveLeFT5uP3a+ahy7znz/EhxTdXoePSDGDFvchCGm0Upe5F2+6013obewM\nJ7InnZhaaf+/6D8vKk6iXqlfdXQItQN1avGJ65y4kPQ77F5+RNJHDd0mJLz7kYMTPVY9iw9f\nt2qPSHvolwZZd0PyrL7bfbL9QOk10qjiS+OV7pf+nznO5sXn66eFT5v/cW/9p2B78SFFt1f9\nNVL8oMV9CEL6jlLfcy/uUWqI//JjZnljwZ50Ymre07EF7tqq3v7LjOJt/qP4WqYjxQPlfqUf\n2Ob438mPCUm7Q/skI6OGbhMS2f3IwYms/of/PRXHObn9WB1VfG10FK+RpHUQ0rLZU/b3n6bt\n5T1bWuTe6Al35Wz3eZH3bbnLlar7h+McrlTf99xHoGHh02Zu8BL7r+7riq/GhxTZXvWQ4gct\n7kMQ0oNKbfC+07aPUsf639/yTqrQnnRiat4pfZITfKvt0dJtjnXb/NQ9tQ894qS3QscvfKC8\n75bd6zh/69VxSOE7TNe+a1ceNXybkMjuRw5OZHWy+4RxiXvjoe3H6pDid+sO4bt20jp6RHK/\nth34wPP/c1ZBNS12VheUGnvr75yl7v+RzS++44fu84vvuLd5xL3vttdc9aWW8Gnztnua7vfS\nPRttqVSfpxfHhRTZXvWQ4gd98txzz/2mO5B7cYezxn1itNHh7hO8nq/5I23tRPakE1NzP9lQ\nOGPeJesrteXq0iRedV+3f+X+37kjbR15/R86UK+6r85ajjyh/04xj0ihO7RPMjJq6DYhkd2P\nHJzI6mPujbe7+dovuQ9zPfx7uw9RF+mXBllXQ3ppaPG1dY+rHP/dpP6P+Eo/OP+m/9OUyf5y\nr5+FTxvvy7dryOtDvJfmcSFFt1f9vXaxg/64/F2ACY7z8vr+Ut01jv/TyB4fRfckeWruJ/ud\n5N+q55PaJG4LvtenNoq+5zB8oL7nL27xB3cObR2EFLpD+yQjo4Y3GhLe/ejBiawe4S+3XOR+\n8B7hPqgrvrOhjnc2SOvwNdKiH28/qNDrc9P/7K28dcB6PTc5x1346Cdf6lvY8Bv3BPdYc/6W\nTQMP+ssDwV1KIa382VbNG33nbWfulg1Db44NKbK9Dt60GjdoOCTnzenDCwMOeNq7Xem9dvqe\nJE/N/WSLM2dUz377/zk0iflTN2nq9YVTl1QcwdCBWn3uFo0bTV/8ljudTzt6WqvfoX2SkVHD\nGw0J73704ESP1c+3bBx40Mt/dafzb4f32pnuWv/FvXGD1vbG6tT/AVJYcZJdHrWmY867v800\n/9wZE70vbPsrdYCBg9b2T326OaTiJDs96tocc/49kqEWuE+1D3jsD993nzg8bOKgNf3j08RT\n+pn1dad2eYCIYJIdjxoZbm2OOf9C1lRntr88Od3IQWv6dQjd/IhUnGTnR639mHu/q+E57dIs\nVofkPHLQ0ELTiEMeM3TQKcH7bbqku0MKJtmFUWs+5gcXf8jQfmkWu0MChBASIICQAAGEBAgg\nJEAAIQECCAkQQEiAAEICBBASIICQAAGEBAggJEAAIQECCAkQQEiAAEICBBASIICQAAGEBAgg\nJEAAIQECCAkQQEiAAEICBBASIICQAAGEBAggJEAAIQECCAkQQEiAAEICBBASIICQAAGEBAgg\nJEAAIQECCAkQQEiAAEICBBASIICQAAGEBAggJEAAIQECCAkQQEiAAEICBBASIICQAAGEBAgg\nJEDAWoe08uXnV0hMBMiz2kN6ZJeNv/6M89AQpfpcLDghII9qDumpBtWnR8tTfYb958H91IOS\nUwLyp+aQ9hv8Z+fdXYePXOY4SzbeR3JKQP7UHNL6P3Y/PKeu8ZZ/0j9y5Zp5c0sevmEtpofU\nPTtX96TEJheHNjl3scQ2TVdzSA3XuR8Wqvu95asaIle+NqBfSatauTYTRLo+Vr37lPVWiwS2\nObOgbbJPYabAJo1Xc0iDznA/PKYu8JZPGRRzwyfVZ7WOgfR9qG59qew+9ZbANo/eS9vkS3sd\nLbBJ49Uc0sT+j372l20+P9w98PP7HRRzQ0IyGiHJqDmkv7UqpfrPH9Fr1x0b6p+NuSEhGY2Q\nZNT+c6SXJo2Z8nfnpR3q1Ka/jbsdIRmNkGSs/VuEPn43/npCMhohyUj/vXaEZDRCkkFIliMk\nGYRkOUKSQUiWIyQZhGQ5QpJBSJYjJBmEZDlCkkFIliMkGYRkOUKSQUiWIyQZhGQ5QpJBSJYj\nJBmEZDlCkkFIliMkGYRkOUKSQUiWIyQZhGQ5QpJBSJYjJBmEZDlCkkFIliMkGYRkOUKSQUiW\nIyQZhGQ5QpJBSJYjJBmEZDlCkkFIliMkGYRkOUKSQUiWIyQZhGQ5QpJBSJYjJBmEZDlCkkFI\nliMkGYRkOUKSQUiWIyQZhGQ5QpJBSJYjJBmEZDlCkkFIliMkGYRkOUKSQUiWIyQZhGQ5QpJB\nSJYjJBmEZDlCkkFIliMkGYRkOUKSQUiWIyQZhGQ5QpJBSJYjJBmEZDlCkkFIliMkGYRkOUKS\nQUiWIyQZhGQ5QpJBSJYjJBmEZDlCkkFIliMkGYRkOUKSQUiWIyQZhGQ5QpJBSJYjJBmEZDlC\nkkFIliMkGYRkOUKSQUiWIyQZhGQ5QpJBSJYjJBmEZDlCkkFIliMkGYRkOUKSQUiWIyQZhGQ5\nQpJBSJYjJBmEZDlCkkFIliMkGYRkOUKSQUiWIyQZhGQ5QpJBSJYjJBmEZDlCkkFIliMkGYRk\nOUKSQUiWIyQZhGQ5QpJBSJYjJBmEZDlCkkFIliMkGYRkOUKSQUiWIyQZhGQ5QpJBSJYjJBmE\nZDlCkkFIliMkGYRkOUKSQUiWIyQZhGQ5QpJBSJYjJBmEZDlCkkFIliMkGYRkOUKSQUiWIyQZ\nhGQ5QpJBSJYjJBmEZDlCkkFIliMkGYRkOUKSQUiWIyQZhGQ5QpJBSJYjJBmEZDlCkkFIliMk\nGYRkOUKSQUiWIyQZhGQ5QpJBSJYjJBmEZDlCkkFIliMkGYRkOUKSQUiWIyQZhGQ5QpJBSJYj\nJBmEZDlCkkFIliMkGYRkOUKSQUiWIyQZhGQ5QpJBSJYjJBmEZDlCkkFIliMkGYRkOUKSQUiW\nIyQZhGQ5QpJBSJYjJBlrG9Lqvzz5RvwtCMlohCSj9pCe9I7P9YOUUiMfj70dIZmMkGTUHNK8\nxt5tzm2q97dm7Nmj6fmYGxKS0QhJRs0h7TJwgeNsMmKhu/hM834xNyQkoxGSjJpD6nOC9z/h\nAn/5u+vF3JCQjEZIMmoOqeV0x1lRd4e/fFbPyJWvDehX0qpWrM0EkS5CklFzSF/Z4lPH2ekE\nb3HFyJGRK9fMm1vyKx6RTEZIMmoO6V41+nerXtjw2k9XPrObmhNzQ57aGY2QZNT+7e8rWlTz\nViNUfb2q+39tMbcjJKMRkoy1+IHsovP3HtHatP52M1+IvRkhGY2QZPAWIcsRkgxCshwhySAk\nyxGSDEKyHCHJICTLEZIMQrIcIckgJMsRkgxCshwhySAkyxGSDEKyHCHJICTLEZIMQrIcIckg\nJMsRkgxCshwhySAkyxGSDEKyHCHJICTLEZIMQrIcIckgJMsRkgxCshwhySAkyxGSDEKyHCHJ\nICTLEZIMQrIcIckgJMsRkgxCshwhySAkyxGSDEKyHCHJICTLEZIMQrIcIckgJMsRkgxCshwh\nySAkyxGSDEKyHCHJICTLEZIMQrIcIckgJMsRkgxCshwhySAkyxGSDEKyHCHJICTLEZIMQrIc\nIckgJMsRkgxCshwhySAkyxGSDEKyHCHJICTLEZIMQrIcIckgJMsRkgxCshwhySAkyxGSDEKy\nHCHJICTLEZIMQrIcIckgJMsRkgxCshwhySAkyxGSDEKyHCHJICTLEZIMQrIcIckgJMsRkgxC\nshwhySAkyxGSDEKyHCHJICTLEZIMQrIcIckgJBu8PFc35yFt5beEJIKQbLBJcx+N6qWt9CYk\nEYRkg2Gz9TNbXaatzCUkEYRkA0JKHSHZgJBSR0g2IKTUEZINCCl1hGQDQkodIdmAkFJHSDYg\npNQRkg0IKXWEZANCSh0h2YCQUkdINiCk1BGSDQgpdYRkA0JKHSHZgJBSR0g2IKTUEZINCCl1\nhGQDQkodIdmAkFJHSDYgpNQRkg0IKXWEZANCSh0h2YCQUkdINiCk1OkhffmyD1MYgZCyR0ip\n00NqUM2THl4jPQIhZY+QUqeH9P6c3evVsFMXyI5ASNkjpNRFXiO9e+muPdTOV/5bcARCyh4h\npa7ymw0LfzlS9freP8RGIKTsEVLqKkJadtuBzWp4oXBmm9AIhJQ9QkpdJKQ/fKePaj5snvPG\ngeoMoREIKXuElDo9pDd+soVSX7xoqbfctsdAoREIKXuElDo9pB6q7/eeb1+5qE5oBELKHiGl\nTg9p7DXLyisL7hQagZCyR0ipC79Gevk978OLoiMQUvYIKXV6SCunqnnuxYVqymrBEQgpe4SU\nOj2kn6txr7kXfz9E/UpwBELKHiGlTg9pm/HFhX03FxyBkLJHSKnTQ2r+eXHhvILgCISUPUJK\nnR7SoGOLCzMGCY5ASNkjpNTpIU3tdb93sfLyhm8LjkBI2SOk1OkhLdxQDd9z/M791Yb/EhyB\nkLJHSKkL/Rxp0ffWV0oN+K7EwSwhpOwRUuoib1pte/uVT4RHIKTsEVLq+OUnNiCk1Okhtd06\nftTWAcERCCl7hJQ6PaTzlerVNyA4AiFlj5BSp4c0dO9XUxiBkLJHSKnTQyo8k8YIhJQ9Qkpd\n6BHp6TRGIKTsEVLq9JB+OCONEQgpe4SUOj2kj/c+9KH5C3yCIxBS9ggpdXpIqkxwBELKHiGl\nTk9m0uRp7QRHIKTsEVLqeGeDDQgpdZGQ/v3yUukRCCl7hJS6UEiPbafUg46z3+8lRyCk7BFS\n6vSQnm1s3dsN6d3Bjc93ePuuI6TsEVLq9JDGDX/zHe8RafHwCYIjEFL2CCl1ekjrn+v4ITmz\n+wmOQEjZI6TUhf705Q3FkH7DbxFatxBS6kLvtTu1GNIRIwRHIKTsEVLq9JCm93vBC2nJKUry\nTXeElD1CSp0e0jvDGkarUaOa1PBFgiMQUvYIKXWhnyMtPsr7LUIbHLVYcgRCyh4hpS76W4QW\nLZB8NPIQUvYIKXVr/167j2b9LfZ6QsoeIaVOD2n3krFd2MKb6t7Y6wkpe4SUuqr/Hql1SPId\nS//iYpLaK/afXRBS9ggpdXpIq3yfvnzCVz/qxB1DYm5ISNkjpNRVTeCk7yXf8Qf1ox5a6vmr\nunlp9N9eLJkxvWQCIYl5fHrIffp1N4WuGnugvta7syHdqQ7T73dyZyf2TGjwrQkp8HQnnto5\nz42qO+pDp/prJEJKx4mDD9KMmKJft++W+nX12+prdZ0NaY4ar91tV7WqkxM7c4A+XF9CCjzc\nqzN3XfXT5iG3882G7nTiWP0MnRAOaap+XfMJ+lp950N6RFu7tvMhba8PsIXlIS0NvDtvVCd/\n9/cru6v93iCk7kNIxqr+W4Su7+zdf9O/9xmE1G0IyVihf9gXOOCoLvxT88UTFSF1G0Iy1tq/\ns+GB4+fHXk9IcgjJWPw6rjwhJGPpIY3cYYxOaARCkkNIxtJDGtSslKpz/2uu9wiNQEhyCMlY\nekhLdj76xeXOR49/c69OvEWo0whJDiEZSw/piPb/Mft8R3AEQpJDSMbSQxpwVXHhvwYKjkBI\ncgjJWHpITecUF05sEhyBkOQQkrH0kL44JPgjsn/YYKTgCIQkh5CMpYd0d73aZI/99thU1d0u\nOAIhySEkY4X/GsXePZVSjbvNlRyBkOQQkrEi72xY89Y/31wtOwIhySEkY/GHxvKEkIzFHxrL\nE0IyFn9oLE8IyVj8obE8ISRj8YfG8oSQjMUfGssTQjIWf2gsTwjJWPyhsTwhJGPxh8byhJCM\nxR8ayxNCMhZ/aCxPCMlYoXd/v5zGCIQkh5CMpYfU86dpjEBIcgjJWHpIe3x9TQojEJIcQjKW\nHtKiSfvc+PwCn+AIhCSHkIxV/ZfoS/7+VUKSQ0jG0pM55NtT2/8wrOAIhCSHkIzF7/7OE0Iy\nVimkC5/wL/4k8cd4QwhJDiEZqxSSOi64EN9rQpJDSMYipDwhJGMRUp4QkrEIKU8IyViElCeE\nZCxCyhNCMhYh5QkhGasc0pgzPOpL/oXgCIQkh5CMVQ4pRHAEQpJDSMYqJXN9iOAIhCSHkIzF\ne+3yhJCMRUh5QkjGIqQ8ISRjEVKeEJKxCClPCMlYhJQnhGQsQsoTQjIWIeUJIRmLkPKEkIxF\nSHlCSMYipDwhJGMRUp4QkrEIKU8IyViElCeEZCxCyhNCMhYh5QkhGYuQ8oSQjEVIeUJIxiKk\nPCEkYxFSnhCSsQgpTwjJWISUJ4RkLELKE0IyFiHlCSEZi5DyhJCMRUh5Eg5py8Z+mkLaIT3V\nXx+u3zh9YnEhbdyk362xJTTp3vpa/6e6+XgKIqQ8CYc0fKcrNK1ph3RHiz7c4Z/TJxYX0qBd\n9fs176+vNRyur7Xc0c3HUxAh5UkkpAn6Wv/UQ+qrb+T0zoc0SV9rnaGvFU7X1/oSUgxCkkNI\nxiKkPCEkYxFSnhCSsQgpTwjJWISUJ4RkLELKE0IyFiHlCSEZi5DyhJCMRUh5QkjGIqQ8ISRj\nEVKeEJKxCClPCMlYhJQnhGQsQsoTQjIWIeUJIRmLkPKEkIxFSHlCSMYipDwhJGMRUp4QkrEI\nKU8IyViElCeEZCxCyhNCMhYh5QkhGYuQ8oSQjEVIeUJIxiKkPCEkYxFSnhCSsQgpTwjJWISU\nJ4RkLELKE0IyFiHlCSEZi5DyhJCMRUh5QkjGIqQ8ISRjEVKeEJKxCClPCMlYhJQnhGQsQsoT\nQjIWIeUJIRmLkPKEkIxFSHlCSMYipDwhJGMRUp4QkrEIKU8IyViElCeEZCxCyhNCMhYh5Qkh\nGYuQ8oSQjEVIeUJIxiKkPCEkYxFSnhCSsQgpTwjJWISUJ4RkLELKE0IyFiHlCSEZi5DyhJCM\nRUh5QkjGIqQ8ISRjEVKeEJKxCClPCMlYhJQnhGQsQsoTQjIWIeUJIRmLkPKEkIwlEtKS12Ou\nJCQ5hGSs2kP6874jdr54tb84K24rhCSHkIxVc0h/aFK9CuprS7xlQuomhGSsmkMaV7irbcUv\nCl/6xCGkbkNIxqo5pGGHex8fadx3dZWQ1sybW/IrQuqS5b+fq7n7Gn3tEELytT2mH5W5L8od\n/ZrVHFLhR/7FdWpmlZBeG9CvpFWtWJsJWufGHn00DaG1ekLy/UW1akelV4vc0a9ZzSEN3T+4\nPFmdx1M7SdcN1s+tXbbS17YhJN+L6mntbpc1yR39mtUc0sy6C1d6l22T1fePJSQ5hJRsXQrp\n/eFqD3+hbaZShCSHkJKtSyE57834fnHpjs0ISQ4hJVunQuosQuoaQkpGSEhESMkICYkIKRkh\nIREhJSMkJCKkZISERISUjJCQiJCSERISEVIyQkIiQkpGSEhESMkICYkIKRkhIREhJSMkJCKk\nZISERISUjJCQiJCSERISEVIyQkIiQkpGSEhESMkICYkIKRkhIREhJSMkJCKkZISERISUjJCQ\niJCSERISEVIyQkIiQkpGSEhESMkICYkIKRkhIREhJSMkJCKkZISERISUjJCQiJCSERISEVIy\nQkIiQkpGSEhESMkICYkIKRkhIREhJSMkJCKkZISERISUjJCQiJCSERISEVIyQkIiQkpGSEhE\nSMkICYkIKRkhIREhJSMkJCKkZISERISUjJCQiJCSERISEVIyQkIiQkpGSEhESMkICYkIKRkh\nIREhJSMkJCKkZISERISUjJCQiJCSERISEVIyQkIiQkpGSEhESMkICYkIKRkhIREhJSMkJCKk\nZISERISUjJCQiJCSERISEVIyQkIiQkpGSEhESMkIKeS+6SGPpz6V7nVyaO+OXdbJu+UkpKPW\n0/duO5GQeu6pb3P05A6PXzik8+r1G07cMXTc5+jH9vHwGXffWv8/LssypCkjDtIMPjH1qXSr\nVWpXbe/Gq3908n45CWnfVv1/3kCRkOp21Db5zZjjFw5pZg99Kpv20te23U4/ticO1q8bMWWt\n/yeXZRpS6EQYu86FdK22d4+scyFtoq/tIBPSL7WVF2KOXySkRn0j3wodvxPCIYWO3wRCygNC\nKiIkGYRUeSLEIaQAIUUQUuWJEIeQAoQUQUiVJ0IcQgoQUgQhVZ4IcQgpQEgRhFR5IsQhpAAh\nRRBS5YkQh5AChBRBSJUnQhxCChBSBCFVnghxCClASBGEVHkixCGkACFFEFLliRCHkAKEFEFI\nlSdCHEIKEFIEIVWeCHEIKUBIEYRUeSLEIaQAIUUQUuWJEIeQAoQUQUiVJ0IcQgoQUgQhVZ4I\ncQgpQEgRhFR5IsQhpAAhRRBS5YkQh5AChBRBSJUnQhxCChBSBCFVnghxCClASBGEVHkixCGk\nACFFEFLliRCHkAKEFEFIlSdCHEIKEFIEIVWeCHEIKUBIEYRUeSLEIaQAIUUQUuWJEIeQAoQU\nQUiVJ0IcQgoQUgQhVZ4IcQgpQEgRhFR5IsQhpAAhRRBS5YkQh5AChBRBSJUnQhxCChBSBCFV\nnghxCClASBGEVHkixCGkAGG54qoAAA7bSURBVCFFEFLliRCHkAKEFEFIlSdCHEIKEFIEIVWe\nCHEIKUBIEYRUeSLEIaQAIUUQUuWJEIeQAoQUQUiVJ0IcQgoQUgQhVZ4IcQgpQEgRhFR5IsQh\npAAhRRBS5YkQh5AChBRBSJUnQhxCChBSBCFVnghxCClASBGEVHkixCGkACFFEFLliRCHkAKE\nFEFIlSdCHEIKEFIEIVWeCHEIKUBIEYRUeSLEIaQAIUUQUuWJEIeQAoQUQUiVJ0IcQgoQUgQh\nVZ4IcQgpQEgRhFR5IsQhpAAhRRBS5YkQh5AChBRBSJUnQhxCChBSBCFVnghxCClASBGEVHki\nxCGkACFFEFLliRCHkAKEFEFIlSdCHEIKEFIEIVWeCHEIKUBIEYRUeSLEIaQAIUUQUuWJEIeQ\nAoQUQUiVJ0IcQgpYFFLbq3PvvPORNxJuRUiVJ0IcQgpYE9KS4wcq3/Czl8XdjpAqT4Q4hBSw\nJaSFm6gtppxx3nmnTRqiRi6JuSEhVZ4IcQgpYEtI0wq3FpdWX1x3XMwNCanyRIhDSAFbQho8\ntbx8yLCYGxJS5YkQh5ACtoRUOKe8fGZj5MrXBvQraVUrO9jEtEIfTY+e/TSNLfpaoVVfa+ij\nr9X31VbWq19PW+tbr9+wtUFf613Q11qa9LXm0FR6NutrTeGJ9Q5NLDRN1aLtXW+lTzM8sT6h\nibX00I9KQ2itviF0xELHr65JX1PhtebQWq/QxHpra71Ca80qdMTq9I0UwtOsD00z/D+2MWaa\nPTucWJ+Y49dHtWrXNSn9bo2hiRVCx7Zn6PgVptV68ldRc0gjDi4vT9g4cuWaeXNLHr6ho00s\nnKu74XZ97Zq79bUr79fX5jykr136sL52ib7y8KX62oOX62v3Xa2v3XWdvnbbjfrajbfpa9fd\npa9dfa++dvmDNU3soTn62v1X6mt3X6Ov3X6DvnbLLfoaxy/Q+eM3d2GtJ38VNYd0XN35K4Kl\nT36kZklNB8inmkNaOlq17j7lmKMn79JLjf1YckpA/tT+c6TPfjGq3vsxUuHLl68WnBCQR2v1\nFqHl/3zhhQUdfU8OsEj677UDLEBIgABCAgQQEiCAkAABhAQIICRAACEBAggJEEBIgABCAgQQ\nEiCAkAABhAQIICRAACEBAggJEJCPkFoU0EktmZyi+QhpyNnPm2L09KxnULLfflnPoGT66Kxn\nUHL2kExO0XyENOy6rGdQ8rUzs55ByRTJ3xS6ds78WtYzKLku7tf+poeQuoiQqiEkQuoiQqqG\nkAipiwipGkIipC4ipGoIiZC6iJCqISRC6iJCqoaQCKmLCKkaQiKkLiKkagiJkLqIkKohpHyE\ntNnNWc+gZM/ZWc+gZPr0rGdQMnvPrGdQcvNmmQybj5D+tSr5Nt3knU+znkHJkiVZz6Dk03ey\nnkHJqn9lMmw+QgIMR0iAAEICBBASIICQAAGEBAggJEAAIQECCAkQQEiAAEICBBASIICQAAGE\nBAggJEAAIQECDA/pB2paeWXpcSMKG05baMJUnJUn9dguo4lEprLk+OGNG0942oSpvPrdTRs3\nmPCsCVOp+ok0mR3Sc/XasfhstDrwnKmFTbL5d6GhqTjzR7dmF1JoKh9srMadflhDz79kP5W/\nr994+BmHFQpPZT+Vqp9IldEhrRo1UjsWv1A/cz/eoo7PfiofNW+/oCmrkMJTOVpd6H68Q+2b\n/VT2rHvc/XinOjj7qVT7RLqMDumndQ9qx2JU6wrvYvOBbZlP5YPjVzqZhRSeyvd3X+l+bGse\nkf1UTjvZ+7i6MDL7qVT7RLpMDumV5qOWlo/F8vrd/csp6tWsp+LLKqQqU3GcFYWvmDKVt9QB\nJkyl6txSZHJIu2/4oXYs/qmCX+N2hpqb9VR8WYVUZSqOc4H/BM+AqXw6b9vW50yYStXDlCKD\nQ/qNut3RjsUL6mj/8nx1Z9ZT8WUUUrWpOI817pzBLyyrMpW+Sh2ewROGyqlUPUxpMjekxf3H\nO+GQjvEvz1N3ZT0VXzYhVZ3KjU2jPzBjKidN36nHzt1fUsVUqh6mVJkb0sTe/wodiwVqsn95\nmvp91lPxZRNSlam0/Ujt828zpuKZ17Ltmsyn0sHcUmRsSA+o0998882/qklvfhR84rOGXfzL\nSaq7f5VmxVR8mYRUZSptU9Wxq82YSuBQNT/rqXQ4t/QYG9Lxqt2s4mfG9PJ+W/CaId3+S9Ir\np+LJJKQqUzlOZfPbyCum8ta23/Yvv6m6+7sNFVOp/n8sVcaGNP9ez81qr3v/5iz/0yvuZy5X\n3h+CuFSdlf1UPJmEVDmVO9RxGcyj6lSGNj7jfvxH797Ls56K/oluYmxIgeBp7kvK+xHS6rFq\nwlkT67bJ6LfY61N5bNasWfWD3Q/vZz6VzdSxs3zZvHNKn8pd9YWJp05pURdlMpPQVLRPdJMc\nheR8fMKIwkZHZ/D9qYqpnNv+zGFB5lMpPYl5PfOpOM8cMKB+vT3uyWQihASsAwgJEEBIgABC\nAgQQEiCAkAABhAQIICRAACEBAggJEEBIgABCAgQQEiCAkAABhAQIICRAACEBAggJEEBIgABC\nAgQQEiCAkAABhAQIICRAACEBAggJEEBIgABCAgQQEiCAkAABhAQIICRAACEBAgipBoeoN7t6\nl6Mbn++WcYRn0IH6Mc7phce8peIFCKkGcSf4ypN6FP9K89LjRhQ2nLbQX75R/Vp4nK6qbQYd\ncENa/dWB77pLxQsQUg1iTvD5o1uLIX02Wh14ztTCJt4fSf54/S8Lj9NVNc6gA25IzoL6qd5i\n8cJ6hFSDjk/wj5q3X9AUhPQL9TP34y3qePfjT9X9suN0WY0z6IAXknNow2tO+cJ2hFQD/wT/\nvylDCuvv96y3ft+XmgfNXDb0i47zwfErnWJIo1pXeBebD2xz1gz+nLe44rxt+/Te5rw17uKz\nB6xfGHH46+7SJLV0+sDmMc9+etyQlh1fcD9xgFo4bWDjlpe0j7NoxvDCBhP+GJ5BeVPj1FJ3\nfZX3t7wrNlW+VXEG2ra1yWjKe6Lvn76ro3sOmLbUD+l5NdMpX9iOkGrgneBvDOz9w2vO2ajp\nCcd5vH7wWRfvsn/fMcG1QUjL64M/Uz9Fveo8F5xrR6hDL73sG+po9+TrOeTsy09qHfi+40xW\ne5z14jU9h4+f9fzt6w1a6W18h1lPPrGnuiIY590RfWddP3toU/g1fXlTWkgVmyrfqjgDbdvl\n6zTanmj7py0+UT9k9hWHjy14u9o2YHOnfGE7QqqBd4JPVne6S/Pr3Zcee6rn3Ffdu6pQSP9U\nU/y1M9Rc51z1W2+x147exx8cuNq5ZPQ8d+lCdaHjTFNHuYsHq4Pcj8epJ72NT3IXP2zaOBjn\nqAZ3484brduHZlDelBZSxabKtyrOQNt2+TqNtifa/mmL+yjvoXFGsKuHqNe1C8sRUg3cE7yt\n76A2b3Fn9b7T03/a9FA4pBeKX+zPd8/Cqeplb7HvkMXaRlYuf8R7/TTNDc1xTlXXux8vUbd7\nG7/bu3oPtdAfZ4PR73j2Vh/rMyhvKhRSZFPlWxVnoG07PJmi8p5o+6ctrmnezFv6U7Crp/nj\ntV9YjpBq4J7gC9Vu/uI09dRSNd5b+nc0pGP8tfPUXc7+apG3eIHq8+2r3/I/e91X11Ou47wN\nzHe8x61H3Y9XqJu8jf/Nu8Vk9aI3ziLV7q/6DMqbCoUU2VT5VsUZaNvWJlOi7Ym2f9riW2pP\nb2l5sKu/9sYoXViOkGrgnuAL1H7+4jFq7ivqYH+xPhTSAjXZXztN/d7ZRS33lx85oEXV7ft/\njnOy2v43jz19ZRDSAsc7+90XIO0h/cu77Qy3B3+cUQ8GloamUNpUKKTIpsq3Ks5A23b5ujJt\nT7T90xb/WVys83f1OnWZdmE5QqqBe4K/U/wyfYR65l9qf2/p0/Aj0mcNu/hrk9xzt/h44Fox\nd3Ld5p8tbx7mPVF7qIOQvMcV5zD15+ARaVQHkwg2VQzp0+ohlW5VekQqbbt8XZm2J9r+aYtv\nBo9IH/OIVIGQauB9E6D/hv4LhzF1Sz/rMdJbejQckjOm16fuxzVDhnmvULTnZUepZ19X3/CW\nTu4gpDu8K3dQ7/rjbNDTfyiq+vYBd1POAcq76uWOQireqjgDbdvl68r0PSnvn7a4qtH/Dt2T\nwa6eHrw4Op3XSA4h1cQ7wb/jvvZxX3bXuSfwDnXuK4/Ve0dCulyd6X68VJ3lOOf6r/GfHnKt\n9/mj1YvL6rwf1PxpI3Vk1ZDGuYv/qNuy+F07dYq7+u7g8foEyptyr3/cXTqxWkjarYIZaNvW\nrtNoe6Ltn7a4i/9du0ODXZ0YfLtuIt+1cwipJt4J/vbg3qdce9bAVvdJ0m1qk/PnjJ3c5J5d\nj82aNat+sPvhfWf1WDXhrIl127iPS3/0HnqcVV9o/O7Fl0ztsXObM14dedPp/R5oGHrjJ1VC\n2mP8ZZdsrP47GGfxcHXENbOHFx7WJ6Bt6mm13aPPnDy2tUpI2q2CGWjb1q7TaHui7Z+2+EDd\nwJPOH7+b/yOztoHBz5EG8nMkh5Bq4r/j4I0jNmwYONF/zXHVlo0jTl3ZuJP32FPkntIfnzCi\nsNHRH7jXrxn0ee9mH3x/s159R852Xx69e+iAvrs94ZzVe/A7VUJa8P0hjVtd0z7OO0cNa1hv\n/2fDMyhvyrlmq+ZB0z8csnOVB7fyrYoz0LatbUFT3hN9/7TFm7dpHDB16TDvEfUFdaxTvrAd\nIUn5KHilXtW56oFOb0f0305EZtC5bcftieawhle1C9sR0tq7+mvev/S5QJ3X4S0+Xn/HTm8t\nnZCCGSRtO3lPyl4J3vb9Cu/+9hHS2numafBZV8xoGL6045t04V8DdXSyr1patrLLcwxmUGXb\noc12Yk/a8e+RwghJwB++PrCw0dS3425yTKf/fWpHId2rymr5wY03gyrbDm+2E3tSdHphnnYB\nQsqLJU+UvWf8Zq1DSIAAQgIEEBIggJAAAYQECCAkQAAhAQIICRBASIAAQgIEEBIggJAAAYQE\nCCAkQAAhAQIICRBASIAAQgIEEBIggJAAAYQECCAkQAAhAQIICRBASIAAQgIE/H8IDRQ7YyTe\nBQAAAABJRU5ErkJggg==",
      "text/plain": [
       "Plot with title “Histogram of log10(sample_sums(ps_good))”"
      ]
     },
     "metadata": {
      "image/png": {
       "height": 420,
       "width": 420
      }
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "ps_good <- subset_samples(ps, sample_sums(ps) > 1000)\n",
    "hist(log10(sample_sums(ps_good)), breaks = 50)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a1d4b790-94d5-484b-84fd-cbe8a29453ea",
   "metadata": {},
   "source": [
    "#### Estimate richness\n",
    "\n",
    "Use `estimate_richness` to calculate various diversity measures for each analysis.\n",
    "\n",
    "To read the documentation on this method, enter `?estimate_richness` in a Code Cell and run it."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "79696fa1-830f-4754-b053-72c35bf219d0",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "#  ?estimate_richness"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3966201c-f453-4cda-9b37-4175b6ca2683",
   "metadata": {},
   "outputs": [],
   "source": [
    "diversity = estimate_richness(ps_good)\n",
    "head(diversity)\n",
    "\n",
    "diversity[diversity$Observed == max(diversity$Observed),]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "341784b5-eb00-4dca-b35d-bb0539f71d9d",
   "metadata": {},
   "source": [
    "#### Normalise the data by rarefication"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a79a572b-cbff-4a0b-b513-a286efbb69ad",
   "metadata": {},
   "outputs": [],
   "source": [
    "ps_rare <- rarefy_even_depth(ps_good)\n",
    "div_rare <- estimate_richness(ps_rare)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4d01b9a4-2290-4528-88b3-2fa663cdb013",
   "metadata": {},
   "outputs": [],
   "source": [
    "merged_df <- merge(div_rare, metadata, by = 0, all.y = F)\n",
    "head(merged_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6425cc11-68c8-4b39-b12c-a024add8e3be",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width=12, repr.plot.height=8)\n",
    "ggplot(merged_df, aes(x=`sample_geo-loc-name`, y=Observed)) + geom_boxplot() + theme(text = element_text(size = 20))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "5bccc5d9-8c27-4e14-a87f-fdb4fa70c8cd",
   "metadata": {},
   "outputs": [
    {
     "ename": "ERROR",
     "evalue": "Error in parse(text = x, srcfile = src): <text>:2:0: unexpected end of input\n1: create.label?\n   ^\n",
     "output_type": "error",
     "traceback": [
      "Error in parse(text = x, srcfile = src): <text>:2:0: unexpected end of input\n1: create.label?\n   ^\nTraceback:\n"
     ]
    }
   ],
   "source": [
    "create.label?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "5f1313af-1804-4f96-b389-0cabd4245073",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "<table width=\"100%\" summary=\"page for create.label {SIAMCAT}\"><tr><td>create.label {SIAMCAT}</td><td style=\"text-align: right;\">R Documentation</td></tr></table>\n",
       "\n",
       "<h2>Create a label list</h2>\n",
       "\n",
       "<h3>Description</h3>\n",
       "\n",
       "<p>This function creates a label object from metadata\n",
       "or an atomic vector\n",
       "</p>\n",
       "\n",
       "\n",
       "<h3>Usage</h3>\n",
       "\n",
       "<pre>\n",
       "create.label(label, case,\n",
       "    meta=NULL, control=NULL,\n",
       "    p.lab = NULL, n.lab = NULL,\n",
       "    remove.meta.column=FALSE,\n",
       "    verbose=1)\n",
       "</pre>\n",
       "\n",
       "\n",
       "<h3>Arguments</h3>\n",
       "\n",
       "<table summary=\"R argblock\">\n",
       "<tr valign=\"top\"><td><code>label</code></td>\n",
       "<td>\n",
       "<p>named vector to create the label or the name of the metadata\n",
       "column that will be used to create the label</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>case</code></td>\n",
       "<td>\n",
       "<p>name of the group that will be used as a positive label. If the\n",
       "variable is binary, the other label will be used as a negative one. If the\n",
       "variable has multiple values, all the other values will be used a negative\n",
       "label (testing one vs rest).</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>meta</code></td>\n",
       "<td>\n",
       "<p>metadata dataframe object or an object of class\n",
       "sample_data-class</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>control</code></td>\n",
       "<td>\n",
       "<p>name of a label or vector with names that will be used as a\n",
       "negative label. All values that are nor equal to case and control will be\n",
       "dropped. Default to NULL in which case: If the variable is binary, the value\n",
       "not equal to case will be used as negative. If the variable has multiple\n",
       "values, all the values not equal to cases will be used a negative label\n",
       "(testing one vs rest).</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>p.lab</code></td>\n",
       "<td>\n",
       "<p>name of the positive group (useful mostly for visualizations).\n",
       "Default to NULL in which case the value of the positive group will be used.</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>n.lab</code></td>\n",
       "<td>\n",
       "<p>name of the negative group (useful mostly for visualizations).\n",
       "Default to NULL in which case the value of the negative group will be used\n",
       "for binary variables and &quot;rest&quot; will be used for variables with multiple\n",
       "values.</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>remove.meta.column</code></td>\n",
       "<td>\n",
       "<p>boolean indicating if the label column in the\n",
       "metadata should be retained. Please note that if this is set to\n",
       "<code>TRUE</code>, the function will return a list as result. Defaults to\n",
       "<code>FALSE</code></p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>verbose</code></td>\n",
       "<td>\n",
       "<p>integer, control output: <code>0</code> for no output at all,\n",
       "<code>1</code> for only information about progress and success, <code>2</code> for\n",
       "normal level of information and <code>3</code> for full debug information,\n",
       "defaults to <code>1</code></p>\n",
       "</td></tr>\n",
       "</table>\n",
       "\n",
       "\n",
       "<h3>Details</h3>\n",
       "\n",
       "<p>The function creates a list to be used as label in a SIAMCAT\n",
       "object. Mainly for interal use, but it can be used to customize your label\n",
       "(<code>p.lab</code> and <code>n.lab</code> will be used as labels during plotting,\n",
       "for example).\n",
       "</p>\n",
       "<p>The input for the function can be either a named vector encoding the label\n",
       "or the name of a column in the metadata (needs to be provided as well) which\n",
       "contains the label information.\n",
       "</p>\n",
       "\n",
       "\n",
       "<h3>Value</h3>\n",
       "\n",
       "<p>return either </p>\n",
       "\n",
       "<ul>\n",
       "<li><p> a list to be used in a SIMCAT object <strong>OR</strong>\n",
       "</p>\n",
       "</li>\n",
       "<li><p> a list with entries <code>meta</code> and <code>label</code>,\n",
       "if <code>remove.meta.column</code> is set to <code>TRUE</code>\n",
       "</p>\n",
       "</li></ul>\n",
       "\n",
       "\n",
       "\n",
       "<h3>Examples</h3>\n",
       "\n",
       "<pre>\n",
       "data('meta_crc_zeller')\n",
       "\n",
       "label &lt;- create.label(label='Group', case='CRC', meta=meta.crc.zeller)\n",
       "</pre>\n",
       "\n",
       "<hr /><div style=\"text-align: center;\">[Package <em>SIAMCAT</em> version 1.14.0 ]</div>"
      ],
      "text/latex": [
       "\\inputencoding{utf8}\n",
       "\\HeaderA{create.label}{Create a label list}{create.label}\n",
       "\\keyword{create.label}{create.label}\n",
       "%\n",
       "\\begin{Description}\\relax\n",
       "This function creates a label object from metadata\n",
       "or an atomic vector\n",
       "\\end{Description}\n",
       "%\n",
       "\\begin{Usage}\n",
       "\\begin{verbatim}\n",
       "create.label(label, case,\n",
       "    meta=NULL, control=NULL,\n",
       "    p.lab = NULL, n.lab = NULL,\n",
       "    remove.meta.column=FALSE,\n",
       "    verbose=1)\n",
       "\\end{verbatim}\n",
       "\\end{Usage}\n",
       "%\n",
       "\\begin{Arguments}\n",
       "\\begin{ldescription}\n",
       "\\item[\\code{label}] named vector to create the label or the name of the metadata\n",
       "column that will be used to create the label\n",
       "\n",
       "\\item[\\code{case}] name of the group that will be used as a positive label. If the\n",
       "variable is binary, the other label will be used as a negative one. If the\n",
       "variable has multiple values, all the other values will be used a negative\n",
       "label (testing one vs rest).\n",
       "\n",
       "\\item[\\code{meta}] metadata dataframe object or an object of class\n",
       "\\LinkA{sample\\_data-class}{sample.Rul.data.Rdash.class}\n",
       "\n",
       "\\item[\\code{control}] name of a label or vector with names that will be used as a\n",
       "negative label. All values that are nor equal to case and control will be\n",
       "dropped. Default to NULL in which case: If the variable is binary, the value\n",
       "not equal to case will be used as negative. If the variable has multiple\n",
       "values, all the values not equal to cases will be used a negative label\n",
       "(testing one vs rest).\n",
       "\n",
       "\\item[\\code{p.lab}] name of the positive group (useful mostly for visualizations).\n",
       "Default to NULL in which case the value of the positive group will be used.\n",
       "\n",
       "\\item[\\code{n.lab}] name of the negative group (useful mostly for visualizations).\n",
       "Default to NULL in which case the value of the negative group will be used\n",
       "for binary variables and \"rest\" will be used for variables with multiple\n",
       "values.\n",
       "\n",
       "\\item[\\code{remove.meta.column}] boolean indicating if the label column in the\n",
       "metadata should be retained. Please note that if this is set to\n",
       "\\code{TRUE}, the function will return a list as result. Defaults to\n",
       "\\code{FALSE}\n",
       "\n",
       "\\item[\\code{verbose}] integer, control output: \\code{0} for no output at all,\n",
       "\\code{1} for only information about progress and success, \\code{2} for\n",
       "normal level of information and \\code{3} for full debug information,\n",
       "defaults to \\code{1}\n",
       "\\end{ldescription}\n",
       "\\end{Arguments}\n",
       "%\n",
       "\\begin{Details}\\relax\n",
       "The function creates a list to be used as label in a SIAMCAT\n",
       "object. Mainly for interal use, but it can be used to customize your label\n",
       "(\\code{p.lab} and \\code{n.lab} will be used as labels during plotting,\n",
       "for example).\n",
       "\n",
       "The input for the function can be either a named vector encoding the label\n",
       "or the name of a column in the metadata (needs to be provided as well) which\n",
       "contains the label information.\n",
       "\\end{Details}\n",
       "%\n",
       "\\begin{Value}\n",
       "return either \\begin{itemize}\n",
       "\n",
       "\\item{} a list to be used in a SIMCAT object \\strong{OR}\n",
       "\\item{} a list with entries \\code{meta} and \\code{label},\n",
       "if \\code{remove.meta.column} is set to \\code{TRUE}\n",
       "\n",
       "\\end{itemize}\n",
       "\n",
       "\\end{Value}\n",
       "%\n",
       "\\begin{Examples}\n",
       "\\begin{ExampleCode}\n",
       "data('meta_crc_zeller')\n",
       "\n",
       "label <- create.label(label='Group', case='CRC', meta=meta.crc.zeller)\n",
       "\\end{ExampleCode}\n",
       "\\end{Examples}"
      ],
      "text/plain": [
       "create.label              package:SIAMCAT              R Documentation\n",
       "\n",
       "_\bC_\br_\be_\ba_\bt_\be _\ba _\bl_\ba_\bb_\be_\bl _\bl_\bi_\bs_\bt\n",
       "\n",
       "_\bD_\be_\bs_\bc_\br_\bi_\bp_\bt_\bi_\bo_\bn:\n",
       "\n",
       "     This function creates a label object from metadata or an atomic\n",
       "     vector\n",
       "\n",
       "_\bU_\bs_\ba_\bg_\be:\n",
       "\n",
       "     create.label(label, case,\n",
       "         meta=NULL, control=NULL,\n",
       "         p.lab = NULL, n.lab = NULL,\n",
       "         remove.meta.column=FALSE,\n",
       "         verbose=1)\n",
       "     \n",
       "_\bA_\br_\bg_\bu_\bm_\be_\bn_\bt_\bs:\n",
       "\n",
       "   label: named vector to create the label or the name of the metadata\n",
       "          column that will be used to create the label\n",
       "\n",
       "    case: name of the group that will be used as a positive label. If\n",
       "          the variable is binary, the other label will be used as a\n",
       "          negative one. If the variable has multiple values, all the\n",
       "          other values will be used a negative label (testing one vs\n",
       "          rest).\n",
       "\n",
       "    meta: metadata dataframe object or an object of class\n",
       "          sample_data-class\n",
       "\n",
       " control: name of a label or vector with names that will be used as a\n",
       "          negative label. All values that are nor equal to case and\n",
       "          control will be dropped. Default to NULL in which case: If\n",
       "          the variable is binary, the value not equal to case will be\n",
       "          used as negative. If the variable has multiple values, all\n",
       "          the values not equal to cases will be used a negative label\n",
       "          (testing one vs rest).\n",
       "\n",
       "   p.lab: name of the positive group (useful mostly for\n",
       "          visualizations). Default to NULL in which case the value of\n",
       "          the positive group will be used.\n",
       "\n",
       "   n.lab: name of the negative group (useful mostly for\n",
       "          visualizations). Default to NULL in which case the value of\n",
       "          the negative group will be used for binary variables and\n",
       "          \"rest\" will be used for variables with multiple values.\n",
       "\n",
       "remove.meta.column: boolean indicating if the label column in the\n",
       "          metadata should be retained. Please note that if this is set\n",
       "          to ‘TRUE’, the function will return a list as result.\n",
       "          Defaults to ‘FALSE’\n",
       "\n",
       " verbose: integer, control output: ‘0’ for no output at all, ‘1’ for\n",
       "          only information about progress and success, ‘2’ for normal\n",
       "          level of information and ‘3’ for full debug information,\n",
       "          defaults to ‘1’\n",
       "\n",
       "_\bD_\be_\bt_\ba_\bi_\bl_\bs:\n",
       "\n",
       "     The function creates a list to be used as label in a SIAMCAT\n",
       "     object. Mainly for interal use, but it can be used to customize\n",
       "     your label (‘p.lab’ and ‘n.lab’ will be used as labels during\n",
       "     plotting, for example).\n",
       "\n",
       "     The input for the function can be either a named vector encoding\n",
       "     the label or the name of a column in the metadata (needs to be\n",
       "     provided as well) which contains the label information.\n",
       "\n",
       "_\bV_\ba_\bl_\bu_\be:\n",
       "\n",
       "     return either\n",
       "\n",
       "        • a list to be used in a SIMCAT object *OR*\n",
       "\n",
       "        • a list with entries ‘meta’ and ‘label’, if\n",
       "          ‘remove.meta.column’ is set to ‘TRUE’\n",
       "\n",
       "_\bE_\bx_\ba_\bm_\bp_\bl_\be_\bs:\n",
       "\n",
       "     data('meta_crc_zeller')\n",
       "     \n",
       "     label <- create.label(label='Group', case='CRC', meta=meta.crc.zeller)\n",
       "     "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "?create.label"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "2bf62d98-2fb7-4e83-ab83-2e2b6978e5d7",
   "metadata": {},
   "outputs": [],
   "source": [
    "accessions_prePS=metadata$analysis_accession"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "b1665c11-4435-4cbc-afbb-b47dd40a4a5c",
   "metadata": {},
   "outputs": [],
   "source": [
    "accessions_postPS=sample_names(ps)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "f103e688-63f3-41ee-9e8d-40a13231a6a4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "78"
      ],
      "text/latex": [
       "78"
      ],
      "text/markdown": [
       "78"
      ],
      "text/plain": [
       "[1] 78"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "77"
      ],
      "text/latex": [
       "77"
      ],
      "text/markdown": [
       "77"
      ],
      "text/plain": [
       "[1] 77"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "length(accessions_prePS)\n",
    "length(accessions_postPS)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "bc6ee458-6b28-4691-9984-b0e066beac1d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "'MGYA00329556'"
      ],
      "text/latex": [
       "'MGYA00329556'"
      ],
      "text/markdown": [
       "'MGYA00329556'"
      ],
      "text/plain": [
       "[1] \"MGYA00329556\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "setdiff(accessions_prePS, accessions_postPS)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "6a1fd010-f349-4bfe-a8d2-cfab5b427980",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "75"
      ],
      "text/latex": [
       "75"
      ],
      "text/markdown": [
       "75"
      ],
      "text/plain": [
       "[1] 75"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "length(sample_names(ps_good))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fb2c8bcd-b39a-4faf-8681-103389fa3e5f",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.1.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
