{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "ca82f5f5-8773-415e-b332-005d2aa29a40",
   "metadata": {},
   "source": [
    "<img src=\"../_resources/mgnify_logo.png\" width=\"200px\">\n",
    "\n",
    "# Search for MGnify Studies or Samples, using MGnifyR\n",
    "\n",
    "The [MGnify API](https://www.ebi.ac.uk/metagenomics/api/v1) returns data and relationships as JSON. \n",
    "[MGnifyR](https://github.com/beadyallen/MGnifyR) is a package to help you read MGnify data into your R analyses.\n",
    "\n",
    "**This example shows you how to perform a search of MGnify Studies or Samples**\n",
    "\n",
    "You can find all of the other \"API endpoints\" using the [Browsable API interface in your web browser](https://www.ebi.ac.uk/metagenomics/api/v1).\n",
    "This interface also lets you inspect the kinds of Filters that can be created for each list.\n",
    "\n",
    "This is an interactive code notebook (a Jupyter Notebook).\n",
    "To run this code, click into each cell and press the â–¶ button in the top toolbar, or press `shift+enter`.\n",
    "\n",
    "---"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "bb99acc5-6bfe-4d93-b258-6885505df8db",
   "metadata": {},
   "outputs": [],
   "source": [
    "library(vegan)\n",
    "library(ggplot2)\n",
    "library(phyloseq)\n",
    "library(MGnifyR)\n",
    "\n",
    "mg <- mgnify_client(usecache = T, cache_dir = '/tmp/mgnify_cache')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1205a39a-25d1-4a04-b7cf-7bf5cfec79b9",
   "metadata": {},
   "source": [
    "## Contents\n",
    "- [Example: Find Polar Samples](#Example:-find-Polar-samples)\n",
    "- [Example: Find Wastewater Samples](#Example:-find-Wastewater-studies)\n",
    "- [More Sample filters](#More-Sample-filters)\n",
    "- [More Study filters](#More-Study-filters)\n",
    "- [Example: Filtering Samples both API-side and client-side](#Example:-adding-additional-filters-to-the-data-frame)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b99a5b00-266b-4d89-9a3e-ce9b52c052e5",
   "metadata": {},
   "source": [
    "### Documentation for `mgnify_query`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "f4396d6a-57ca-4a05-9315-c8b661b8ea4c",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "<table width=\"100%\" summary=\"page for mgnify_query {MGnifyR}\"><tr><td>mgnify_query {MGnifyR}</td><td style=\"text-align: right;\">R Documentation</td></tr></table>\n",
       "\n",
       "<h2>#' Search MGnify database for studies, samples and runs</h2>\n",
       "\n",
       "<h3>Description</h3>\n",
       "\n",
       "<p><code>mgnify_query</code> is a flexible query function, harnessing the &quot;full&quot; power of the JSONAPI MGnify\n",
       "search filters. Search results may be filtered by metadata value, associated study/sample/analyese etc. Details\n",
       "of the capabilities may be found [here](https://emg-docs.readthedocs.io/en/latest/api.html#customising-queries). Currently,\n",
       "the following filters are available (based on examination of the Python source code):\n",
       "</p>\n",
       "\n",
       "<ul>\n",
       "<li><p><strong>Studies</strong> : accession, biome_name, lineage, centre_name\n",
       "</p>\n",
       "</li>\n",
       "<li><p><strong>Samples</strong> : accession, experiment_type, biome_name,\n",
       "lineage, geo_loc_name, latitude_gte, latitude_lte,\n",
       "longitude_gte, longitude_lte, species, instrument_model, instrument_platform,\n",
       "metadata_key, metadata_value_gte, metadata_value_lte, metadata_value,\n",
       "environment_material, environment_feature, study_accession\n",
       "</p>\n",
       "</li>\n",
       "<li><p><strong>Runs</strong> accession, experiment_type, biome_name, lineage, species,\n",
       "instrument_platform, instrument_model\n",
       "</p>\n",
       "</li></ul>\n",
       "\n",
       "<p>Unfortunately (from testing) it appears that some of these filters don't work as expected, so it is important\n",
       "to check the results returned match up with what's expected. Even more unfortunately if there's an error in the\n",
       "parameter specification, the query will run as if no filter parameters were present at all. Thus the\n",
       "result will appear superficially correct but will infact correspond to something completely different. This beahviour\n",
       "will hopefully be fixed in future incarnations of the API, but for now users should double check returned\n",
       "values.\n",
       "</p>\n",
       "\n",
       "\n",
       "<h3>Usage</h3>\n",
       "\n",
       "<pre>\n",
       "mgnify_query(client, qtype = \"samples\", accession = NULL,\n",
       "  asDataFrame = T, maxhits = 200, usecache = F, ...)\n",
       "</pre>\n",
       "\n",
       "\n",
       "<h3>Arguments</h3>\n",
       "\n",
       "<table summary=\"R argblock\">\n",
       "<tr valign=\"top\"><td><code>qtype</code></td>\n",
       "<td>\n",
       "<p>Type of objects to query. One of <code>studies</code>,<code>samples</code>,<code>runs</code> or\n",
       "<code>analyses</code></p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>accession</code></td>\n",
       "<td>\n",
       "<p>Either a single known MGnify accession identifier (of type <code>qtype</code>), or a list/vector\n",
       "of accessions to query. Note that multiple values only work for samples, runs and assemblies ... not sure why.</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>asDataFrame</code></td>\n",
       "<td>\n",
       "<p>Boolean flag to choose whether to return the results as a data.frame or leave as a nested list. In\n",
       "most cases, <code>asDataFrame = TRUE</code> will make the most sense.</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>maxhits</code></td>\n",
       "<td>\n",
       "<p>determines the maximum number of results to return. The actual number of results will actually be higher than <code>maxhits</code>,\n",
       "as clipping only occurs on pagination page boundaries. To disable the limit, set <code>maxhits</code> &lt; 0</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>usecache</code></td>\n",
       "<td>\n",
       "<p>Whether to cache the result - and reuse any existing cache entry instead of issuing a\n",
       "new callout. In generl the use of caching for queries is discouraged, as new data is being uploaded to MGnify\n",
       "all the time, which might potentially be missed. However, for some purposes (such as analysis reproducibility)\n",
       "caching makes sense.</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>...</code></td>\n",
       "<td>\n",
       "<p>Remaining parameter key/value pairs may be supplied to filter the returned values. Available options differ\n",
       "between <code>qtypes</code>.See discussion above for details.</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>mgnify_client</code></td>\n",
       "<td>\n",
       "<p>Client instance</p>\n",
       "</td></tr>\n",
       "</table>\n",
       "\n",
       "\n",
       "<h3>Details</h3>\n",
       "\n",
       "<p>It is currently not possible to combine queries of the same type in a single call (for example to search for samples\n",
       "<em>between</em> latitude). However, it is possible to run multiple queries and combine the results using set operations in R to get the\n",
       "desired behaviour.\n",
       "</p>\n",
       "\n",
       "\n",
       "<h3>Value</h3>\n",
       "\n",
       "<p>A nested list or data.frame (depending on <code>asDataFrame</code>) containing the results of the query.\n",
       "</p>\n",
       "\n",
       "\n",
       "<h3>Examples</h3>\n",
       "\n",
       "<pre>\n",
       "mg &lt;- mgnify_client(cache_dir=\"/tmp/mgcache\")\n",
       "\n",
       "#Get a list of studies from the Agricultural Wastewater :\n",
       "agwaste_studies &lt;- mgnify_query(mg, \"studies\", biome_name=\"Agricultural wastewater\")\n",
       "\n",
       "#Get all samples from a particular study\n",
       "samps &lt;- mgnify_query(mg, \"samples\", study_accession=\"MGYS00004521\")\n",
       "\n",
       "#Search for all polar samples\n",
       "samps_np &lt;- mgnify_query(mg, \"samples\", latitude_gte=66, maxhits=-1)\n",
       "samps_sp &lt;- mgnify_query(mg, \"samples\", latitude_lte=-66, maxhits=-1)\n",
       "samps_polar &lt;- rbind(samps_np, samps_sp)\n",
       "\n",
       "</pre>\n",
       "\n",
       "<hr /><div style=\"text-align: center;\">[Package <em>MGnifyR</em> version 0.1.0 ]</div>"
      ],
      "text/latex": [
       "\\inputencoding{utf8}\n",
       "\\HeaderA{mgnify\\_query}{\\#' Search MGnify database for studies, samples and runs}{mgnify.Rul.query}\n",
       "%\n",
       "\\begin{Description}\\relax\n",
       "\\code{mgnify\\_query} is a flexible query function, harnessing the \"full\" power of the JSONAPI MGnify\n",
       "search filters. Search results may be filtered by metadata value, associated study/sample/analyese etc. Details\n",
       "of the capabilities may be found [here](https://emg-docs.readthedocs.io/en/latest/api.html\\#customising-queries). Currently,\n",
       "the following filters are available (based on examination of the Python source code):\n",
       "\\begin{itemize}\n",
       "\n",
       "\\item{} \\strong{Studies} : accession, biome\\_name, lineage, centre\\_name\n",
       "\\item{} \\strong{Samples} : accession, experiment\\_type, biome\\_name,\n",
       "lineage, geo\\_loc\\_name, latitude\\_gte, latitude\\_lte,\n",
       "longitude\\_gte, longitude\\_lte, species, instrument\\_model, instrument\\_platform,\n",
       "metadata\\_key, metadata\\_value\\_gte, metadata\\_value\\_lte, metadata\\_value,\n",
       "environment\\_material, environment\\_feature, study\\_accession\n",
       "\\item{} \\strong{Runs} accession, experiment\\_type, biome\\_name, lineage, species,\n",
       "instrument\\_platform, instrument\\_model\n",
       "\n",
       "\\end{itemize}\n",
       "\n",
       "Unfortunately (from testing) it appears that some of these filters don't work as expected, so it is important\n",
       "to check the results returned match up with what's expected. Even more unfortunately if there's an error in the\n",
       "parameter specification, the query will run as if no filter parameters were present at all. Thus the\n",
       "result will appear superficially correct but will infact correspond to something completely different. This beahviour\n",
       "will hopefully be fixed in future incarnations of the API, but for now users should double check returned\n",
       "values.\n",
       "\\end{Description}\n",
       "%\n",
       "\\begin{Usage}\n",
       "\\begin{verbatim}\n",
       "mgnify_query(client, qtype = \"samples\", accession = NULL,\n",
       "  asDataFrame = T, maxhits = 200, usecache = F, ...)\n",
       "\\end{verbatim}\n",
       "\\end{Usage}\n",
       "%\n",
       "\\begin{Arguments}\n",
       "\\begin{ldescription}\n",
       "\\item[\\code{qtype}] Type of objects to query. One of \\code{studies},\\code{samples},\\code{runs} or\n",
       "\\code{analyses}\n",
       "\n",
       "\\item[\\code{accession}] Either a single known MGnify accession identifier (of type \\code{qtype}), or a list/vector\n",
       "of accessions to query. Note that multiple values only work for samples, runs and assemblies ... not sure why.\n",
       "\n",
       "\\item[\\code{asDataFrame}] Boolean flag to choose whether to return the results as a data.frame or leave as a nested list. In\n",
       "most cases, \\code{asDataFrame = TRUE} will make the most sense.\n",
       "\n",
       "\\item[\\code{maxhits}] determines the maximum number of results to return. The actual number of results will actually be higher than \\code{maxhits},\n",
       "as clipping only occurs on pagination page boundaries. To disable the limit, set \\code{maxhits} < 0\n",
       "\n",
       "\\item[\\code{usecache}] Whether to cache the result - and reuse any existing cache entry instead of issuing a\n",
       "new callout. In generl the use of caching for queries is discouraged, as new data is being uploaded to MGnify\n",
       "all the time, which might potentially be missed. However, for some purposes (such as analysis reproducibility)\n",
       "caching makes sense.\n",
       "\n",
       "\\item[\\code{...}] Remaining parameter key/value pairs may be supplied to filter the returned values. Available options differ\n",
       "between \\code{qtypes}.See discussion above for details.\n",
       "\n",
       "\\item[\\code{mgnify\\_client}] Client instance\n",
       "\\end{ldescription}\n",
       "\\end{Arguments}\n",
       "%\n",
       "\\begin{Details}\\relax\n",
       "It is currently not possible to combine queries of the same type in a single call (for example to search for samples\n",
       "\\emph{between} latitude). However, it is possible to run multiple queries and combine the results using set operations in R to get the\n",
       "desired behaviour.\n",
       "\\end{Details}\n",
       "%\n",
       "\\begin{Value}\n",
       "A nested list or data.frame (depending on \\code{asDataFrame}) containing the results of the query.\n",
       "\\end{Value}\n",
       "%\n",
       "\\begin{Examples}\n",
       "\\begin{ExampleCode}\n",
       "mg <- mgnify_client(cache_dir=\"/tmp/mgcache\")\n",
       "\n",
       "#Get a list of studies from the Agricultural Wastewater :\n",
       "agwaste_studies <- mgnify_query(mg, \"studies\", biome_name=\"Agricultural wastewater\")\n",
       "\n",
       "#Get all samples from a particular study\n",
       "samps <- mgnify_query(mg, \"samples\", study_accession=\"MGYS00004521\")\n",
       "\n",
       "#Search for all polar samples\n",
       "samps_np <- mgnify_query(mg, \"samples\", latitude_gte=66, maxhits=-1)\n",
       "samps_sp <- mgnify_query(mg, \"samples\", latitude_lte=-66, maxhits=-1)\n",
       "samps_polar <- rbind(samps_np, samps_sp)\n",
       "\n",
       "\\end{ExampleCode}\n",
       "\\end{Examples}"
      ],
      "text/plain": [
       "mgnify_query              package:MGnifyR              R Documentation\n",
       "\n",
       "#' _\bS_\be_\ba_\br_\bc_\bh _\bM_\bG_\bn_\bi_\bf_\by _\bd_\ba_\bt_\ba_\bb_\ba_\bs_\be _\bf_\bo_\br _\bs_\bt_\bu_\bd_\bi_\be_\bs, _\bs_\ba_\bm_\bp_\bl_\be_\bs _\ba_\bn_\bd _\br_\bu_\bn_\bs\n",
       "\n",
       "_\bD_\be_\bs_\bc_\br_\bi_\bp_\bt_\bi_\bo_\bn:\n",
       "\n",
       "     â€˜mgnify_queryâ€™ is a flexible query function, harnessing the \"full\"\n",
       "     power of the JSONAPI MGnify search filters. Search results may be\n",
       "     filtered by metadata value, associated study/sample/analyese etc.\n",
       "     Details of the capabilities may be found\n",
       "     [here](https://emg-docs.readthedocs.io/en/latest/api.html#customising-queries).\n",
       "     Currently, the following filters are available (based on\n",
       "     examination of the Python source code):\n",
       "\n",
       "        â€¢ *Studies* : accession, biome_name, lineage, centre_name\n",
       "\n",
       "        â€¢ *Samples* : accession, experiment_type, biome_name, lineage,\n",
       "          geo_loc_name, latitude_gte, latitude_lte, longitude_gte,\n",
       "          longitude_lte, species, instrument_model,\n",
       "          instrument_platform, metadata_key, metadata_value_gte,\n",
       "          metadata_value_lte, metadata_value, environment_material,\n",
       "          environment_feature, study_accession\n",
       "\n",
       "        â€¢ *Runs* accession, experiment_type, biome_name, lineage,\n",
       "          species, instrument_platform, instrument_model\n",
       "\n",
       "     Unfortunately (from testing) it appears that some of these filters\n",
       "     don't work as expected, so it is important to check the results\n",
       "     returned match up with what's expected. Even more unfortunately if\n",
       "     there's an error in the parameter specification, the query will\n",
       "     run as if no filter parameters were present at all. Thus the\n",
       "     result will appear superficially correct but will infact\n",
       "     correspond to something completely different. This beahviour will\n",
       "     hopefully be fixed in future incarnations of the API, but for now\n",
       "     users should double check returned values.\n",
       "\n",
       "_\bU_\bs_\ba_\bg_\be:\n",
       "\n",
       "     mgnify_query(client, qtype = \"samples\", accession = NULL,\n",
       "       asDataFrame = T, maxhits = 200, usecache = F, ...)\n",
       "     \n",
       "_\bA_\br_\bg_\bu_\bm_\be_\bn_\bt_\bs:\n",
       "\n",
       "   qtype: Type of objects to query. One of â€˜studiesâ€™,â€˜samplesâ€™,â€˜runsâ€™\n",
       "          or â€˜analysesâ€™\n",
       "\n",
       "accession: Either a single known MGnify accession identifier (of type\n",
       "          â€˜qtypeâ€™), or a list/vector of accessions to query. Note that\n",
       "          multiple values only work for samples, runs and assemblies\n",
       "          ... not sure why.\n",
       "\n",
       "asDataFrame: Boolean flag to choose whether to return the results as a\n",
       "          data.frame or leave as a nested list. In most cases,\n",
       "          â€˜asDataFrame = TRUEâ€™ will make the most sense.\n",
       "\n",
       " maxhits: determines the maximum number of results to return. The\n",
       "          actual number of results will actually be higher than\n",
       "          â€˜maxhitsâ€™, as clipping only occurs on pagination page\n",
       "          boundaries. To disable the limit, set â€˜maxhitsâ€™ < 0\n",
       "\n",
       "usecache: Whether to cache the result - and reuse any existing cache\n",
       "          entry instead of issuing a new callout. In generl the use of\n",
       "          caching for queries is discouraged, as new data is being\n",
       "          uploaded to MGnify all the time, which might potentially be\n",
       "          missed. However, for some purposes (such as analysis\n",
       "          reproducibility) caching makes sense.\n",
       "\n",
       "     ...: Remaining parameter key/value pairs may be supplied to filter\n",
       "          the returned values. Available options differ between\n",
       "          â€˜qtypesâ€™.See discussion above for details.\n",
       "\n",
       "mgnify_client: Client instance\n",
       "\n",
       "_\bD_\be_\bt_\ba_\bi_\bl_\bs:\n",
       "\n",
       "     It is currently not possible to combine queries of the same type\n",
       "     in a single call (for example to search for samples _between_\n",
       "     latitude). However, it is possible to run multiple queries and\n",
       "     combine the results using set operations in R to get the desired\n",
       "     behaviour.\n",
       "\n",
       "_\bV_\ba_\bl_\bu_\be:\n",
       "\n",
       "     A nested list or data.frame (depending on â€˜asDataFrameâ€™)\n",
       "     containing the results of the query.\n",
       "\n",
       "_\bE_\bx_\ba_\bm_\bp_\bl_\be_\bs:\n",
       "\n",
       "     mg <- mgnify_client(cache_dir=\"/tmp/mgcache\")\n",
       "     \n",
       "     #Get a list of studies from the Agricultural Wastewater :\n",
       "     agwaste_studies <- mgnify_query(mg, \"studies\", biome_name=\"Agricultural wastewater\")\n",
       "     \n",
       "     #Get all samples from a particular study\n",
       "     samps <- mgnify_query(mg, \"samples\", study_accession=\"MGYS00004521\")\n",
       "     \n",
       "     #Search for all polar samples\n",
       "     samps_np <- mgnify_query(mg, \"samples\", latitude_gte=66, maxhits=-1)\n",
       "     samps_sp <- mgnify_query(mg, \"samples\", latitude_lte=-66, maxhits=-1)\n",
       "     samps_polar <- rbind(samps_np, samps_sp)\n",
       "     "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "?mgnify_query"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "31501a56-24ae-4045-be4d-0df51293cf61",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Example: find Polar samples "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "83e3d3f1-701b-4528-a6a6-44ac8637f385",
   "metadata": {},
   "outputs": [],
   "source": [
    "samps_np <- mgnify_query(mg, \"samples\", latitude_gte=88, maxhits=-1)\n",
    "samps_sp <- mgnify_query(mg, \"samples\", latitude_lte=-88, maxhits=-1)\n",
    "samps_polar <- rbind(samps_np, samps_sp)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "bbeaf601-0b08-4954-a169-41c2a5971f18",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 6 Ã— 70</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>biosample</th><th scope=col>latitude</th><th scope=col>longitude</th><th scope=col>accession</th><th scope=col>analysis-completed</th><th scope=col>collection-date</th><th scope=col>geo-loc-name</th><th scope=col>sample-desc</th><th scope=col>environment-biome</th><th scope=col>environment-feature</th><th scope=col>â‹¯</th><th scope=col>pcr conditions</th><th scope=col>host common name</th><th scope=col>host age</th><th scope=col>host body habitat</th><th scope=col>host diet</th><th scope=col>host genotype</th><th scope=col>host phenotype</th><th scope=col>host sex</th><th scope=col>sample volume or weight for DNA extraction</th><th scope=col>chemical administration</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>â‹¯</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>ERS1568988</th><td>SAMEA98155168 </td><td>89.9903</td><td>-89.2525 </td><td>ERS1568988</td><td>2017-03-13</td><td>2015-09-07</td><td>Arctic Ocean</td><td>Arctic seawater metagenome, 5 m  </td><td>marine biome</td><td>marine water body</td><td>â‹¯</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td></tr>\n",
       "\t<tr><th scope=row>ERS1972379</th><td>SAMEA104347401</td><td>88.4072</td><td>-176.7614</td><td>ERS1972379</td><td>2017-11-20</td><td>NA        </td><td>NA          </td><td>Station 31 sea ice, depth 0.1m   </td><td>marine biome</td><td>marine water body</td><td>â‹¯</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td></tr>\n",
       "\t<tr><th scope=row>ERS1972376</th><td>SAMEA104347398</td><td>88.4072</td><td>-176.7614</td><td>ERS1972376</td><td>2017-11-20</td><td>NA        </td><td>NA          </td><td>Station 31 sea ice, depth 0.5m   </td><td>marine biome</td><td>marine water body</td><td>â‹¯</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td></tr>\n",
       "\t<tr><th scope=row>ERS1568987</th><td>SAMEA98154418 </td><td>89.9903</td><td>-89.2525 </td><td>ERS1568987</td><td>2017-03-13</td><td>2015-09-07</td><td>Arctic Ocean</td><td>Arctic seawater metagenome, 1.5 m</td><td>marine biome</td><td>marine water body</td><td>â‹¯</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td></tr>\n",
       "\t<tr><th scope=row>ERS1972377</th><td>SAMEA104347399</td><td>88.4072</td><td>-176.7614</td><td>ERS1972377</td><td>2017-11-20</td><td>NA        </td><td>NA          </td><td>Station 31 sea ice, depth 0.5m   </td><td>marine biome</td><td>marine water body</td><td>â‹¯</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td></tr>\n",
       "\t<tr><th scope=row>ERS1972391</th><td>SAMEA104347413</td><td>89.9903</td><td>-89.2525 </td><td>ERS1972391</td><td>2017-11-20</td><td>NA        </td><td>NA          </td><td>Station 33 sea ice, depth 0.1m   </td><td>marine biome</td><td>marine water body</td><td>â‹¯</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 6 Ã— 70\n",
       "\\begin{tabular}{r|lllllllllllllllllllll}\n",
       "  & biosample & latitude & longitude & accession & analysis-completed & collection-date & geo-loc-name & sample-desc & environment-biome & environment-feature & â‹¯ & pcr conditions & host common name & host age & host body habitat & host diet & host genotype & host phenotype & host sex & sample volume or weight for DNA extraction & chemical administration\\\\\n",
       "  & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & â‹¯ & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\tERS1568988 & SAMEA98155168  & 89.9903 & -89.2525  & ERS1568988 & 2017-03-13 & 2015-09-07 & Arctic Ocean & Arctic seawater metagenome, 5 m   & marine biome & marine water body & â‹¯ & NA & NA & NA & NA & NA & NA & NA & NA & NA & NA\\\\\n",
       "\tERS1972379 & SAMEA104347401 & 88.4072 & -176.7614 & ERS1972379 & 2017-11-20 & NA         & NA           & Station 31 sea ice, depth 0.1m    & marine biome & marine water body & â‹¯ & NA & NA & NA & NA & NA & NA & NA & NA & NA & NA\\\\\n",
       "\tERS1972376 & SAMEA104347398 & 88.4072 & -176.7614 & ERS1972376 & 2017-11-20 & NA         & NA           & Station 31 sea ice, depth 0.5m    & marine biome & marine water body & â‹¯ & NA & NA & NA & NA & NA & NA & NA & NA & NA & NA\\\\\n",
       "\tERS1568987 & SAMEA98154418  & 89.9903 & -89.2525  & ERS1568987 & 2017-03-13 & 2015-09-07 & Arctic Ocean & Arctic seawater metagenome, 1.5 m & marine biome & marine water body & â‹¯ & NA & NA & NA & NA & NA & NA & NA & NA & NA & NA\\\\\n",
       "\tERS1972377 & SAMEA104347399 & 88.4072 & -176.7614 & ERS1972377 & 2017-11-20 & NA         & NA           & Station 31 sea ice, depth 0.5m    & marine biome & marine water body & â‹¯ & NA & NA & NA & NA & NA & NA & NA & NA & NA & NA\\\\\n",
       "\tERS1972391 & SAMEA104347413 & 89.9903 & -89.2525  & ERS1972391 & 2017-11-20 & NA         & NA           & Station 33 sea ice, depth 0.1m    & marine biome & marine water body & â‹¯ & NA & NA & NA & NA & NA & NA & NA & NA & NA & NA\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 6 Ã— 70\n",
       "\n",
       "| <!--/--> | biosample &lt;chr&gt; | latitude &lt;chr&gt; | longitude &lt;chr&gt; | accession &lt;chr&gt; | analysis-completed &lt;chr&gt; | collection-date &lt;chr&gt; | geo-loc-name &lt;chr&gt; | sample-desc &lt;chr&gt; | environment-biome &lt;chr&gt; | environment-feature &lt;chr&gt; | â‹¯ â‹¯ | pcr conditions &lt;chr&gt; | host common name &lt;chr&gt; | host age &lt;chr&gt; | host body habitat &lt;chr&gt; | host diet &lt;chr&gt; | host genotype &lt;chr&gt; | host phenotype &lt;chr&gt; | host sex &lt;chr&gt; | sample volume or weight for DNA extraction &lt;chr&gt; | chemical administration &lt;chr&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| ERS1568988 | SAMEA98155168  | 89.9903 | -89.2525  | ERS1568988 | 2017-03-13 | 2015-09-07 | Arctic Ocean | Arctic seawater metagenome, 5 m   | marine biome | marine water body | â‹¯ | NA | NA | NA | NA | NA | NA | NA | NA | NA | NA |\n",
       "| ERS1972379 | SAMEA104347401 | 88.4072 | -176.7614 | ERS1972379 | 2017-11-20 | NA         | NA           | Station 31 sea ice, depth 0.1m    | marine biome | marine water body | â‹¯ | NA | NA | NA | NA | NA | NA | NA | NA | NA | NA |\n",
       "| ERS1972376 | SAMEA104347398 | 88.4072 | -176.7614 | ERS1972376 | 2017-11-20 | NA         | NA           | Station 31 sea ice, depth 0.5m    | marine biome | marine water body | â‹¯ | NA | NA | NA | NA | NA | NA | NA | NA | NA | NA |\n",
       "| ERS1568987 | SAMEA98154418  | 89.9903 | -89.2525  | ERS1568987 | 2017-03-13 | 2015-09-07 | Arctic Ocean | Arctic seawater metagenome, 1.5 m | marine biome | marine water body | â‹¯ | NA | NA | NA | NA | NA | NA | NA | NA | NA | NA |\n",
       "| ERS1972377 | SAMEA104347399 | 88.4072 | -176.7614 | ERS1972377 | 2017-11-20 | NA         | NA           | Station 31 sea ice, depth 0.5m    | marine biome | marine water body | â‹¯ | NA | NA | NA | NA | NA | NA | NA | NA | NA | NA |\n",
       "| ERS1972391 | SAMEA104347413 | 89.9903 | -89.2525  | ERS1972391 | 2017-11-20 | NA         | NA           | Station 33 sea ice, depth 0.1m    | marine biome | marine water body | â‹¯ | NA | NA | NA | NA | NA | NA | NA | NA | NA | NA |\n",
       "\n"
      ],
      "text/plain": [
       "           biosample      latitude longitude accession  analysis-completed\n",
       "ERS1568988 SAMEA98155168  89.9903  -89.2525  ERS1568988 2017-03-13        \n",
       "ERS1972379 SAMEA104347401 88.4072  -176.7614 ERS1972379 2017-11-20        \n",
       "ERS1972376 SAMEA104347398 88.4072  -176.7614 ERS1972376 2017-11-20        \n",
       "ERS1568987 SAMEA98154418  89.9903  -89.2525  ERS1568987 2017-03-13        \n",
       "ERS1972377 SAMEA104347399 88.4072  -176.7614 ERS1972377 2017-11-20        \n",
       "ERS1972391 SAMEA104347413 89.9903  -89.2525  ERS1972391 2017-11-20        \n",
       "           collection-date geo-loc-name sample-desc                      \n",
       "ERS1568988 2015-09-07      Arctic Ocean Arctic seawater metagenome, 5 m  \n",
       "ERS1972379 NA              NA           Station 31 sea ice, depth 0.1m   \n",
       "ERS1972376 NA              NA           Station 31 sea ice, depth 0.5m   \n",
       "ERS1568987 2015-09-07      Arctic Ocean Arctic seawater metagenome, 1.5 m\n",
       "ERS1972377 NA              NA           Station 31 sea ice, depth 0.5m   \n",
       "ERS1972391 NA              NA           Station 33 sea ice, depth 0.1m   \n",
       "           environment-biome environment-feature â‹¯ pcr conditions\n",
       "ERS1568988 marine biome      marine water body   â‹¯ NA            \n",
       "ERS1972379 marine biome      marine water body   â‹¯ NA            \n",
       "ERS1972376 marine biome      marine water body   â‹¯ NA            \n",
       "ERS1568987 marine biome      marine water body   â‹¯ NA            \n",
       "ERS1972377 marine biome      marine water body   â‹¯ NA            \n",
       "ERS1972391 marine biome      marine water body   â‹¯ NA            \n",
       "           host common name host age host body habitat host diet host genotype\n",
       "ERS1568988 NA               NA       NA                NA        NA           \n",
       "ERS1972379 NA               NA       NA                NA        NA           \n",
       "ERS1972376 NA               NA       NA                NA        NA           \n",
       "ERS1568987 NA               NA       NA                NA        NA           \n",
       "ERS1972377 NA               NA       NA                NA        NA           \n",
       "ERS1972391 NA               NA       NA                NA        NA           \n",
       "           host phenotype host sex sample volume or weight for DNA extraction\n",
       "ERS1568988 NA             NA       NA                                        \n",
       "ERS1972379 NA             NA       NA                                        \n",
       "ERS1972376 NA             NA       NA                                        \n",
       "ERS1568987 NA             NA       NA                                        \n",
       "ERS1972377 NA             NA       NA                                        \n",
       "ERS1972391 NA             NA       NA                                        \n",
       "           chemical administration\n",
       "ERS1568988 NA                     \n",
       "ERS1972379 NA                     \n",
       "ERS1972376 NA                     \n",
       "ERS1568987 NA                     \n",
       "ERS1972377 NA                     \n",
       "ERS1972391 NA                     "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(samps_polar)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4b2c4b5a-0915-4da0-a9df-c85a692f86db",
   "metadata": {},
   "source": [
    "## Example: find Wastewater studies"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "1587a68f-d855-4bb8-a540-829aebc80f44",
   "metadata": {},
   "outputs": [],
   "source": [
    "studies_ww <- mgnify_query(mg, \"studies\", biome_name=\"wastewater\", maxhits=-1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "d825d7a4-319c-419a-b5c0-9b2bab3dd87a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 6 Ã— 13</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>samples-count</th><th scope=col>accession</th><th scope=col>bioproject</th><th scope=col>secondary-accession</th><th scope=col>centre-name</th><th scope=col>is-public</th><th scope=col>study-abstract</th><th scope=col>study-name</th><th scope=col>data-origination</th><th scope=col>last-update</th><th scope=col>acc_type</th><th scope=col>type</th><th scope=col>public-release-date</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>MGYS00005846</th><td>110</td><td>MGYS00005846</td><td>PRJEB47494 </td><td>ERP131768</td><td>EMG                                                                                                                                     </td><td>TRUE</td><td>The Third Party Annotation (TPA) assembly was derived from the primary whole genome shotgun (WGS) data set PRJEB27054, and was assembled with metaSPAdes v3.12.0. This project includes samples from the following biomes: root:Engineered:Wastewater:Water and sludge.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                            </td><td>EMG produced TPA metagenomics assembly of PRJEB27054 data set (Global surveillance of antimicrobial resistance).</td><td>SUBMITTED</td><td>2021-11-18T06:32:39</td><td>studies</td><td>studies</td><td>NA</td></tr>\n",
       "\t<tr><th scope=row>MGYS00005847</th><td>109</td><td>MGYS00005847</td><td>PRJEB27054 </td><td>ERP109094</td><td>DTU-GE                                                                                                                                  </td><td>TRUE</td><td>Antimicrobial resistance (AMR) is one of the most serious global public health threats, however, obtaining representative data on AMR for healthy human populations is difficult. We characterized the bacterial resistome from untreated sewage from 79 sites in 60 countries. We found systematic differences in abundance and diversity of AMR genes between Europe/North-America/Oceania and Africa/Asia/South-America. Antimicrobial use data only explained a minor part of the AMR variation and no evidence for cross-selection between antimicrobial classes nor effect of travel by flight between sites were found. However, AMR abundance was strongly correlated with socio-economic, health and environmental factors, which we used to predict AMR abundances in all countries in the world. Our findings suggest that the global AMR gene diversity and abundance varies by region and are caused by national circumstances. Improving sanitation and health could potentially limit the global burden of AMR. We propose to use sewage for an ethically acceptable and economically feasible continuous global surveillance and prediction of AMR.</td><td>Global surveillance of antimicrobial resistance                                                                 </td><td>SUBMITTED</td><td>2021-11-17T00:54:21</td><td>studies</td><td>studies</td><td>NA</td></tr>\n",
       "\t<tr><th scope=row>MGYS00005802</th><td>6  </td><td>MGYS00005802</td><td>PRJEB43967 </td><td>ERP127957</td><td>EMG                                                                                                                                     </td><td>TRUE</td><td>The Third Party Annotation (TPA) assembly was derived from the primary whole genome shotgun (WGS) data set PRJEB8087, and was assembled with SPAdes v3.11.1, metaSPAdes v3.12.0. This project includes samples from the following biomes: root:Engineered:Wastewater:Nutrient removal:Biological phosphorus removal:Activated sludge.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                              </td><td>EMG produced TPA metagenomics assembly of PRJEB8087 data set (Metagenomes of Danish EBPR WWTPs).                </td><td>SUBMITTED</td><td>2021-10-08T10:16:00</td><td>studies</td><td>studies</td><td>NA</td></tr>\n",
       "\t<tr><th scope=row>MGYS00005769</th><td>11 </td><td>MGYS00005769</td><td>PRJEB42552 </td><td>ERP126430</td><td>EMG                                                                                                                                     </td><td>TRUE</td><td>The Third Party Annotation (TPA) assembly was derived from the primary whole genome shotgun (WGS) data set PRJNA479723, and was assembled with Flye v2.8.1. This project includes samples from the following biomes: root:Engineered:Wastewater:Water and sludge.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                  </td><td>EMG produced TPA metagenomics assembly of PRJNA479723 data set (Nanopore metagenomics).                         </td><td>SUBMITTED</td><td>2021-07-28T17:07:01</td><td>studies</td><td>studies</td><td>NA</td></tr>\n",
       "\t<tr><th scope=row>MGYS00005770</th><td>10 </td><td>MGYS00005770</td><td>PRJNA479723</td><td>SRP152758</td><td>The University of Hong Kong                                                                                                             </td><td>TRUE</td><td>WWTPs nanopore metagenomics                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                        </td><td>Nanopore metagenomics                                                                                           </td><td>HARVESTED</td><td>2021-07-28T16:48:03</td><td>studies</td><td>studies</td><td>NA</td></tr>\n",
       "\t<tr><th scope=row>MGYS00005741</th><td>76 </td><td>MGYS00005741</td><td>PRJDB4240  </td><td>DRP003823</td><td>Mino/Satoh Laboratory, Department of Socio-cultural Environmental Studies, Graduate School of Frontier Sciences, The University of Tokyo</td><td>TRUE</td><td>The data is the whole sequence data for the \"Bacterial population dynamics in a laboratory activated sludge reactor monitored by pyrosequencing of 16S rRNA\", Satoh et al. (2013), Microbes and Environments, 28(1), 65-70.  Bacterial community change in a laboratory activated sludge reactor was analyzed by NGS for around 8 months.  A sudden disappearance of a major population was observed around day 50th, and periodical increase and recessions were observed for many of the observed OTUs.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          </td><td>WL Reactor published in ME                                                                                      </td><td>HARVESTED</td><td>2021-06-07T13:09:08</td><td>studies</td><td>studies</td><td>NA</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 6 Ã— 13\n",
       "\\begin{tabular}{r|lllllllllllll}\n",
       "  & samples-count & accession & bioproject & secondary-accession & centre-name & is-public & study-abstract & study-name & data-origination & last-update & acc\\_type & type & public-release-date\\\\\n",
       "  & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\tMGYS00005846 & 110 & MGYS00005846 & PRJEB47494  & ERP131768 & EMG                                                                                                                                      & TRUE & The Third Party Annotation (TPA) assembly was derived from the primary whole genome shotgun (WGS) data set PRJEB27054, and was assembled with metaSPAdes v3.12.0. This project includes samples from the following biomes: root:Engineered:Wastewater:Water and sludge.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                             & EMG produced TPA metagenomics assembly of PRJEB27054 data set (Global surveillance of antimicrobial resistance). & SUBMITTED & 2021-11-18T06:32:39 & studies & studies & NA\\\\\n",
       "\tMGYS00005847 & 109 & MGYS00005847 & PRJEB27054  & ERP109094 & DTU-GE                                                                                                                                   & TRUE & Antimicrobial resistance (AMR) is one of the most serious global public health threats, however, obtaining representative data on AMR for healthy human populations is difficult. We characterized the bacterial resistome from untreated sewage from 79 sites in 60 countries. We found systematic differences in abundance and diversity of AMR genes between Europe/North-America/Oceania and Africa/Asia/South-America. Antimicrobial use data only explained a minor part of the AMR variation and no evidence for cross-selection between antimicrobial classes nor effect of travel by flight between sites were found. However, AMR abundance was strongly correlated with socio-economic, health and environmental factors, which we used to predict AMR abundances in all countries in the world. Our findings suggest that the global AMR gene diversity and abundance varies by region and are caused by national circumstances. Improving sanitation and health could potentially limit the global burden of AMR. We propose to use sewage for an ethically acceptable and economically feasible continuous global surveillance and prediction of AMR. & Global surveillance of antimicrobial resistance                                                                  & SUBMITTED & 2021-11-17T00:54:21 & studies & studies & NA\\\\\n",
       "\tMGYS00005802 & 6   & MGYS00005802 & PRJEB43967  & ERP127957 & EMG                                                                                                                                      & TRUE & The Third Party Annotation (TPA) assembly was derived from the primary whole genome shotgun (WGS) data set PRJEB8087, and was assembled with SPAdes v3.11.1, metaSPAdes v3.12.0. This project includes samples from the following biomes: root:Engineered:Wastewater:Nutrient removal:Biological phosphorus removal:Activated sludge.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                               & EMG produced TPA metagenomics assembly of PRJEB8087 data set (Metagenomes of Danish EBPR WWTPs).                 & SUBMITTED & 2021-10-08T10:16:00 & studies & studies & NA\\\\\n",
       "\tMGYS00005769 & 11  & MGYS00005769 & PRJEB42552  & ERP126430 & EMG                                                                                                                                      & TRUE & The Third Party Annotation (TPA) assembly was derived from the primary whole genome shotgun (WGS) data set PRJNA479723, and was assembled with Flye v2.8.1. This project includes samples from the following biomes: root:Engineered:Wastewater:Water and sludge.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   & EMG produced TPA metagenomics assembly of PRJNA479723 data set (Nanopore metagenomics).                          & SUBMITTED & 2021-07-28T17:07:01 & studies & studies & NA\\\\\n",
       "\tMGYS00005770 & 10  & MGYS00005770 & PRJNA479723 & SRP152758 & The University of Hong Kong                                                                                                              & TRUE & WWTPs nanopore metagenomics                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                         & Nanopore metagenomics                                                                                            & HARVESTED & 2021-07-28T16:48:03 & studies & studies & NA\\\\\n",
       "\tMGYS00005741 & 76  & MGYS00005741 & PRJDB4240   & DRP003823 & Mino/Satoh Laboratory, Department of Socio-cultural Environmental Studies, Graduate School of Frontier Sciences, The University of Tokyo & TRUE & The data is the whole sequence data for the \"Bacterial population dynamics in a laboratory activated sludge reactor monitored by pyrosequencing of 16S rRNA\", Satoh et al. (2013), Microbes and Environments, 28(1), 65-70.  Bacterial community change in a laboratory activated sludge reactor was analyzed by NGS for around 8 months.  A sudden disappearance of a major population was observed around day 50th, and periodical increase and recessions were observed for many of the observed OTUs.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                           & WL Reactor published in ME                                                                                       & HARVESTED & 2021-06-07T13:09:08 & studies & studies & NA\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 6 Ã— 13\n",
       "\n",
       "| <!--/--> | samples-count &lt;chr&gt; | accession &lt;chr&gt; | bioproject &lt;chr&gt; | secondary-accession &lt;chr&gt; | centre-name &lt;chr&gt; | is-public &lt;chr&gt; | study-abstract &lt;chr&gt; | study-name &lt;chr&gt; | data-origination &lt;chr&gt; | last-update &lt;chr&gt; | acc_type &lt;chr&gt; | type &lt;chr&gt; | public-release-date &lt;chr&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| MGYS00005846 | 110 | MGYS00005846 | PRJEB47494  | ERP131768 | EMG                                                                                                                                      | TRUE | The Third Party Annotation (TPA) assembly was derived from the primary whole genome shotgun (WGS) data set PRJEB27054, and was assembled with metaSPAdes v3.12.0. This project includes samples from the following biomes: root:Engineered:Wastewater:Water and sludge.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                             | EMG produced TPA metagenomics assembly of PRJEB27054 data set (Global surveillance of antimicrobial resistance). | SUBMITTED | 2021-11-18T06:32:39 | studies | studies | NA |\n",
       "| MGYS00005847 | 109 | MGYS00005847 | PRJEB27054  | ERP109094 | DTU-GE                                                                                                                                   | TRUE | Antimicrobial resistance (AMR) is one of the most serious global public health threats, however, obtaining representative data on AMR for healthy human populations is difficult. We characterized the bacterial resistome from untreated sewage from 79 sites in 60 countries. We found systematic differences in abundance and diversity of AMR genes between Europe/North-America/Oceania and Africa/Asia/South-America. Antimicrobial use data only explained a minor part of the AMR variation and no evidence for cross-selection between antimicrobial classes nor effect of travel by flight between sites were found. However, AMR abundance was strongly correlated with socio-economic, health and environmental factors, which we used to predict AMR abundances in all countries in the world. Our findings suggest that the global AMR gene diversity and abundance varies by region and are caused by national circumstances. Improving sanitation and health could potentially limit the global burden of AMR. We propose to use sewage for an ethically acceptable and economically feasible continuous global surveillance and prediction of AMR. | Global surveillance of antimicrobial resistance                                                                  | SUBMITTED | 2021-11-17T00:54:21 | studies | studies | NA |\n",
       "| MGYS00005802 | 6   | MGYS00005802 | PRJEB43967  | ERP127957 | EMG                                                                                                                                      | TRUE | The Third Party Annotation (TPA) assembly was derived from the primary whole genome shotgun (WGS) data set PRJEB8087, and was assembled with SPAdes v3.11.1, metaSPAdes v3.12.0. This project includes samples from the following biomes: root:Engineered:Wastewater:Nutrient removal:Biological phosphorus removal:Activated sludge.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                               | EMG produced TPA metagenomics assembly of PRJEB8087 data set (Metagenomes of Danish EBPR WWTPs).                 | SUBMITTED | 2021-10-08T10:16:00 | studies | studies | NA |\n",
       "| MGYS00005769 | 11  | MGYS00005769 | PRJEB42552  | ERP126430 | EMG                                                                                                                                      | TRUE | The Third Party Annotation (TPA) assembly was derived from the primary whole genome shotgun (WGS) data set PRJNA479723, and was assembled with Flye v2.8.1. This project includes samples from the following biomes: root:Engineered:Wastewater:Water and sludge.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   | EMG produced TPA metagenomics assembly of PRJNA479723 data set (Nanopore metagenomics).                          | SUBMITTED | 2021-07-28T17:07:01 | studies | studies | NA |\n",
       "| MGYS00005770 | 10  | MGYS00005770 | PRJNA479723 | SRP152758 | The University of Hong Kong                                                                                                              | TRUE | WWTPs nanopore metagenomics                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                         | Nanopore metagenomics                                                                                            | HARVESTED | 2021-07-28T16:48:03 | studies | studies | NA |\n",
       "| MGYS00005741 | 76  | MGYS00005741 | PRJDB4240   | DRP003823 | Mino/Satoh Laboratory, Department of Socio-cultural Environmental Studies, Graduate School of Frontier Sciences, The University of Tokyo | TRUE | The data is the whole sequence data for the \"Bacterial population dynamics in a laboratory activated sludge reactor monitored by pyrosequencing of 16S rRNA\", Satoh et al. (2013), Microbes and Environments, 28(1), 65-70.  Bacterial community change in a laboratory activated sludge reactor was analyzed by NGS for around 8 months.  A sudden disappearance of a major population was observed around day 50th, and periodical increase and recessions were observed for many of the observed OTUs.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                           | WL Reactor published in ME                                                                                       | HARVESTED | 2021-06-07T13:09:08 | studies | studies | NA |\n",
       "\n"
      ],
      "text/plain": [
       "             samples-count accession    bioproject  secondary-accession\n",
       "MGYS00005846 110           MGYS00005846 PRJEB47494  ERP131768          \n",
       "MGYS00005847 109           MGYS00005847 PRJEB27054  ERP109094          \n",
       "MGYS00005802 6             MGYS00005802 PRJEB43967  ERP127957          \n",
       "MGYS00005769 11            MGYS00005769 PRJEB42552  ERP126430          \n",
       "MGYS00005770 10            MGYS00005770 PRJNA479723 SRP152758          \n",
       "MGYS00005741 76            MGYS00005741 PRJDB4240   DRP003823          \n",
       "             centre-name                                                                                                                             \n",
       "MGYS00005846 EMG                                                                                                                                     \n",
       "MGYS00005847 DTU-GE                                                                                                                                  \n",
       "MGYS00005802 EMG                                                                                                                                     \n",
       "MGYS00005769 EMG                                                                                                                                     \n",
       "MGYS00005770 The University of Hong Kong                                                                                                             \n",
       "MGYS00005741 Mino/Satoh Laboratory, Department of Socio-cultural Environmental Studies, Graduate School of Frontier Sciences, The University of Tokyo\n",
       "             is-public\n",
       "MGYS00005846 TRUE     \n",
       "MGYS00005847 TRUE     \n",
       "MGYS00005802 TRUE     \n",
       "MGYS00005769 TRUE     \n",
       "MGYS00005770 TRUE     \n",
       "MGYS00005741 TRUE     \n",
       "             study-abstract                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     \n",
       "MGYS00005846 The Third Party Annotation (TPA) assembly was derived from the primary whole genome shotgun (WGS) data set PRJEB27054, and was assembled with metaSPAdes v3.12.0. This project includes samples from the following biomes: root:Engineered:Wastewater:Water and sludge.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                            \n",
       "MGYS00005847 Antimicrobial resistance (AMR) is one of the most serious global public health threats, however, obtaining representative data on AMR for healthy human populations is difficult. We characterized the bacterial resistome from untreated sewage from 79 sites in 60 countries. We found systematic differences in abundance and diversity of AMR genes between Europe/North-America/Oceania and Africa/Asia/South-America. Antimicrobial use data only explained a minor part of the AMR variation and no evidence for cross-selection between antimicrobial classes nor effect of travel by flight between sites were found. However, AMR abundance was strongly correlated with socio-economic, health and environmental factors, which we used to predict AMR abundances in all countries in the world. Our findings suggest that the global AMR gene diversity and abundance varies by region and are caused by national circumstances. Improving sanitation and health could potentially limit the global burden of AMR. We propose to use sewage for an ethically acceptable and economically feasible continuous global surveillance and prediction of AMR.\n",
       "MGYS00005802 The Third Party Annotation (TPA) assembly was derived from the primary whole genome shotgun (WGS) data set PRJEB8087, and was assembled with SPAdes v3.11.1, metaSPAdes v3.12.0. This project includes samples from the following biomes: root:Engineered:Wastewater:Nutrient removal:Biological phosphorus removal:Activated sludge.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                              \n",
       "MGYS00005769 The Third Party Annotation (TPA) assembly was derived from the primary whole genome shotgun (WGS) data set PRJNA479723, and was assembled with Flye v2.8.1. This project includes samples from the following biomes: root:Engineered:Wastewater:Water and sludge.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                  \n",
       "MGYS00005770 WWTPs nanopore metagenomics                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                        \n",
       "MGYS00005741 The data is the whole sequence data for the \"Bacterial population dynamics in a laboratory activated sludge reactor monitored by pyrosequencing of 16S rRNA\", Satoh et al. (2013), Microbes and Environments, 28(1), 65-70.  Bacterial community change in a laboratory activated sludge reactor was analyzed by NGS for around 8 months.  A sudden disappearance of a major population was observed around day 50th, and periodical increase and recessions were observed for many of the observed OTUs.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          \n",
       "             study-name                                                                                                      \n",
       "MGYS00005846 EMG produced TPA metagenomics assembly of PRJEB27054 data set (Global surveillance of antimicrobial resistance).\n",
       "MGYS00005847 Global surveillance of antimicrobial resistance                                                                 \n",
       "MGYS00005802 EMG produced TPA metagenomics assembly of PRJEB8087 data set (Metagenomes of Danish EBPR WWTPs).                \n",
       "MGYS00005769 EMG produced TPA metagenomics assembly of PRJNA479723 data set (Nanopore metagenomics).                         \n",
       "MGYS00005770 Nanopore metagenomics                                                                                           \n",
       "MGYS00005741 WL Reactor published in ME                                                                                      \n",
       "             data-origination last-update         acc_type type   \n",
       "MGYS00005846 SUBMITTED        2021-11-18T06:32:39 studies  studies\n",
       "MGYS00005847 SUBMITTED        2021-11-17T00:54:21 studies  studies\n",
       "MGYS00005802 SUBMITTED        2021-10-08T10:16:00 studies  studies\n",
       "MGYS00005769 SUBMITTED        2021-07-28T17:07:01 studies  studies\n",
       "MGYS00005770 HARVESTED        2021-07-28T16:48:03 studies  studies\n",
       "MGYS00005741 HARVESTED        2021-06-07T13:09:08 studies  studies\n",
       "             public-release-date\n",
       "MGYS00005846 NA                 \n",
       "MGYS00005847 NA                 \n",
       "MGYS00005802 NA                 \n",
       "MGYS00005769 NA                 \n",
       "MGYS00005770 NA                 \n",
       "MGYS00005741 NA                 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(studies_ww)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f6edc0c5-48a8-4b8d-a227-b82991007935",
   "metadata": {},
   "source": [
    "## More Sample filters"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e975a310-262a-4870-b208-35b8c7868133",
   "metadata": {},
   "source": [
    "### By location"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cea4da31-788d-4f1d-8f22-fa5bf131dd07",
   "metadata": {},
   "outputs": [],
   "source": [
    "more_northerly_than <- mgnify_query(mg, \"samples\", latitude_gte=88, maxhits=-1)\n",
    "\n",
    "more_southerly_than <- mgnify_query(mg, \"samples\", latitude_lte=-88, maxhits=-1)\n",
    "\n",
    "more_easterly_than <- mgnify_query(mg, \"samples\", longitude_gte=170, maxhits=-1)\n",
    "\n",
    "more_westerly_than <- mgnify_query(mg, \"samples\", longitude_lte=170, maxhits=-1)\n",
    "\n",
    "at_location <- mgnify_query(mg, \"samples\", geo_loc_name=\"usa\", maxhits=-1)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "612eae01-4049-4f8e-90bc-866aedffd908",
   "metadata": {},
   "source": [
    "### By biome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "97075a7b-3159-42c0-96c4-6a517370b1c4",
   "metadata": {},
   "outputs": [],
   "source": [
    "biome_within_wastewater <- mgnify_query(mg, \"samples\", biome_name=\"wastewater\", maxhits=-1)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "533541ce-b64e-43a5-984b-0e13970bc2cb",
   "metadata": {},
   "source": [
    "### By metadata\n",
    "There are a large number of metadata key:value pairs, because these are author-submitted, along with the samples, to the ENA archive.\n",
    "\n",
    "If you know how to specify the metadata key:value query for the samples you're interested in, you can use this form to find matching Samples:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "5b2a26cb-98ce-483c-88d1-c6b0fe149030",
   "metadata": {},
   "outputs": [],
   "source": [
    "from_ex_smokers <- mgnify_query(mg, \"samples\", metadata_key=\"smoker\", metadata_value=\"ex-smoker\", maxhits=-1)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c554cea5-5ccd-44ed-bfef-a790eea6bfed",
   "metadata": {},
   "source": [
    "To find `metadata_key`s and values, it is best to browse the [interactive API Browser](https://www.ebi.ac.uk/metagenomics/v1/samples), and use the `Filters` button to construct queries interactively at first."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6eb4e359-5905-4722-ab88-65a9b3e6e9c2",
   "metadata": {},
   "source": [
    "--- \n",
    "## More Study filters"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0d3ea380-2f3f-4fdd-97c3-eef9d24a3800",
   "metadata": {},
   "source": [
    "### By Centre Name"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "c670e399-7297-44f3-951b-58040a9ac0d4",
   "metadata": {},
   "outputs": [],
   "source": [
    "from_smithsonian <- mgnify_query(mg, \"studies\", centre_name=\"Smithsonian\", maxhits=-1)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fe1f615b-68ec-4b60-bd87-74d3f6fe016b",
   "metadata": {},
   "source": [
    "---\n",
    "\n",
    "## Example: adding additional filters to the data frame"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a35c8957-f3de-4526-9f27-fc9e2b28c3df",
   "metadata": {},
   "source": [
    "First, fetch some samples from the Lentic biome. We can specify the entire Biome lineage, too."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "1dd0f6c8-0430-4ad2-adef-fb7c5c776fa5",
   "metadata": {},
   "outputs": [],
   "source": [
    "lentic_samples <- mgnify_query(mg, \"samples\", biome_name=\"root:Environmental:Aquatic:Lentic\", usecache=T)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7d5c00be-be94-4368-9141-2ec66e8e16ce",
   "metadata": {},
   "source": [
    "Not, also filter by depth *within* the returned results, using normal R syntax."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "8e0a11f8-bb7c-4cd6-a314-9cb3ed7b4e0f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 16 Ã— 37</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>latitude</th><th scope=col>biosample</th><th scope=col>longitude</th><th scope=col>accession</th><th scope=col>collection-date</th><th scope=col>sample-desc</th><th scope=col>sample-name</th><th scope=col>sample-alias</th><th scope=col>last-update</th><th scope=col>geographic location (longitude)</th><th scope=col>â‹¯</th><th scope=col>instrument model</th><th scope=col>last update date</th><th scope=col>investigation type</th><th scope=col>project name</th><th scope=col>geographic location (depth)</th><th scope=col>geographic location (altitude)</th><th scope=col>environmental package</th><th scope=col>sequencing method</th><th scope=col>NCBI sample classification</th><th scope=col>ENA checklist</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>â‹¯</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>SRS992699</th><td>17.39</td><td>SAMN03860260</td><td>40.54</td><td>SRS992699</td><td>2011-10-15</td><td>12 </td><td>sample03</td><td>sample03</td><td>2020-05-18T00:52:00</td><td>40.54</td><td>â‹¯</td><td>Illumina HiSeq 2000</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td></tr>\n",
       "\t<tr><th scope=row>SRS992702</th><td>20.31</td><td>SAMN03860274</td><td>38.46</td><td>SRS992702</td><td>2011-10-15</td><td>91 </td><td>sample17</td><td>sample17</td><td>2020-05-18T00:51:47</td><td>38.46</td><td>â‹¯</td><td>Illumina HiSeq 2000</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td></tr>\n",
       "\t<tr><th scope=row>SRS992693</th><td>17.39</td><td>SAMN03860259</td><td>40.54</td><td>SRS992693</td><td>2011-10-15</td><td>12 </td><td>sample02</td><td>sample02</td><td>2020-05-18T00:50:43</td><td>40.54</td><td>â‹¯</td><td>Illumina HiSeq 2000</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td></tr>\n",
       "\t<tr><th scope=row>SRS992705</th><td>23.36</td><td>SAMN03860286</td><td>37.3 </td><td>SRS992705</td><td>2011-10-15</td><td>149</td><td>sample29</td><td>sample29</td><td>2020-05-18T00:46:05</td><td>37.3 </td><td>â‹¯</td><td>Illumina HiSeq 2000</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td></tr>\n",
       "\t<tr><th scope=row>SRS992692</th><td>18.34</td><td>SAMN03860268</td><td>40.44</td><td>SRS992692</td><td>2011-10-15</td><td>34 </td><td>sample11</td><td>sample11</td><td>2020-05-18T00:45:26</td><td>40.44</td><td>â‹¯</td><td>Illumina HiSeq 2000</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td></tr>\n",
       "\t<tr><th scope=row>SRS992710</th><td>22.2 </td><td>SAMN03860281</td><td>37.55</td><td>SRS992710</td><td>2011-10-15</td><td>108</td><td>sample24</td><td>sample24</td><td>2020-05-18T00:35:28</td><td>37.55</td><td>â‹¯</td><td>Illumina HiSeq 2000</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td></tr>\n",
       "\t<tr><th scope=row>SRS992714</th><td>25.46</td><td>SAMN03860292</td><td>36.6 </td><td>SRS992714</td><td>2011-10-15</td><td>169</td><td>sample35</td><td>sample35</td><td>2020-05-18T00:35:15</td><td>36.6 </td><td>â‹¯</td><td>Illumina HiSeq 2000</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td></tr>\n",
       "\t<tr><th scope=row>SRS992704</th><td>23.36</td><td>SAMN03860287</td><td>37.3 </td><td>SRS992704</td><td>2011-10-15</td><td>149</td><td>sample30</td><td>sample30</td><td>2020-05-18T00:27:10</td><td>37.3 </td><td>â‹¯</td><td>Illumina HiSeq 2000</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td></tr>\n",
       "\t<tr><th scope=row>SRS992713</th><td>25.46</td><td>SAMN03860293</td><td>36.6 </td><td>SRS992713</td><td>2011-10-15</td><td>169</td><td>sample36</td><td>sample36</td><td>2020-05-18T00:13:07</td><td>36.6 </td><td>â‹¯</td><td>Illumina HiSeq 2000</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td></tr>\n",
       "\t<tr><th scope=row>SRS992696</th><td>22.2 </td><td>SAMN03860280</td><td>37.55</td><td>SRS992696</td><td>2011-10-15</td><td>108</td><td>sample23</td><td>sample23</td><td>2020-05-18T00:06:42</td><td>37.55</td><td>â‹¯</td><td>Illumina HiSeq 2000</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td></tr>\n",
       "\t<tr><th scope=row>SRS992729</th><td>17.59</td><td>SAMN03860263</td><td>39.47</td><td>SRS992729</td><td>2011-10-15</td><td>22 </td><td>sample06</td><td>sample06</td><td>2020-05-17T10:16:43</td><td>39.47</td><td>â‹¯</td><td>Illumina HiSeq 2000</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td></tr>\n",
       "\t<tr><th scope=row>SRS992701</th><td>20.31</td><td>SAMN03860275</td><td>38.46</td><td>SRS992701</td><td>2011-10-15</td><td>91 </td><td>sample18</td><td>sample18</td><td>2020-05-17T10:11:28</td><td>38.46</td><td>â‹¯</td><td>Illumina HiSeq 2000</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td></tr>\n",
       "\t<tr><th scope=row>SRS992691</th><td>18.34</td><td>SAMN03860269</td><td>40.44</td><td>SRS992691</td><td>2011-10-15</td><td>34 </td><td>sample12</td><td>sample12</td><td>2020-05-17T10:01:44</td><td>40.44</td><td>â‹¯</td><td>Illumina HiSeq 2000</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td></tr>\n",
       "\t<tr><th scope=row>SRS992721</th><td>17.59</td><td>SAMN03860262</td><td>39.47</td><td>SRS992721</td><td>2011-10-15</td><td>22 </td><td>sample05</td><td>sample05</td><td>2020-05-17T10:01:21</td><td>39.47</td><td>â‹¯</td><td>Illumina HiSeq 2000</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td></tr>\n",
       "\t<tr><th scope=row>SRS992720</th><td>27.53</td><td>SAMN03860299</td><td>34.3 </td><td>SRS992720</td><td>2011-10-15</td><td>192</td><td>sample42</td><td>sample42</td><td>2020-05-17T10:00:58</td><td>34.3 </td><td>â‹¯</td><td>Illumina HiSeq 2000</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td></tr>\n",
       "\t<tr><th scope=row>SRS992719</th><td>27.53</td><td>SAMN03860298</td><td>34.3 </td><td>SRS992719</td><td>2011-10-15</td><td>192</td><td>sample41</td><td>sample41</td><td>2020-05-17T10:00:45</td><td>34.3 </td><td>â‹¯</td><td>Illumina HiSeq 2000</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 16 Ã— 37\n",
       "\\begin{tabular}{r|lllllllllllllllllllll}\n",
       "  & latitude & biosample & longitude & accession & collection-date & sample-desc & sample-name & sample-alias & last-update & geographic location (longitude) & â‹¯ & instrument model & last update date & investigation type & project name & geographic location (depth) & geographic location (altitude) & environmental package & sequencing method & NCBI sample classification & ENA checklist\\\\\n",
       "  & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & â‹¯ & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\tSRS992699 & 17.39 & SAMN03860260 & 40.54 & SRS992699 & 2011-10-15 & 12  & sample03 & sample03 & 2020-05-18T00:52:00 & 40.54 & â‹¯ & Illumina HiSeq 2000 & NA & NA & NA & NA & NA & NA & NA & NA & NA\\\\\n",
       "\tSRS992702 & 20.31 & SAMN03860274 & 38.46 & SRS992702 & 2011-10-15 & 91  & sample17 & sample17 & 2020-05-18T00:51:47 & 38.46 & â‹¯ & Illumina HiSeq 2000 & NA & NA & NA & NA & NA & NA & NA & NA & NA\\\\\n",
       "\tSRS992693 & 17.39 & SAMN03860259 & 40.54 & SRS992693 & 2011-10-15 & 12  & sample02 & sample02 & 2020-05-18T00:50:43 & 40.54 & â‹¯ & Illumina HiSeq 2000 & NA & NA & NA & NA & NA & NA & NA & NA & NA\\\\\n",
       "\tSRS992705 & 23.36 & SAMN03860286 & 37.3  & SRS992705 & 2011-10-15 & 149 & sample29 & sample29 & 2020-05-18T00:46:05 & 37.3  & â‹¯ & Illumina HiSeq 2000 & NA & NA & NA & NA & NA & NA & NA & NA & NA\\\\\n",
       "\tSRS992692 & 18.34 & SAMN03860268 & 40.44 & SRS992692 & 2011-10-15 & 34  & sample11 & sample11 & 2020-05-18T00:45:26 & 40.44 & â‹¯ & Illumina HiSeq 2000 & NA & NA & NA & NA & NA & NA & NA & NA & NA\\\\\n",
       "\tSRS992710 & 22.2  & SAMN03860281 & 37.55 & SRS992710 & 2011-10-15 & 108 & sample24 & sample24 & 2020-05-18T00:35:28 & 37.55 & â‹¯ & Illumina HiSeq 2000 & NA & NA & NA & NA & NA & NA & NA & NA & NA\\\\\n",
       "\tSRS992714 & 25.46 & SAMN03860292 & 36.6  & SRS992714 & 2011-10-15 & 169 & sample35 & sample35 & 2020-05-18T00:35:15 & 36.6  & â‹¯ & Illumina HiSeq 2000 & NA & NA & NA & NA & NA & NA & NA & NA & NA\\\\\n",
       "\tSRS992704 & 23.36 & SAMN03860287 & 37.3  & SRS992704 & 2011-10-15 & 149 & sample30 & sample30 & 2020-05-18T00:27:10 & 37.3  & â‹¯ & Illumina HiSeq 2000 & NA & NA & NA & NA & NA & NA & NA & NA & NA\\\\\n",
       "\tSRS992713 & 25.46 & SAMN03860293 & 36.6  & SRS992713 & 2011-10-15 & 169 & sample36 & sample36 & 2020-05-18T00:13:07 & 36.6  & â‹¯ & Illumina HiSeq 2000 & NA & NA & NA & NA & NA & NA & NA & NA & NA\\\\\n",
       "\tSRS992696 & 22.2  & SAMN03860280 & 37.55 & SRS992696 & 2011-10-15 & 108 & sample23 & sample23 & 2020-05-18T00:06:42 & 37.55 & â‹¯ & Illumina HiSeq 2000 & NA & NA & NA & NA & NA & NA & NA & NA & NA\\\\\n",
       "\tSRS992729 & 17.59 & SAMN03860263 & 39.47 & SRS992729 & 2011-10-15 & 22  & sample06 & sample06 & 2020-05-17T10:16:43 & 39.47 & â‹¯ & Illumina HiSeq 2000 & NA & NA & NA & NA & NA & NA & NA & NA & NA\\\\\n",
       "\tSRS992701 & 20.31 & SAMN03860275 & 38.46 & SRS992701 & 2011-10-15 & 91  & sample18 & sample18 & 2020-05-17T10:11:28 & 38.46 & â‹¯ & Illumina HiSeq 2000 & NA & NA & NA & NA & NA & NA & NA & NA & NA\\\\\n",
       "\tSRS992691 & 18.34 & SAMN03860269 & 40.44 & SRS992691 & 2011-10-15 & 34  & sample12 & sample12 & 2020-05-17T10:01:44 & 40.44 & â‹¯ & Illumina HiSeq 2000 & NA & NA & NA & NA & NA & NA & NA & NA & NA\\\\\n",
       "\tSRS992721 & 17.59 & SAMN03860262 & 39.47 & SRS992721 & 2011-10-15 & 22  & sample05 & sample05 & 2020-05-17T10:01:21 & 39.47 & â‹¯ & Illumina HiSeq 2000 & NA & NA & NA & NA & NA & NA & NA & NA & NA\\\\\n",
       "\tSRS992720 & 27.53 & SAMN03860299 & 34.3  & SRS992720 & 2011-10-15 & 192 & sample42 & sample42 & 2020-05-17T10:00:58 & 34.3  & â‹¯ & Illumina HiSeq 2000 & NA & NA & NA & NA & NA & NA & NA & NA & NA\\\\\n",
       "\tSRS992719 & 27.53 & SAMN03860298 & 34.3  & SRS992719 & 2011-10-15 & 192 & sample41 & sample41 & 2020-05-17T10:00:45 & 34.3  & â‹¯ & Illumina HiSeq 2000 & NA & NA & NA & NA & NA & NA & NA & NA & NA\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 16 Ã— 37\n",
       "\n",
       "| <!--/--> | latitude &lt;chr&gt; | biosample &lt;chr&gt; | longitude &lt;chr&gt; | accession &lt;chr&gt; | collection-date &lt;chr&gt; | sample-desc &lt;chr&gt; | sample-name &lt;chr&gt; | sample-alias &lt;chr&gt; | last-update &lt;chr&gt; | geographic location (longitude) &lt;chr&gt; | â‹¯ â‹¯ | instrument model &lt;chr&gt; | last update date &lt;chr&gt; | investigation type &lt;chr&gt; | project name &lt;chr&gt; | geographic location (depth) &lt;chr&gt; | geographic location (altitude) &lt;chr&gt; | environmental package &lt;chr&gt; | sequencing method &lt;chr&gt; | NCBI sample classification &lt;chr&gt; | ENA checklist &lt;chr&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| SRS992699 | 17.39 | SAMN03860260 | 40.54 | SRS992699 | 2011-10-15 | 12  | sample03 | sample03 | 2020-05-18T00:52:00 | 40.54 | â‹¯ | Illumina HiSeq 2000 | NA | NA | NA | NA | NA | NA | NA | NA | NA |\n",
       "| SRS992702 | 20.31 | SAMN03860274 | 38.46 | SRS992702 | 2011-10-15 | 91  | sample17 | sample17 | 2020-05-18T00:51:47 | 38.46 | â‹¯ | Illumina HiSeq 2000 | NA | NA | NA | NA | NA | NA | NA | NA | NA |\n",
       "| SRS992693 | 17.39 | SAMN03860259 | 40.54 | SRS992693 | 2011-10-15 | 12  | sample02 | sample02 | 2020-05-18T00:50:43 | 40.54 | â‹¯ | Illumina HiSeq 2000 | NA | NA | NA | NA | NA | NA | NA | NA | NA |\n",
       "| SRS992705 | 23.36 | SAMN03860286 | 37.3  | SRS992705 | 2011-10-15 | 149 | sample29 | sample29 | 2020-05-18T00:46:05 | 37.3  | â‹¯ | Illumina HiSeq 2000 | NA | NA | NA | NA | NA | NA | NA | NA | NA |\n",
       "| SRS992692 | 18.34 | SAMN03860268 | 40.44 | SRS992692 | 2011-10-15 | 34  | sample11 | sample11 | 2020-05-18T00:45:26 | 40.44 | â‹¯ | Illumina HiSeq 2000 | NA | NA | NA | NA | NA | NA | NA | NA | NA |\n",
       "| SRS992710 | 22.2  | SAMN03860281 | 37.55 | SRS992710 | 2011-10-15 | 108 | sample24 | sample24 | 2020-05-18T00:35:28 | 37.55 | â‹¯ | Illumina HiSeq 2000 | NA | NA | NA | NA | NA | NA | NA | NA | NA |\n",
       "| SRS992714 | 25.46 | SAMN03860292 | 36.6  | SRS992714 | 2011-10-15 | 169 | sample35 | sample35 | 2020-05-18T00:35:15 | 36.6  | â‹¯ | Illumina HiSeq 2000 | NA | NA | NA | NA | NA | NA | NA | NA | NA |\n",
       "| SRS992704 | 23.36 | SAMN03860287 | 37.3  | SRS992704 | 2011-10-15 | 149 | sample30 | sample30 | 2020-05-18T00:27:10 | 37.3  | â‹¯ | Illumina HiSeq 2000 | NA | NA | NA | NA | NA | NA | NA | NA | NA |\n",
       "| SRS992713 | 25.46 | SAMN03860293 | 36.6  | SRS992713 | 2011-10-15 | 169 | sample36 | sample36 | 2020-05-18T00:13:07 | 36.6  | â‹¯ | Illumina HiSeq 2000 | NA | NA | NA | NA | NA | NA | NA | NA | NA |\n",
       "| SRS992696 | 22.2  | SAMN03860280 | 37.55 | SRS992696 | 2011-10-15 | 108 | sample23 | sample23 | 2020-05-18T00:06:42 | 37.55 | â‹¯ | Illumina HiSeq 2000 | NA | NA | NA | NA | NA | NA | NA | NA | NA |\n",
       "| SRS992729 | 17.59 | SAMN03860263 | 39.47 | SRS992729 | 2011-10-15 | 22  | sample06 | sample06 | 2020-05-17T10:16:43 | 39.47 | â‹¯ | Illumina HiSeq 2000 | NA | NA | NA | NA | NA | NA | NA | NA | NA |\n",
       "| SRS992701 | 20.31 | SAMN03860275 | 38.46 | SRS992701 | 2011-10-15 | 91  | sample18 | sample18 | 2020-05-17T10:11:28 | 38.46 | â‹¯ | Illumina HiSeq 2000 | NA | NA | NA | NA | NA | NA | NA | NA | NA |\n",
       "| SRS992691 | 18.34 | SAMN03860269 | 40.44 | SRS992691 | 2011-10-15 | 34  | sample12 | sample12 | 2020-05-17T10:01:44 | 40.44 | â‹¯ | Illumina HiSeq 2000 | NA | NA | NA | NA | NA | NA | NA | NA | NA |\n",
       "| SRS992721 | 17.59 | SAMN03860262 | 39.47 | SRS992721 | 2011-10-15 | 22  | sample05 | sample05 | 2020-05-17T10:01:21 | 39.47 | â‹¯ | Illumina HiSeq 2000 | NA | NA | NA | NA | NA | NA | NA | NA | NA |\n",
       "| SRS992720 | 27.53 | SAMN03860299 | 34.3  | SRS992720 | 2011-10-15 | 192 | sample42 | sample42 | 2020-05-17T10:00:58 | 34.3  | â‹¯ | Illumina HiSeq 2000 | NA | NA | NA | NA | NA | NA | NA | NA | NA |\n",
       "| SRS992719 | 27.53 | SAMN03860298 | 34.3  | SRS992719 | 2011-10-15 | 192 | sample41 | sample41 | 2020-05-17T10:00:45 | 34.3  | â‹¯ | Illumina HiSeq 2000 | NA | NA | NA | NA | NA | NA | NA | NA | NA |\n",
       "\n"
      ],
      "text/plain": [
       "          latitude biosample    longitude accession collection-date sample-desc\n",
       "SRS992699 17.39    SAMN03860260 40.54     SRS992699 2011-10-15      12         \n",
       "SRS992702 20.31    SAMN03860274 38.46     SRS992702 2011-10-15      91         \n",
       "SRS992693 17.39    SAMN03860259 40.54     SRS992693 2011-10-15      12         \n",
       "SRS992705 23.36    SAMN03860286 37.3      SRS992705 2011-10-15      149        \n",
       "SRS992692 18.34    SAMN03860268 40.44     SRS992692 2011-10-15      34         \n",
       "SRS992710 22.2     SAMN03860281 37.55     SRS992710 2011-10-15      108        \n",
       "SRS992714 25.46    SAMN03860292 36.6      SRS992714 2011-10-15      169        \n",
       "SRS992704 23.36    SAMN03860287 37.3      SRS992704 2011-10-15      149        \n",
       "SRS992713 25.46    SAMN03860293 36.6      SRS992713 2011-10-15      169        \n",
       "SRS992696 22.2     SAMN03860280 37.55     SRS992696 2011-10-15      108        \n",
       "SRS992729 17.59    SAMN03860263 39.47     SRS992729 2011-10-15      22         \n",
       "SRS992701 20.31    SAMN03860275 38.46     SRS992701 2011-10-15      91         \n",
       "SRS992691 18.34    SAMN03860269 40.44     SRS992691 2011-10-15      34         \n",
       "SRS992721 17.59    SAMN03860262 39.47     SRS992721 2011-10-15      22         \n",
       "SRS992720 27.53    SAMN03860299 34.3      SRS992720 2011-10-15      192        \n",
       "SRS992719 27.53    SAMN03860298 34.3      SRS992719 2011-10-15      192        \n",
       "          sample-name sample-alias last-update        \n",
       "SRS992699 sample03    sample03     2020-05-18T00:52:00\n",
       "SRS992702 sample17    sample17     2020-05-18T00:51:47\n",
       "SRS992693 sample02    sample02     2020-05-18T00:50:43\n",
       "SRS992705 sample29    sample29     2020-05-18T00:46:05\n",
       "SRS992692 sample11    sample11     2020-05-18T00:45:26\n",
       "SRS992710 sample24    sample24     2020-05-18T00:35:28\n",
       "SRS992714 sample35    sample35     2020-05-18T00:35:15\n",
       "SRS992704 sample30    sample30     2020-05-18T00:27:10\n",
       "SRS992713 sample36    sample36     2020-05-18T00:13:07\n",
       "SRS992696 sample23    sample23     2020-05-18T00:06:42\n",
       "SRS992729 sample06    sample06     2020-05-17T10:16:43\n",
       "SRS992701 sample18    sample18     2020-05-17T10:11:28\n",
       "SRS992691 sample12    sample12     2020-05-17T10:01:44\n",
       "SRS992721 sample05    sample05     2020-05-17T10:01:21\n",
       "SRS992720 sample42    sample42     2020-05-17T10:00:58\n",
       "SRS992719 sample41    sample41     2020-05-17T10:00:45\n",
       "          geographic location (longitude) â‹¯ instrument model   \n",
       "SRS992699 40.54                           â‹¯ Illumina HiSeq 2000\n",
       "SRS992702 38.46                           â‹¯ Illumina HiSeq 2000\n",
       "SRS992693 40.54                           â‹¯ Illumina HiSeq 2000\n",
       "SRS992705 37.3                            â‹¯ Illumina HiSeq 2000\n",
       "SRS992692 40.44                           â‹¯ Illumina HiSeq 2000\n",
       "SRS992710 37.55                           â‹¯ Illumina HiSeq 2000\n",
       "SRS992714 36.6                            â‹¯ Illumina HiSeq 2000\n",
       "SRS992704 37.3                            â‹¯ Illumina HiSeq 2000\n",
       "SRS992713 36.6                            â‹¯ Illumina HiSeq 2000\n",
       "SRS992696 37.55                           â‹¯ Illumina HiSeq 2000\n",
       "SRS992729 39.47                           â‹¯ Illumina HiSeq 2000\n",
       "SRS992701 38.46                           â‹¯ Illumina HiSeq 2000\n",
       "SRS992691 40.44                           â‹¯ Illumina HiSeq 2000\n",
       "SRS992721 39.47                           â‹¯ Illumina HiSeq 2000\n",
       "SRS992720 34.3                            â‹¯ Illumina HiSeq 2000\n",
       "SRS992719 34.3                            â‹¯ Illumina HiSeq 2000\n",
       "          last update date investigation type project name\n",
       "SRS992699 NA               NA                 NA          \n",
       "SRS992702 NA               NA                 NA          \n",
       "SRS992693 NA               NA                 NA          \n",
       "SRS992705 NA               NA                 NA          \n",
       "SRS992692 NA               NA                 NA          \n",
       "SRS992710 NA               NA                 NA          \n",
       "SRS992714 NA               NA                 NA          \n",
       "SRS992704 NA               NA                 NA          \n",
       "SRS992713 NA               NA                 NA          \n",
       "SRS992696 NA               NA                 NA          \n",
       "SRS992729 NA               NA                 NA          \n",
       "SRS992701 NA               NA                 NA          \n",
       "SRS992691 NA               NA                 NA          \n",
       "SRS992721 NA               NA                 NA          \n",
       "SRS992720 NA               NA                 NA          \n",
       "SRS992719 NA               NA                 NA          \n",
       "          geographic location (depth) geographic location (altitude)\n",
       "SRS992699 NA                          NA                            \n",
       "SRS992702 NA                          NA                            \n",
       "SRS992693 NA                          NA                            \n",
       "SRS992705 NA                          NA                            \n",
       "SRS992692 NA                          NA                            \n",
       "SRS992710 NA                          NA                            \n",
       "SRS992714 NA                          NA                            \n",
       "SRS992704 NA                          NA                            \n",
       "SRS992713 NA                          NA                            \n",
       "SRS992696 NA                          NA                            \n",
       "SRS992729 NA                          NA                            \n",
       "SRS992701 NA                          NA                            \n",
       "SRS992691 NA                          NA                            \n",
       "SRS992721 NA                          NA                            \n",
       "SRS992720 NA                          NA                            \n",
       "SRS992719 NA                          NA                            \n",
       "          environmental package sequencing method NCBI sample classification\n",
       "SRS992699 NA                    NA                NA                        \n",
       "SRS992702 NA                    NA                NA                        \n",
       "SRS992693 NA                    NA                NA                        \n",
       "SRS992705 NA                    NA                NA                        \n",
       "SRS992692 NA                    NA                NA                        \n",
       "SRS992710 NA                    NA                NA                        \n",
       "SRS992714 NA                    NA                NA                        \n",
       "SRS992704 NA                    NA                NA                        \n",
       "SRS992713 NA                    NA                NA                        \n",
       "SRS992696 NA                    NA                NA                        \n",
       "SRS992729 NA                    NA                NA                        \n",
       "SRS992701 NA                    NA                NA                        \n",
       "SRS992691 NA                    NA                NA                        \n",
       "SRS992721 NA                    NA                NA                        \n",
       "SRS992720 NA                    NA                NA                        \n",
       "SRS992719 NA                    NA                NA                        \n",
       "          ENA checklist\n",
       "SRS992699 NA           \n",
       "SRS992702 NA           \n",
       "SRS992693 NA           \n",
       "SRS992705 NA           \n",
       "SRS992692 NA           \n",
       "SRS992710 NA           \n",
       "SRS992714 NA           \n",
       "SRS992704 NA           \n",
       "SRS992713 NA           \n",
       "SRS992696 NA           \n",
       "SRS992729 NA           \n",
       "SRS992701 NA           \n",
       "SRS992691 NA           \n",
       "SRS992721 NA           \n",
       "SRS992720 NA           \n",
       "SRS992719 NA           "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "depth_numeric = as.numeric(lentic_samples$depth)  # We must convert data from MGnifyR (always strings) to numerical format.\n",
    "depth_numeric[is.na(depth_numeric)] = 0.0  # If depth data is missing, assume it is surface-level.\n",
    "lentic_subset = lentic_samples[depth_numeric >=25 & depth_numeric <=50,]  # Filter to samples collected between 25m and 50m down.\n",
    "lentic_subset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7934c0bf-aa6c-4c8e-85e7-aca272db3daf",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.1.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
