{
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "---\n",
        "title: \"Fetch analyses metadata for a study\"\n",
        "author:\n",
        "  - name: Noah de Gunst\n",
        "    affiliation:\n",
        "      - id: mia\n",
        "        name: Department of Computing, University of Turku, Finland\n",
        "  - name: Sam Hillman\n",
        "    affiliation:\n",
        "      - id: mia\n",
        "        name: Department of Computing, University of Turku, Finland\n",
        "categories: [R]\n",
        "execute: \n",
        "  eval: true\n",
        "---\n",
        "\n",
        "::: {style=\"max-width:1200px\"}\n",
        "![](./../_resources/mgnify_logo.png)\n",
        "\n",
        "# Fetch a study using MGnifyR; download the metadata for all of its analyses\n",
        "\n",
        "The [MGnify API](https://www.ebi.ac.uk/metagenomics/api/v1) returns data and \n",
        "relationships as JSON. [MGnifyR](https://www.bioconductor.org/packages/release/bioc/html/MGnifyR.html) \n",
        "is a package to help you read MGnify data into your R analyses.\n",
        "\n",
        "You can find all of the other \"API endpoints\" using the [Browsable API interface in your web browser](https://www.ebi.ac.uk/metagenomics/api/v1).\n",
        "\n",
        "This is an interactive code notebook (a Jupyter Notebook). To run this code, click \n",
        "into each cell and press the â–¶ button in the top toolbar, or press `shift+enter`.\n",
        "\n",
        "------------------------------------------------------------------------\n",
        ":::\n",
        "\n",
        "#### Setting the access code\n",
        "First, we need to specify the accession number of the study we're working with. \n",
        "This can be done by setting the `mgnify_study_accession` variable. The accession \n",
        "number uniquely identifies the study in the MGnify database.\n",
        "\n",
        "```{r}\n",
        "#| output: false\n",
        "\n",
        "source(\"./utils/variable_utils.r\")\n",
        "\n",
        "mgnify_study_accession <- get_variable_from_link_or_input('MGYS', 'Study Accession', \n",
        "                                                          'MGYS00005116')\n",
        "\n",
        "# You can also just directly set the accession variable in code, like this:\n",
        "# mgnify_study_accession <- \"MGYS00005292\"\n",
        "```\n",
        "\n",
        "#### Constructing a MgnifyClient object to access the database\n",
        "To interact with the MGnify database, we need to create an MgnifyClient object. \n",
        "This object allows us to fetch data from MGnify, and we can configure it to use \n",
        "a cache for efficiency. \n",
        "\n",
        "```{r}\n",
        "#| output: false\n",
        "# Importing the libraries\n",
        "library(vegan)\n",
        "library(ggplot2)\n",
        "library(mia)\n",
        "library(MGnifyR)\n",
        "\n",
        "# Check if the cache directory exists, if not, create it\n",
        "if (!dir.exists(\"./.mgnify_cache\")) {\n",
        "  dir.create(\"./.mgnify_cache\", recursive = TRUE)\n",
        "}\n",
        "\n",
        "# Create the MgnifyClient object with caching enabled\n",
        "mg <- MgnifyClient(usecache = TRUE, cache_dir = \"./.mgnify_cache\")\n",
        "```\n",
        "\n",
        "#### Displaying the help file\n",
        "\n",
        "```{r}\n",
        "#| output: false\n",
        "library(IRdisplay)\n",
        "display_markdown(file = '../_resources/mgnifyr_help.md')\n",
        "```\n",
        "\n",
        "## Fetch a list of the Analyses for the Study\n",
        "Using the MgnifyClient object, we can search for all analyses associated with the \n",
        "study accession number we set earlier. This will return a list of analysis accession \n",
        "numbers.\n",
        "\n",
        "```{r}\n",
        "#| output: false\n",
        "analyses_accessions <- searchAnalysis(mg, \"studies\", mgnify_study_accession)\n",
        "analyses_accessions\n",
        "```\n",
        "\n",
        "## Download metadata for the first 10 Analyses\n",
        "...and put it into a dataframe.\n",
        "\n",
        "```{r}\n",
        "#| output: false\n",
        "analyses_metadata_df <- getMetadata(mg, head(analyses_accessions, 10))\n",
        "```\n",
        "\n",
        "## Display metadata\n",
        "The table could be big, so let's look at a sample of it (`head`).\n",
        "\n",
        "```{r}\n",
        "#| output: false\n",
        "t(head(analyses_metadata_df))\n",
        "```\n",
        "\n",
        "## Download the data to a multi-assay data object\n",
        "\n",
        "> [mia](https://microbiome.github.io/mia/) is a Bioconductor package designed to \n",
        "import, store and analyze microbiome data using an object called a `TreeSummarizedExperiment.` \n",
        "This is a tailored data container optimized for microbiome data analysis.Being \n",
        "built on the `SummarizedExperiment` class, miaverse seamlessly integrates into the \n",
        "extensive `SummarizedExperiment` ecosystem. In this example we download the MGnifyR \n",
        "data to an MAE, which contains multiple `TreeSummarizedExperiment` objects.\n",
        "\n",
        "\n",
        "```{r}\n",
        "#| output: false\n",
        "mae <- getResult(mg, accession = analyses_accessions)\n",
        "```\n",
        "\n",
        "You use `MGnifyR` features further, for example to download data. Check the Cheat \n",
        "Sheet at the top for more."
      ]
    }
  ],
  "metadata": {
    "kernelspec": {
      "display_name": "Python 3",
      "language": "python",
      "name": "python3"
    }
  },
  "nbformat": 4,
  "nbformat_minor": 4
}