{
 "cells": [
  {
   "cell_type": "raw",
   "id": "003ddb34-a276-4b23-8888-9f514272b7b8",
   "metadata": {
    "tags": []
   },
   "source": [
    "---\n",
    "title: \"Interactive map for AtlantECO project\"\n",
    "author: \"Kate S [Ekaterina Sakharova] (MGnify team)\"\n",
    "categories: [Python]\n",
    "execute: \n",
    "  enabled: false\n",
    "  eval: false\n",
    "---"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9c3ee428-a8a6-42a8-b2d8-f7df41987a78",
   "metadata": {},
   "source": [
    "<div style=\"max-width:1200px\"><img src=\"../_resources/mgnify_banner.png\" width=\"100%\"></div>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "94de3e3b-c12a-4472-8ffe-710494e44bd1",
   "metadata": {},
   "source": [
    "<img src=\"../_resources/mgnify_logo.png\" width=\"200px\">\n",
    "\n",
    "# Mapping samples from the [AtlantECO Super Study](https://www.ebi.ac.uk/metagenomics/super-studies/atlanteco)\n",
    "### ... using the MGnify API and an interactive map widget\n",
    "\n",
    "The [MGnify API](https://www.ebi.ac.uk/metagenomics/api/v1) returns JSON data. The `jsonapi_client` package can help you load this data into Python, e.g. into a Pandas dataframe.\n",
    "\n",
    "**This example shows you how to load a MGnify AtlantECO Super Study's data from the MGnify API and display it on an interactive world map**\n",
    "\n",
    "You can find all of the other \"API endpoints\" using the [Browsable API interface in your web browser](https://www.ebi.ac.uk/metagenomics/api/v1). The URL you see in the browsable API is exactly the same as the one you can use in this code.\n",
    "\n",
    "This is an interactive code notebook (a Jupyter Notebook).\n",
    "To run this code, click into each cell and press the ▶ button in the top toolbar, or press `shift+enter`.\n",
    "\n",
    "**Content:**\n",
    "- Fetch AtlantECO studies data\n",
    "- Show study' samples on the interactive world map\n",
    "- Check functional annotation terms presence/absense\n",
    "    - GO-term\n",
    "    - InterPro \n",
    "    - Biosynthetic Gene Clusters (BGC)\n",
    "    \n",
    "\n",
    "---"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3cd6dd3f-2ff7-4d27-947c-54e648b0912b",
   "metadata": {},
   "source": [
    "## Fetch all [AtlantECO](https://www.ebi.ac.uk/metagenomics/super-studies/atlanteco) studies\n",
    "A Super Study is a collection of MGnify Studies originating from a major project. AtlantECO is one such project, aiming to develop and apply a novel, unifying framework that provides knowledge-based resources for a better understanding and management of the Atlantic Ocean and its ecosystem services.\n",
    "\n",
    "Fetch the Super Study's Studies from the MGnify API, into a [Pandas dataframe](https://pandas.pydata.org/docs/user_guide/index.html):"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "id": "78657238-8961-4fe9-ac64-02dda587c8b8",
   "metadata": {
    "is_executing": true,
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>type</th>\n",
       "      <th>id</th>\n",
       "      <th>attributes.bioproject</th>\n",
       "      <th>attributes.accession</th>\n",
       "      <th>attributes.samples-count</th>\n",
       "      <th>attributes.is-private</th>\n",
       "      <th>attributes.last-update</th>\n",
       "      <th>attributes.secondary-accession</th>\n",
       "      <th>attributes.centre-name</th>\n",
       "      <th>attributes.study-abstract</th>\n",
       "      <th>attributes.study-name</th>\n",
       "      <th>attributes.data-origination</th>\n",
       "      <th>relationships.biomes.data</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>studies</td>\n",
       "      <td>MGYS00006613</td>\n",
       "      <td>PRJEB40759</td>\n",
       "      <td>MGYS00006613</td>\n",
       "      <td>58</td>\n",
       "      <td>False</td>\n",
       "      <td>2024-03-01T18:29:37</td>\n",
       "      <td>ERP124426</td>\n",
       "      <td>Ocean Sampling Day Consortium</td>\n",
       "      <td>Ocean Sampling Day was initiated by the EU-fun...</td>\n",
       "      <td>18S rRNA amplicon sequencing from the Ocean Sa...</td>\n",
       "      <td>SUBMITTED</td>\n",
       "      <td>[{'id': 'root:Environmental:Aquatic:Marine', '...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>studies</td>\n",
       "      <td>MGYS00006612</td>\n",
       "      <td>PRJEB40763</td>\n",
       "      <td>MGYS00006612</td>\n",
       "      <td>48</td>\n",
       "      <td>False</td>\n",
       "      <td>2024-03-01T18:14:18</td>\n",
       "      <td>ERP124432</td>\n",
       "      <td>Ocean Sampling Day Consortium</td>\n",
       "      <td>Ocean Sampling Day was initiated by the EU-fun...</td>\n",
       "      <td>18S rRNA amplicon sequencing from the Ocean Sa...</td>\n",
       "      <td>SUBMITTED</td>\n",
       "      <td>[{'id': 'root:Environmental:Aquatic:Marine', '...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>studies</td>\n",
       "      <td>MGYS00006611</td>\n",
       "      <td>PRJEB55999</td>\n",
       "      <td>MGYS00006611</td>\n",
       "      <td>63</td>\n",
       "      <td>False</td>\n",
       "      <td>2024-03-01T18:01:09</td>\n",
       "      <td>ERP140920</td>\n",
       "      <td>Ocean Sampling Day Consortium</td>\n",
       "      <td>Ocean Sampling Day was initiated by the EU-fun...</td>\n",
       "      <td>18S rRNA amplicon sequencing from the Ocean Sa...</td>\n",
       "      <td>SUBMITTED</td>\n",
       "      <td>[{'id': 'root:Environmental:Aquatic:Marine', '...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>studies</td>\n",
       "      <td>MGYS00006610</td>\n",
       "      <td>PRJEB56005</td>\n",
       "      <td>MGYS00006610</td>\n",
       "      <td>50</td>\n",
       "      <td>False</td>\n",
       "      <td>2024-03-01T17:44:36</td>\n",
       "      <td>ERP140926</td>\n",
       "      <td>Ocean Sampling Day Consortium</td>\n",
       "      <td>Ocean Sampling Day was initiated by the EU-fun...</td>\n",
       "      <td>18S rRNA amplicon sequencing from the Ocean Sa...</td>\n",
       "      <td>SUBMITTED</td>\n",
       "      <td>[{'id': 'root:Environmental:Aquatic:Marine', '...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>studies</td>\n",
       "      <td>MGYS00006609</td>\n",
       "      <td>PRJEB9737</td>\n",
       "      <td>MGYS00006609</td>\n",
       "      <td>193</td>\n",
       "      <td>False</td>\n",
       "      <td>2024-03-01T17:30:57</td>\n",
       "      <td>ERP010877</td>\n",
       "      <td>GSC</td>\n",
       "      <td>Analysis of 18S DNA in Tara Oceans Polar Circl...</td>\n",
       "      <td>Amplicon sequencing of Tara Oceans Polar Circl...</td>\n",
       "      <td>SUBMITTED</td>\n",
       "      <td>[{'id': 'root:Environmental:Aquatic:Marine', '...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      type            id attributes.bioproject attributes.accession  \\\n",
       "0  studies  MGYS00006613            PRJEB40759         MGYS00006613   \n",
       "1  studies  MGYS00006612            PRJEB40763         MGYS00006612   \n",
       "2  studies  MGYS00006611            PRJEB55999         MGYS00006611   \n",
       "3  studies  MGYS00006610            PRJEB56005         MGYS00006610   \n",
       "4  studies  MGYS00006609             PRJEB9737         MGYS00006609   \n",
       "\n",
       "   attributes.samples-count  attributes.is-private attributes.last-update  \\\n",
       "0                        58                  False    2024-03-01T18:29:37   \n",
       "1                        48                  False    2024-03-01T18:14:18   \n",
       "2                        63                  False    2024-03-01T18:01:09   \n",
       "3                        50                  False    2024-03-01T17:44:36   \n",
       "4                       193                  False    2024-03-01T17:30:57   \n",
       "\n",
       "  attributes.secondary-accession         attributes.centre-name  \\\n",
       "0                      ERP124426  Ocean Sampling Day Consortium   \n",
       "1                      ERP124432  Ocean Sampling Day Consortium   \n",
       "2                      ERP140920  Ocean Sampling Day Consortium   \n",
       "3                      ERP140926  Ocean Sampling Day Consortium   \n",
       "4                      ERP010877                            GSC   \n",
       "\n",
       "                           attributes.study-abstract  \\\n",
       "0  Ocean Sampling Day was initiated by the EU-fun...   \n",
       "1  Ocean Sampling Day was initiated by the EU-fun...   \n",
       "2  Ocean Sampling Day was initiated by the EU-fun...   \n",
       "3  Ocean Sampling Day was initiated by the EU-fun...   \n",
       "4  Analysis of 18S DNA in Tara Oceans Polar Circl...   \n",
       "\n",
       "                               attributes.study-name  \\\n",
       "0  18S rRNA amplicon sequencing from the Ocean Sa...   \n",
       "1  18S rRNA amplicon sequencing from the Ocean Sa...   \n",
       "2  18S rRNA amplicon sequencing from the Ocean Sa...   \n",
       "3  18S rRNA amplicon sequencing from the Ocean Sa...   \n",
       "4  Amplicon sequencing of Tara Oceans Polar Circl...   \n",
       "\n",
       "  attributes.data-origination  \\\n",
       "0                   SUBMITTED   \n",
       "1                   SUBMITTED   \n",
       "2                   SUBMITTED   \n",
       "3                   SUBMITTED   \n",
       "4                   SUBMITTED   \n",
       "\n",
       "                           relationships.biomes.data  \n",
       "0  [{'id': 'root:Environmental:Aquatic:Marine', '...  \n",
       "1  [{'id': 'root:Environmental:Aquatic:Marine', '...  \n",
       "2  [{'id': 'root:Environmental:Aquatic:Marine', '...  \n",
       "3  [{'id': 'root:Environmental:Aquatic:Marine', '...  \n",
       "4  [{'id': 'root:Environmental:Aquatic:Marine', '...  "
      ]
     },
     "execution_count": 75,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "from jsonapi_client import Session, Modifier\n",
    "\n",
    "atlanteco_endpoint = 'super-studies/atlanteco/flagship-studies'\n",
    "with Session(\"https://www.ebi.ac.uk/metagenomics/api/v1\") as mgnify:\n",
    "    studies = map(lambda r: r.json, mgnify.iterate(atlanteco_endpoint))\n",
    "    studies = pd.json_normalize(studies)\n",
    "studies[:5]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5ae10b66-889b-4932-b62d-37b89feac050",
   "metadata": {},
   "source": [
    "## Show the studies' samples on a map\n",
    "\n",
    "We can fetch the Samples for each Study, and concatenate them all into one Dataframe.\n",
    "Each sample has geolocation data in its `attributes` - this is what we need to build a map.\n",
    "\n",
    "It takes time to fetch data for all samples, so **let's show samples from chosen PRJEB46727 study only.** This study contain assembly data https://www.ebi.ac.uk/metagenomics/studies/MGYS00005810#overview."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "id": "bc5d0fc9-5c05-4bb1-b401-507b6e8c8877",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "fetching MGYS00005810 samples\n"
     ]
    }
   ],
   "source": [
    "substudy = studies[studies['attributes.bioproject'] == 'PRJEB46727']\n",
    "studies_samples = []\n",
    "\n",
    "with Session(\"https://www.ebi.ac.uk/metagenomics/api/v1\") as mgnify:\n",
    "    for idx, study in substudy.iterrows():\n",
    "        print(f\"fetching {study.id} samples\")\n",
    "        samples = map(lambda r: r.json, mgnify.iterate(f'studies/{study.id}/samples?page_size=1000'))\n",
    "        samples = pd.json_normalize(samples)\n",
    "        samples = pd.DataFrame(data={\n",
    "            'accession': samples['id'],\n",
    "            'sample_id': samples['id'],\n",
    "            'study': study.id, \n",
    "            'lon': samples['attributes.longitude'],\n",
    "            'lat': samples['attributes.latitude'],\n",
    "            'color': \"#FF0000\",\n",
    "        })\n",
    "        samples.set_index('accession', inplace=True)\n",
    "        studies_samples.append(samples)\n",
    "studies_samples = pd.concat(studies_samples)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "id": "2f503c45-d3c0-45b8-a9d5-e3d3eb69ce95",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "fetched 83 samples\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_id</th>\n",
       "      <th>study</th>\n",
       "      <th>lon</th>\n",
       "      <th>lat</th>\n",
       "      <th>color</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>accession</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>SRS836773</th>\n",
       "      <td>SRS836773</td>\n",
       "      <td>MGYS00005810</td>\n",
       "      <td>-50.6225</td>\n",
       "      <td>-0.1569</td>\n",
       "      <td>#FF0000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRS567311</th>\n",
       "      <td>SRS567311</td>\n",
       "      <td>MGYS00005810</td>\n",
       "      <td>-53.0000</td>\n",
       "      <td>7.2900</td>\n",
       "      <td>#FF0000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRS567886</th>\n",
       "      <td>SRS567886</td>\n",
       "      <td>MGYS00005810</td>\n",
       "      <td>-54.4200</td>\n",
       "      <td>10.6800</td>\n",
       "      <td>#FF0000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRS580500</th>\n",
       "      <td>SRS580500</td>\n",
       "      <td>MGYS00005810</td>\n",
       "      <td>-52.2200</td>\n",
       "      <td>12.4100</td>\n",
       "      <td>#FF0000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRS580495</th>\n",
       "      <td>SRS580495</td>\n",
       "      <td>MGYS00005810</td>\n",
       "      <td>-54.5100</td>\n",
       "      <td>10.2900</td>\n",
       "      <td>#FF0000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           sample_id         study      lon      lat    color\n",
       "accession                                                    \n",
       "SRS836773  SRS836773  MGYS00005810 -50.6225  -0.1569  #FF0000\n",
       "SRS567311  SRS567311  MGYS00005810 -53.0000   7.2900  #FF0000\n",
       "SRS567886  SRS567886  MGYS00005810 -54.4200  10.6800  #FF0000\n",
       "SRS580500  SRS580500  MGYS00005810 -52.2200  12.4100  #FF0000\n",
       "SRS580495  SRS580495  MGYS00005810 -54.5100  10.2900  #FF0000"
      ]
     },
     "execution_count": 77,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(f\"fetched {len(studies_samples)} samples\")\n",
    "\n",
    "studies_samples.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "id": "6c09ddf3-746e-4891-96a5-8f70682d530f",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "6ef86a64cf4c422da77e6367d252b6bb",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Map(center=[0, 0], controls=(ZoomControl(options=['position', 'zoom_in_text', 'zoom_in_title', 'zoom_out_text'…"
      ]
     },
     "execution_count": 78,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import leafmap\n",
    "m = leafmap.Map(center=(0, 0), zoom=2)\n",
    "m.add_points_from_xy(\n",
    "    studies_samples,\n",
    "    x='lon', \n",
    "    y='lat', \n",
    "    popup=[\"study\", \"sample_id\"], \n",
    "    color_column='color',\n",
    "    add_legend=False\n",
    ")\n",
    "m"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7fc7b990-0804-43a3-bfc8-dc14e5a61a88",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Check functional annotation terms presence\n",
    "Let's check whether a specific identifier is present in each sample. \n",
    "\n",
    "We will work with MGnify analyses (`MGYA`s) corresponding to chosen samples. We filter analyses by \n",
    "- pipeline version: 5.0\n",
    "- experiment type: assembly\n",
    "\n",
    "This example shows how to process **just the first 10 samples** (again, because the full dataset takes a while to fetch).\n",
    "Firstly, get analyses for each sample."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "id": "8c7d6790-1cec-4ecf-b8be-56e3e03fe669",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "processing SRS836773\n",
      "processing SRS567311\n",
      "processing SRS567886\n",
      "processing SRS580500\n",
      "processing SRS580495\n",
      "processing SRS565748\n",
      "processing SRS1776191\n",
      "processing SRS581965\n",
      "processing SRS837379\n",
      "processing SRS1776201\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>type</th>\n",
       "      <th>id</th>\n",
       "      <th>attributes.analysis-status</th>\n",
       "      <th>attributes.accession</th>\n",
       "      <th>attributes.experiment-type</th>\n",
       "      <th>attributes.analysis-summary</th>\n",
       "      <th>attributes.pipeline-version</th>\n",
       "      <th>attributes.is-private</th>\n",
       "      <th>attributes.last-update</th>\n",
       "      <th>attributes.complete-time</th>\n",
       "      <th>attributes.instrument-platform</th>\n",
       "      <th>attributes.instrument-model</th>\n",
       "      <th>relationships.assembly.data.id</th>\n",
       "      <th>relationships.assembly.data.type</th>\n",
       "      <th>relationships.study.data.id</th>\n",
       "      <th>relationships.study.data.type</th>\n",
       "      <th>relationships.sample.data.id</th>\n",
       "      <th>relationships.sample.data.type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>analysis-jobs</td>\n",
       "      <td>MGYA00593142</td>\n",
       "      <td>completed</td>\n",
       "      <td>MGYA00593142</td>\n",
       "      <td>assembly</td>\n",
       "      <td>[{'key': 'Submitted nucleotide sequences', 'va...</td>\n",
       "      <td>5.0</td>\n",
       "      <td>False</td>\n",
       "      <td>2024-01-29T15:29:19.757516</td>\n",
       "      <td>2021-12-06T19:31:17</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>Illumina HiSeq 2500</td>\n",
       "      <td>ERZ2945023</td>\n",
       "      <td>assemblies</td>\n",
       "      <td>MGYS00005810</td>\n",
       "      <td>studies</td>\n",
       "      <td>SRS836773</td>\n",
       "      <td>samples</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>analysis-jobs</td>\n",
       "      <td>MGYA00589840</td>\n",
       "      <td>completed</td>\n",
       "      <td>MGYA00589840</td>\n",
       "      <td>assembly</td>\n",
       "      <td>[{'key': 'Submitted nucleotide sequences', 'va...</td>\n",
       "      <td>5.0</td>\n",
       "      <td>False</td>\n",
       "      <td>2024-01-29T15:29:19.757516</td>\n",
       "      <td>2021-10-22T17:25:45</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>Illumina Genome Analyzer IIx</td>\n",
       "      <td>ERZ2945090</td>\n",
       "      <td>assemblies</td>\n",
       "      <td>MGYS00005810</td>\n",
       "      <td>studies</td>\n",
       "      <td>SRS567311</td>\n",
       "      <td>samples</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>analysis-jobs</td>\n",
       "      <td>MGYA00589562</td>\n",
       "      <td>completed</td>\n",
       "      <td>MGYA00589562</td>\n",
       "      <td>assembly</td>\n",
       "      <td>[{'key': 'Submitted nucleotide sequences', 'va...</td>\n",
       "      <td>5.0</td>\n",
       "      <td>False</td>\n",
       "      <td>2024-01-29T15:29:19.757516</td>\n",
       "      <td>2021-10-21T08:23:57</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>Illumina Genome Analyzer IIx</td>\n",
       "      <td>ERZ2945101</td>\n",
       "      <td>assemblies</td>\n",
       "      <td>MGYS00005810</td>\n",
       "      <td>studies</td>\n",
       "      <td>SRS567886</td>\n",
       "      <td>samples</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>analysis-jobs</td>\n",
       "      <td>MGYA00589561</td>\n",
       "      <td>completed</td>\n",
       "      <td>MGYA00589561</td>\n",
       "      <td>assembly</td>\n",
       "      <td>[{'key': 'Submitted nucleotide sequences', 'va...</td>\n",
       "      <td>5.0</td>\n",
       "      <td>False</td>\n",
       "      <td>2024-01-29T15:29:19.757516</td>\n",
       "      <td>2021-10-21T08:16:10</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>Illumina Genome Analyzer IIx</td>\n",
       "      <td>ERZ2944669</td>\n",
       "      <td>assemblies</td>\n",
       "      <td>MGYS00005810</td>\n",
       "      <td>studies</td>\n",
       "      <td>SRS580500</td>\n",
       "      <td>samples</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>analysis-jobs</td>\n",
       "      <td>MGYA00589560</td>\n",
       "      <td>completed</td>\n",
       "      <td>MGYA00589560</td>\n",
       "      <td>assembly</td>\n",
       "      <td>[{'key': 'Submitted nucleotide sequences', 'va...</td>\n",
       "      <td>5.0</td>\n",
       "      <td>False</td>\n",
       "      <td>2024-01-29T15:29:19.757516</td>\n",
       "      <td>2021-10-21T08:13:12</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>Illumina Genome Analyzer IIx</td>\n",
       "      <td>ERZ2944798</td>\n",
       "      <td>assemblies</td>\n",
       "      <td>MGYS00005810</td>\n",
       "      <td>studies</td>\n",
       "      <td>SRS580495</td>\n",
       "      <td>samples</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            type            id attributes.analysis-status  \\\n",
       "0  analysis-jobs  MGYA00593142                  completed   \n",
       "0  analysis-jobs  MGYA00589840                  completed   \n",
       "0  analysis-jobs  MGYA00589562                  completed   \n",
       "0  analysis-jobs  MGYA00589561                  completed   \n",
       "0  analysis-jobs  MGYA00589560                  completed   \n",
       "\n",
       "  attributes.accession attributes.experiment-type  \\\n",
       "0         MGYA00593142                   assembly   \n",
       "0         MGYA00589840                   assembly   \n",
       "0         MGYA00589562                   assembly   \n",
       "0         MGYA00589561                   assembly   \n",
       "0         MGYA00589560                   assembly   \n",
       "\n",
       "                         attributes.analysis-summary  \\\n",
       "0  [{'key': 'Submitted nucleotide sequences', 'va...   \n",
       "0  [{'key': 'Submitted nucleotide sequences', 'va...   \n",
       "0  [{'key': 'Submitted nucleotide sequences', 'va...   \n",
       "0  [{'key': 'Submitted nucleotide sequences', 'va...   \n",
       "0  [{'key': 'Submitted nucleotide sequences', 'va...   \n",
       "\n",
       "  attributes.pipeline-version  attributes.is-private  \\\n",
       "0                         5.0                  False   \n",
       "0                         5.0                  False   \n",
       "0                         5.0                  False   \n",
       "0                         5.0                  False   \n",
       "0                         5.0                  False   \n",
       "\n",
       "       attributes.last-update attributes.complete-time  \\\n",
       "0  2024-01-29T15:29:19.757516      2021-12-06T19:31:17   \n",
       "0  2024-01-29T15:29:19.757516      2021-10-22T17:25:45   \n",
       "0  2024-01-29T15:29:19.757516      2021-10-21T08:23:57   \n",
       "0  2024-01-29T15:29:19.757516      2021-10-21T08:16:10   \n",
       "0  2024-01-29T15:29:19.757516      2021-10-21T08:13:12   \n",
       "\n",
       "  attributes.instrument-platform   attributes.instrument-model  \\\n",
       "0                       ILLUMINA           Illumina HiSeq 2500   \n",
       "0                       ILLUMINA  Illumina Genome Analyzer IIx   \n",
       "0                       ILLUMINA  Illumina Genome Analyzer IIx   \n",
       "0                       ILLUMINA  Illumina Genome Analyzer IIx   \n",
       "0                       ILLUMINA  Illumina Genome Analyzer IIx   \n",
       "\n",
       "  relationships.assembly.data.id relationships.assembly.data.type  \\\n",
       "0                     ERZ2945023                       assemblies   \n",
       "0                     ERZ2945090                       assemblies   \n",
       "0                     ERZ2945101                       assemblies   \n",
       "0                     ERZ2944669                       assemblies   \n",
       "0                     ERZ2944798                       assemblies   \n",
       "\n",
       "  relationships.study.data.id relationships.study.data.type  \\\n",
       "0                MGYS00005810                       studies   \n",
       "0                MGYS00005810                       studies   \n",
       "0                MGYS00005810                       studies   \n",
       "0                MGYS00005810                       studies   \n",
       "0                MGYS00005810                       studies   \n",
       "\n",
       "  relationships.sample.data.id relationships.sample.data.type  \n",
       "0                    SRS836773                        samples  \n",
       "0                    SRS567311                        samples  \n",
       "0                    SRS567886                        samples  \n",
       "0                    SRS580500                        samples  \n",
       "0                    SRS580495                        samples  "
      ]
     },
     "execution_count": 79,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "analyses = []\n",
    "with Session(\"https://www.ebi.ac.uk/metagenomics/api/v1\") as mgnify:\n",
    "    for idx, sample in studies_samples[:10].iterrows():\n",
    "        print(f\"processing {sample.sample_id}\")\n",
    "        filtering = Modifier(f\"pipeline_version=5.0&sample_accession={sample.sample_id}&experiment_type=assembly\")\n",
    "        analysis = map(lambda r: r.json, mgnify.iterate('analyses', filter=filtering))\n",
    "        analysis = pd.json_normalize(analysis)\n",
    "        analyses.append(analysis)\n",
    "analyses = pd.concat(analyses)\n",
    "analyses[:5]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0b4ad3aa-f19f-4007-b64f-59edf10b5734",
   "metadata": {},
   "source": [
    "Define functions:\n",
    "- `identify_existance` to check each analysis for identifier presence/absence. We add a column to the dataframe with a colour: blue if identifier was found and red if not.\n",
    "- `show_on_map` to plot results on the world map. Join the analyses and sample tables to have geolocation data and identifier presence data together (we'll create a new sub-DataFrame with a subset of the fields and add them to the map)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "id": "7454b97b-2f9d-45b9-bc96-f1877cf3b58a",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "def identify_existance(input_analyses, identifier, term):\n",
    "    data = []\n",
    "    with Session(\"https://www.ebi.ac.uk/metagenomics/api/v1\") as mgnify:\n",
    "        for idx, mgya in input_analyses.iterrows():\n",
    "            print(f\"processing {mgya.id}\")\n",
    "            analysis_identifier = map(lambda r: r.json, mgnify.iterate(f'analyses/{mgya.id}/{identifier}'))\n",
    "            analysis_identifier = pd.json_normalize(analysis_identifier)\n",
    "            data.append(\"#0000FF\" if term in list(analysis_identifier.id) else \"#FF0000\")\n",
    "        presented = sum([1 for i in data if i == \"#0000FF\"])\n",
    "        print(f\"Presented {presented} of {identifier} {term}\")\n",
    "        input_analyses.insert(2, identifier, data, True)\n",
    "    return input_analyses\n",
    "\n",
    "def show_on_map(input_analyses, studies_samples, identifier):\n",
    "    df = input_analyses.join(studies_samples.set_index('sample_id'), on='relationships.sample.data.id')\n",
    "    df2 = df[[identifier, 'lon', 'lat', 'study', 'attributes.accession', 'relationships.study.data.id', 'relationships.sample.data.id', 'relationships.assembly.data.id']].copy()\n",
    "    df2 = df2.set_index(\"study\")\n",
    "    df2 = df2.rename(columns={\"attributes.accession\": \"analysis_ID\", \n",
    "                              'relationships.study.data.id': \"study_ID\",\n",
    "                              'relationships.sample.data.id': \"sample_ID\", \n",
    "                              'relationships.assembly.data.id': \"assembly_ID\"\n",
    "                             })\n",
    "    m = leafmap.Map(center=(0, 0), zoom=2)\n",
    "    m.add_points_from_xy(df2, \n",
    "                         x='lon', \n",
    "                         y='lat', \n",
    "                         popup=[\"study_ID\", \"sample_ID\", \"assembly_ID\", \"analysis_ID\"],\n",
    "                        color_column=identifier, add_legend=False)\n",
    "    return m"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "720a6d00-7976-4881-aed1-cd1fc756fb82",
   "metadata": {},
   "source": [
    "## GO term\n",
    "This example is written for GO-term for biotin transport [GO:0015878](http://www.candidagenome.org/cgi-bin/GO/go.pl?goid=15878)\n",
    "\n",
    "Other GO identifiers are available on the MGnify API."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "id": "20035a8f-7957-4d84-9099-0f930287fd69",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "processing MGYA00593142\n",
      "processing MGYA00589840\n",
      "processing MGYA00589562\n",
      "processing MGYA00589561\n",
      "processing MGYA00589560\n",
      "processing MGYA00589559\n",
      "processing MGYA00589558\n",
      "processing MGYA00589557\n",
      "processing MGYA00589556\n",
      "processing MGYA00589555\n",
      "Presented 9 of go-terms GO:0015878\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "1ad12d9989244fbf8884f3bd481fb4e8",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Map(center=[0, 0], controls=(ZoomControl(options=['position', 'zoom_in_text', 'zoom_in_title', 'zoom_out_text'…"
      ]
     },
     "execution_count": 81,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "identifier = \"go-terms\"\n",
    "go_term = 'GO:0015878'\n",
    "go_analyses = analyses\n",
    "go_data = identify_existance(go_analyses, identifier, go_term)\n",
    "map_vis = show_on_map(go_data, studies_samples, identifier)\n",
    "map_vis"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "923ad300-d764-499c-bd76-c3bd89e78d36",
   "metadata": {
    "tags": []
   },
   "source": [
    "## InterPro entry\n",
    "This example is written for InterPro entry [IPR001650](https://www.ebi.ac.uk/interpro/entry/InterPro/IPR001650): Helicase, C-terminal domain-like \n",
    "\n",
    "Other IPS identifiers are available on the MGnify API."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "id": "0ec7bbb7-0538-4d3d-b58e-efc10db32549",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "processing MGYA00593142\n",
      "processing MGYA00589840\n",
      "processing MGYA00589562\n",
      "processing MGYA00589561\n",
      "processing MGYA00589560\n",
      "processing MGYA00589559\n",
      "processing MGYA00589558\n",
      "processing MGYA00589557\n",
      "processing MGYA00589556\n",
      "processing MGYA00589555\n",
      "Presented 0 of interpro-identifiers GO:0015878\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "6bbf06b4f71e4a838371b7c5864e29a6",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Map(center=[0, 0], controls=(ZoomControl(options=['position', 'zoom_in_text', 'zoom_in_title', 'zoom_out_text'…"
      ]
     },
     "execution_count": 86,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "identifier = \"interpro-identifiers\"\n",
    "ips_term = 'IPR001650'\n",
    "ips_analyses = analyses\n",
    "ips_data = identify_existance(ips_analyses, identifier, ips_term)\n",
    "map_vis = show_on_map(ips_data, studies_samples, identifier)\n",
    "map_vis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ae2b8468-58e4-4032-ac1e-dd0f655fdaa3",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python (mgnify-py-env)",
   "language": "python",
   "name": "conda-env-mgnify-py-env-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
