{
 "cells": [
  {
   "cell_type": "raw",
   "id": "003ddb34-a276-4b23-8888-9f514272b7b8",
   "metadata": {
    "jupyter": {
     "source_hidden": true
    },
    "tags": []
   },
   "source": [
    "---\n",
    "title: \"Interactive map for AtlantECO project\"\n",
    "author: \"Kate S [Ekaterina Sakharova] (MGnify team)\"\n",
    "categories: [Python]\n",
    "execute: \n",
    "  enabled: true\n",
    "---"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9c3ee428-a8a6-42a8-b2d8-f7df41987a78",
   "metadata": {},
   "source": [
    "<div style=\"max-width:1200px\"><img src=\"../_resources/mgnify_banner.png\" width=\"100%\"></div>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "94de3e3b-c12a-4472-8ffe-710494e44bd1",
   "metadata": {},
   "source": [
    "<img src=\"../_resources/mgnify_logo.png\" width=\"200px\">\n",
    "\n",
    "# Mapping samples from the [AtlantECO Super Study](https://www.ebi.ac.uk/metagenomics/super-studies/atlanteco)\n",
    "### ... using the MGnify API and an interactive map widget\n",
    "\n",
    "The [MGnify API](https://www.ebi.ac.uk/metagenomics/api/v1) returns JSON data. The `jsonapi_client` package can help you load this data into Python, e.g. into a Pandas dataframe.\n",
    "\n",
    "**This example shows you how to load a MGnify Super Study's data from the MGnify API and display it on an interactive world map**\n",
    "\n",
    "You can find all of the other \"API endpoints\" using the [Browsable API interface in your web browser](https://www.ebi.ac.uk/metagenomics/api/v1). The URL you see in the browsable API is exactly the same as the one you can use in this code.\n",
    "\n",
    "This is an interactive code notebook (a Jupyter Notebook).\n",
    "To run this code, click into each cell and press the ▶ button in the top toolbar, or press `shift+enter`.\n",
    "\n",
    "---"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3cd6dd3f-2ff7-4d27-947c-54e648b0912b",
   "metadata": {},
   "source": [
    "## Fetch all [AtlantECO](https://www.ebi.ac.uk/metagenomics/super-studies/atlanteco) studies\n",
    "A Super Study is a collection of MGnify Studies originating from a major project. AtlantECO is one such project, aiming to develop and apply a novel, unifying framework that provides knowledge-based resources for a better understanding and management of the Atlantic Ocean and its ecosystem services.\n",
    "\n",
    "Fetch the Super Study's Studies from the MGnify API, into a [Pandas dataframe](https://pandas.pydata.org/docs/user_guide/index.html):"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "78657238-8961-4fe9-ac64-02dda587c8b8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>type</th>\n",
       "      <th>id</th>\n",
       "      <th>attributes.samples-count</th>\n",
       "      <th>attributes.accession</th>\n",
       "      <th>attributes.bioproject</th>\n",
       "      <th>attributes.is-private</th>\n",
       "      <th>attributes.secondary-accession</th>\n",
       "      <th>attributes.centre-name</th>\n",
       "      <th>attributes.study-abstract</th>\n",
       "      <th>attributes.study-name</th>\n",
       "      <th>attributes.data-origination</th>\n",
       "      <th>attributes.last-update</th>\n",
       "      <th>relationships.biomes.data</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>studies</td>\n",
       "      <td>MGYS00006074</td>\n",
       "      <td>75</td>\n",
       "      <td>MGYS00006074</td>\n",
       "      <td>PRJEB51168</td>\n",
       "      <td>False</td>\n",
       "      <td>ERP135767</td>\n",
       "      <td>EMG</td>\n",
       "      <td>The Third Party Annotation (TPA) assembly was ...</td>\n",
       "      <td>EMG produced TPA metagenomics assembly of PRJE...</td>\n",
       "      <td>SUBMITTED</td>\n",
       "      <td>2022-10-21T19:56:09</td>\n",
       "      <td>[{'id': 'root:Environmental:Aquatic:Marine', '...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>studies</td>\n",
       "      <td>MGYS00006072</td>\n",
       "      <td>130</td>\n",
       "      <td>MGYS00006072</td>\n",
       "      <td>PRJEB54918</td>\n",
       "      <td>False</td>\n",
       "      <td>ERP139784</td>\n",
       "      <td>EMG</td>\n",
       "      <td>The Third Party Annotation (TPA) assembly was ...</td>\n",
       "      <td>EMG produced TPA metagenomics assembly of PRJN...</td>\n",
       "      <td>SUBMITTED</td>\n",
       "      <td>2022-10-13T14:03:52</td>\n",
       "      <td>[{'id': 'root:Environmental:Aquatic:Marine:Oce...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>studies</td>\n",
       "      <td>MGYS00006034</td>\n",
       "      <td>971</td>\n",
       "      <td>MGYS00006034</td>\n",
       "      <td>PRJEB50181</td>\n",
       "      <td>False</td>\n",
       "      <td>ERP134737</td>\n",
       "      <td>EMG</td>\n",
       "      <td>The Third Party Annotation (TPA) assembly was ...</td>\n",
       "      <td>EMG produced TPA metagenomics assembly of PRJN...</td>\n",
       "      <td>SUBMITTED</td>\n",
       "      <td>2022-10-10T12:22:12</td>\n",
       "      <td>[{'id': 'root:Environmental:Aquatic:Marine', '...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>studies</td>\n",
       "      <td>MGYS00006061</td>\n",
       "      <td>3</td>\n",
       "      <td>MGYS00006061</td>\n",
       "      <td>PRJEB52804</td>\n",
       "      <td>False</td>\n",
       "      <td>ERP137544</td>\n",
       "      <td>EMG</td>\n",
       "      <td>The Third Party Annotation (TPA) assembly was ...</td>\n",
       "      <td>EMG produced TPA metagenomics assembly of PRJE...</td>\n",
       "      <td>SUBMITTED</td>\n",
       "      <td>2022-09-05T14:38:08</td>\n",
       "      <td>[{'id': 'root:Environmental:Aquatic:Marine', '...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>studies</td>\n",
       "      <td>MGYS00006058</td>\n",
       "      <td>41</td>\n",
       "      <td>MGYS00006058</td>\n",
       "      <td>PRJEB51989</td>\n",
       "      <td>False</td>\n",
       "      <td>ERP136659</td>\n",
       "      <td>EMG</td>\n",
       "      <td>The Third Party Annotation (TPA) assembly was ...</td>\n",
       "      <td>EMG produced TPA metagenomics assembly of PRJE...</td>\n",
       "      <td>SUBMITTED</td>\n",
       "      <td>2022-08-27T19:19:53</td>\n",
       "      <td>[{'id': 'root:Environmental:Aquatic:Marine', '...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      type            id  attributes.samples-count attributes.accession  \\\n",
       "0  studies  MGYS00006074                        75         MGYS00006074   \n",
       "1  studies  MGYS00006072                       130         MGYS00006072   \n",
       "2  studies  MGYS00006034                       971         MGYS00006034   \n",
       "3  studies  MGYS00006061                         3         MGYS00006061   \n",
       "4  studies  MGYS00006058                        41         MGYS00006058   \n",
       "\n",
       "  attributes.bioproject  attributes.is-private attributes.secondary-accession  \\\n",
       "0            PRJEB51168                  False                      ERP135767   \n",
       "1            PRJEB54918                  False                      ERP139784   \n",
       "2            PRJEB50181                  False                      ERP134737   \n",
       "3            PRJEB52804                  False                      ERP137544   \n",
       "4            PRJEB51989                  False                      ERP136659   \n",
       "\n",
       "  attributes.centre-name                          attributes.study-abstract  \\\n",
       "0                    EMG  The Third Party Annotation (TPA) assembly was ...   \n",
       "1                    EMG  The Third Party Annotation (TPA) assembly was ...   \n",
       "2                    EMG  The Third Party Annotation (TPA) assembly was ...   \n",
       "3                    EMG  The Third Party Annotation (TPA) assembly was ...   \n",
       "4                    EMG  The Third Party Annotation (TPA) assembly was ...   \n",
       "\n",
       "                               attributes.study-name  \\\n",
       "0  EMG produced TPA metagenomics assembly of PRJE...   \n",
       "1  EMG produced TPA metagenomics assembly of PRJN...   \n",
       "2  EMG produced TPA metagenomics assembly of PRJN...   \n",
       "3  EMG produced TPA metagenomics assembly of PRJE...   \n",
       "4  EMG produced TPA metagenomics assembly of PRJE...   \n",
       "\n",
       "  attributes.data-origination attributes.last-update  \\\n",
       "0                   SUBMITTED    2022-10-21T19:56:09   \n",
       "1                   SUBMITTED    2022-10-13T14:03:52   \n",
       "2                   SUBMITTED    2022-10-10T12:22:12   \n",
       "3                   SUBMITTED    2022-09-05T14:38:08   \n",
       "4                   SUBMITTED    2022-08-27T19:19:53   \n",
       "\n",
       "                           relationships.biomes.data  \n",
       "0  [{'id': 'root:Environmental:Aquatic:Marine', '...  \n",
       "1  [{'id': 'root:Environmental:Aquatic:Marine:Oce...  \n",
       "2  [{'id': 'root:Environmental:Aquatic:Marine', '...  \n",
       "3  [{'id': 'root:Environmental:Aquatic:Marine', '...  \n",
       "4  [{'id': 'root:Environmental:Aquatic:Marine', '...  "
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "from jsonapi_client import Session, Modifier\n",
    "\n",
    "atlanteco_endpoint = 'super-studies/atlanteco/flagship-studies'\n",
    "with Session(\"https://www.ebi.ac.uk/metagenomics/api/v1\") as mgnify:\n",
    "    studies = map(lambda r: r.json, mgnify.iterate(atlanteco_endpoint))\n",
    "    studies = pd.json_normalize(studies)\n",
    "studies[:5]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5ae10b66-889b-4932-b62d-37b89feac050",
   "metadata": {},
   "source": [
    "## Show the studies' samples on a map\n",
    "\n",
    "We can fetch the Samples for each Study, and concatenate them all into one Dataframe.\n",
    "Each sample has geolocation data in its `attributes` - this is what we need to build a map.\n",
    "\n",
    "It takes time to fetch data for all samples, so **let's show samples from the first 6 studies only.** "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "bc5d0fc9-5c05-4bb1-b401-507b6e8c8877",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "fetching MGYS00006074 samples\n",
      "fetching MGYS00006072 samples\n",
      "fetching MGYS00006034 samples\n",
      "fetching MGYS00006061 samples\n",
      "fetching MGYS00006058 samples\n",
      "fetching MGYS00006057 samples\n"
     ]
    }
   ],
   "source": [
    "studies_samples = []\n",
    "\n",
    "with Session(\"https://www.ebi.ac.uk/metagenomics/api/v1\") as mgnify:\n",
    "    for idx, study in studies[:6].iterrows():\n",
    "        print(f\"fetching {study.id} samples\")\n",
    "        samples = map(lambda r: r.json, mgnify.iterate(f'studies/{study.id}/samples'))\n",
    "        samples = pd.json_normalize(samples)\n",
    "        samples = pd.DataFrame(data={\n",
    "            'accession': samples['id'],\n",
    "            'sample_id': samples['id'],\n",
    "            'study': study.id, \n",
    "            'lon': samples['attributes.longitude'],\n",
    "            'lat': samples['attributes.latitude'],\n",
    "            'color': \"#FF0000\",\n",
    "        })\n",
    "        samples.set_index('accession', inplace=True)\n",
    "        studies_samples.append(samples)\n",
    "studies_samples = pd.concat(studies_samples)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "2f503c45-d3c0-45b8-a9d5-e3d3eb69ce95",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_id</th>\n",
       "      <th>study</th>\n",
       "      <th>lon</th>\n",
       "      <th>lat</th>\n",
       "      <th>color</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>accession</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ERS492820</th>\n",
       "      <td>ERS492820</td>\n",
       "      <td>MGYS00006074</td>\n",
       "      <td>-140.5216</td>\n",
       "      <td>-9.1504</td>\n",
       "      <td>#FF0000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERS477953</th>\n",
       "      <td>ERS477953</td>\n",
       "      <td>MGYS00006074</td>\n",
       "      <td>1.9478</td>\n",
       "      <td>37.0541</td>\n",
       "      <td>#FF0000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERS488658</th>\n",
       "      <td>ERS488658</td>\n",
       "      <td>MGYS00006074</td>\n",
       "      <td>39.8750</td>\n",
       "      <td>18.3967</td>\n",
       "      <td>#FF0000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERS490548</th>\n",
       "      <td>ERS490548</td>\n",
       "      <td>MGYS00006074</td>\n",
       "      <td>-35.1803</td>\n",
       "      <td>-20.9354</td>\n",
       "      <td>#FF0000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERS488147</th>\n",
       "      <td>ERS488147</td>\n",
       "      <td>MGYS00006074</td>\n",
       "      <td>5.9422</td>\n",
       "      <td>39.0609</td>\n",
       "      <td>#FF0000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           sample_id         study       lon      lat    color\n",
       "accession                                                     \n",
       "ERS492820  ERS492820  MGYS00006074 -140.5216  -9.1504  #FF0000\n",
       "ERS477953  ERS477953  MGYS00006074    1.9478  37.0541  #FF0000\n",
       "ERS488658  ERS488658  MGYS00006074   39.8750  18.3967  #FF0000\n",
       "ERS490548  ERS490548  MGYS00006074  -35.1803 -20.9354  #FF0000\n",
       "ERS488147  ERS488147  MGYS00006074    5.9422  39.0609  #FF0000"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "studies_samples[:5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "6c09ddf3-746e-4891-96a5-8f70682d530f",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "58c5461cb9994b958be0689cfea3f2d3",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Map(center=[0, 0], controls=(ZoomControl(options=['position', 'zoom_in_text', 'zoom_in_title', 'zoom_out_text'…"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import leafmap\n",
    "m = leafmap.Map(center=(0, 0), zoom=2)\n",
    "m.add_points_from_xy(studies_samples, \n",
    "                     x='lon', \n",
    "                     y='lat', \n",
    "                     popup=[\"study\", \"sample_id\"], \n",
    "                     color_column='color',\n",
    "                     add_legend=False)\n",
    "m"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7fc7b990-0804-43a3-bfc8-dc14e5a61a88",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Check GO term presence\n",
    "Let's check whether a specific identifier is present in each sample. This example is written for GO-term 'GO:0015878', but other identifier types are available on the MGnify API.\n",
    "\n",
    "We will work with MGnify analyses (`MGYA`s) corresponding to chosen samples. We filter analyses by \n",
    "- pipeline version: 5.0\n",
    "- experiment type: assembly\n",
    "\n",
    "This example shows how to process **just the first 10 samples** (again, because the full dataset takes a while to fetch).\n",
    "Firstly, get analyses for each sample."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "8c7d6790-1cec-4ecf-b8be-56e3e03fe669",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "processing ERS492820\n",
      "processing ERS477953\n",
      "processing ERS488658\n",
      "processing ERS490548\n",
      "processing ERS488147\n",
      "processing ERS490665\n",
      "processing ERS490382\n",
      "processing ERS492932\n",
      "processing ERS477931\n",
      "processing ERS488723\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>type</th>\n",
       "      <th>id</th>\n",
       "      <th>attributes.pipeline-version</th>\n",
       "      <th>attributes.experiment-type</th>\n",
       "      <th>attributes.analysis-summary</th>\n",
       "      <th>attributes.analysis-status</th>\n",
       "      <th>attributes.accession</th>\n",
       "      <th>attributes.is-private</th>\n",
       "      <th>attributes.complete-time</th>\n",
       "      <th>attributes.instrument-platform</th>\n",
       "      <th>attributes.instrument-model</th>\n",
       "      <th>relationships.study.data.id</th>\n",
       "      <th>relationships.study.data.type</th>\n",
       "      <th>relationships.sample.data.id</th>\n",
       "      <th>relationships.sample.data.type</th>\n",
       "      <th>relationships.assembly.data.id</th>\n",
       "      <th>relationships.assembly.data.type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>analysis-jobs</td>\n",
       "      <td>MGYA00615698</td>\n",
       "      <td>5.0</td>\n",
       "      <td>assembly</td>\n",
       "      <td>[{'key': 'Submitted nucleotide sequences', 'va...</td>\n",
       "      <td>completed</td>\n",
       "      <td>MGYA00615698</td>\n",
       "      <td>False</td>\n",
       "      <td>2022-10-21T19:56:12</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>Illumina HiSeq 2000</td>\n",
       "      <td>MGYS00006074</td>\n",
       "      <td>studies</td>\n",
       "      <td>ERS492820</td>\n",
       "      <td>samples</td>\n",
       "      <td>ERZ8733628</td>\n",
       "      <td>assemblies</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>analysis-jobs</td>\n",
       "      <td>MGYA00615631</td>\n",
       "      <td>5.0</td>\n",
       "      <td>assembly</td>\n",
       "      <td>[{'key': 'Submitted nucleotide sequences', 'va...</td>\n",
       "      <td>completed</td>\n",
       "      <td>MGYA00615631</td>\n",
       "      <td>False</td>\n",
       "      <td>2022-10-16T16:11:02</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>Illumina HiSeq 2000</td>\n",
       "      <td>MGYS00006074</td>\n",
       "      <td>studies</td>\n",
       "      <td>ERS477953</td>\n",
       "      <td>samples</td>\n",
       "      <td>ERZ6778699</td>\n",
       "      <td>assemblies</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>analysis-jobs</td>\n",
       "      <td>MGYA00615679</td>\n",
       "      <td>5.0</td>\n",
       "      <td>assembly</td>\n",
       "      <td>[{'key': 'Submitted nucleotide sequences', 'va...</td>\n",
       "      <td>completed</td>\n",
       "      <td>MGYA00615679</td>\n",
       "      <td>False</td>\n",
       "      <td>2022-10-17T13:44:04</td>\n",
       "      <td>LS454</td>\n",
       "      <td>454 GS FLX Titanium</td>\n",
       "      <td>MGYS00006074</td>\n",
       "      <td>studies</td>\n",
       "      <td>ERS477953</td>\n",
       "      <td>samples</td>\n",
       "      <td>ERZ5402186</td>\n",
       "      <td>assemblies</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>analysis-jobs</td>\n",
       "      <td>MGYA00615697</td>\n",
       "      <td>5.0</td>\n",
       "      <td>assembly</td>\n",
       "      <td>[{'key': 'Submitted nucleotide sequences', 'va...</td>\n",
       "      <td>completed</td>\n",
       "      <td>MGYA00615697</td>\n",
       "      <td>False</td>\n",
       "      <td>2022-10-21T19:54:50</td>\n",
       "      <td>LS454</td>\n",
       "      <td>454 GS FLX Titanium</td>\n",
       "      <td>MGYS00006074</td>\n",
       "      <td>studies</td>\n",
       "      <td>ERS477953</td>\n",
       "      <td>samples</td>\n",
       "      <td>ERZ5402156</td>\n",
       "      <td>assemblies</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>analysis-jobs</td>\n",
       "      <td>MGYA00615696</td>\n",
       "      <td>5.0</td>\n",
       "      <td>assembly</td>\n",
       "      <td>[{'key': 'Submitted nucleotide sequences', 'va...</td>\n",
       "      <td>completed</td>\n",
       "      <td>MGYA00615696</td>\n",
       "      <td>False</td>\n",
       "      <td>2022-10-21T19:37:54</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>Illumina HiSeq 2000</td>\n",
       "      <td>MGYS00006074</td>\n",
       "      <td>studies</td>\n",
       "      <td>ERS488658</td>\n",
       "      <td>samples</td>\n",
       "      <td>ERZ5402542</td>\n",
       "      <td>assemblies</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            type            id attributes.pipeline-version  \\\n",
       "0  analysis-jobs  MGYA00615698                         5.0   \n",
       "0  analysis-jobs  MGYA00615631                         5.0   \n",
       "1  analysis-jobs  MGYA00615679                         5.0   \n",
       "2  analysis-jobs  MGYA00615697                         5.0   \n",
       "0  analysis-jobs  MGYA00615696                         5.0   \n",
       "\n",
       "  attributes.experiment-type  \\\n",
       "0                   assembly   \n",
       "0                   assembly   \n",
       "1                   assembly   \n",
       "2                   assembly   \n",
       "0                   assembly   \n",
       "\n",
       "                         attributes.analysis-summary  \\\n",
       "0  [{'key': 'Submitted nucleotide sequences', 'va...   \n",
       "0  [{'key': 'Submitted nucleotide sequences', 'va...   \n",
       "1  [{'key': 'Submitted nucleotide sequences', 'va...   \n",
       "2  [{'key': 'Submitted nucleotide sequences', 'va...   \n",
       "0  [{'key': 'Submitted nucleotide sequences', 'va...   \n",
       "\n",
       "  attributes.analysis-status attributes.accession  attributes.is-private  \\\n",
       "0                  completed         MGYA00615698                  False   \n",
       "0                  completed         MGYA00615631                  False   \n",
       "1                  completed         MGYA00615679                  False   \n",
       "2                  completed         MGYA00615697                  False   \n",
       "0                  completed         MGYA00615696                  False   \n",
       "\n",
       "  attributes.complete-time attributes.instrument-platform  \\\n",
       "0      2022-10-21T19:56:12                       ILLUMINA   \n",
       "0      2022-10-16T16:11:02                       ILLUMINA   \n",
       "1      2022-10-17T13:44:04                          LS454   \n",
       "2      2022-10-21T19:54:50                          LS454   \n",
       "0      2022-10-21T19:37:54                       ILLUMINA   \n",
       "\n",
       "  attributes.instrument-model relationships.study.data.id  \\\n",
       "0         Illumina HiSeq 2000                MGYS00006074   \n",
       "0         Illumina HiSeq 2000                MGYS00006074   \n",
       "1         454 GS FLX Titanium                MGYS00006074   \n",
       "2         454 GS FLX Titanium                MGYS00006074   \n",
       "0         Illumina HiSeq 2000                MGYS00006074   \n",
       "\n",
       "  relationships.study.data.type relationships.sample.data.id  \\\n",
       "0                       studies                    ERS492820   \n",
       "0                       studies                    ERS477953   \n",
       "1                       studies                    ERS477953   \n",
       "2                       studies                    ERS477953   \n",
       "0                       studies                    ERS488658   \n",
       "\n",
       "  relationships.sample.data.type relationships.assembly.data.id  \\\n",
       "0                        samples                     ERZ8733628   \n",
       "0                        samples                     ERZ6778699   \n",
       "1                        samples                     ERZ5402186   \n",
       "2                        samples                     ERZ5402156   \n",
       "0                        samples                     ERZ5402542   \n",
       "\n",
       "  relationships.assembly.data.type  \n",
       "0                       assemblies  \n",
       "0                       assemblies  \n",
       "1                       assemblies  \n",
       "2                       assemblies  \n",
       "0                       assemblies  "
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "analyses = []\n",
    "with Session(\"https://www.ebi.ac.uk/metagenomics/api/v1\") as mgnify:\n",
    "    for idx, sample in studies_samples[:10].iterrows():\n",
    "        print(f\"processing {sample.sample_id}\")\n",
    "        filtering = Modifier(f\"pipeline_version=5.0&sample_accession={sample.sample_id}&experiment_type=assembly\")\n",
    "        analysis = map(lambda r: r.json, mgnify.iterate('analyses', filter=filtering))\n",
    "        analysis = pd.json_normalize(analysis)\n",
    "        analyses.append(analysis)\n",
    "analyses = pd.concat(analyses)\n",
    "analyses[:5]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0b4ad3aa-f19f-4007-b64f-59edf10b5734",
   "metadata": {},
   "source": [
    "Next, check each analysis for GO term presence/absence. We add a column to the dataframe with a colour: blue if GO term was found and red if not."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "c6980125-7fd7-4fdb-9f15-c1b9c93a21c8",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "processing MGYA00615698\n",
      "processing MGYA00615631\n",
      "processing MGYA00615679\n",
      "processing MGYA00615697\n",
      "processing MGYA00615696\n",
      "processing MGYA00615695\n",
      "processing MGYA00615673\n",
      "processing MGYA00615694\n",
      "processing MGYA00615693\n",
      "processing MGYA00615692\n",
      "processing MGYA00615674\n",
      "processing MGYA00615658\n",
      "processing MGYA00615684\n",
      "processing MGYA00615691\n",
      "processing MGYA00615689\n",
      "processing MGYA00615690\n"
     ]
    }
   ],
   "source": [
    "identifier = \"go-terms\"\n",
    "go_term = 'GO:0015878'\n",
    "go_data = []\n",
    "with Session(\"https://www.ebi.ac.uk/metagenomics/api/v1\") as mgnify:\n",
    "    for idx, mgya in analyses.iterrows():\n",
    "        print(f\"processing {mgya.id}\")\n",
    "        analysis_identifier = map(lambda r: r.json, mgnify.iterate(f'analyses/{mgya.id}/{identifier}'))\n",
    "        analysis_identifier = pd.json_normalize(analysis_identifier)\n",
    "        go_data.append(\"#0000FF\" if go_term in list(analysis_identifier.id) else \"#FF0000\")\n",
    "analyses.insert(2, identifier, go_data, True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8f4242af-617a-4ad9-b265-f7b629da9b52",
   "metadata": {},
   "source": [
    "Join the analyses and sample tables to have geolocation data and identifier presence data together.\n",
    "\n",
    "We'll create a new sub-DataFrame with a subset of the fields and add them to the map."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "0decacbe-04ed-4fdc-93fe-9f3acd78624b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "1604a72cd7b94f9b9bae6878b6d5134a",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Map(center=[0, 0], controls=(ZoomControl(options=['position', 'zoom_in_text', 'zoom_in_title', 'zoom_out_text'…"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = analyses.join(studies_samples.set_index('sample_id'), on='relationships.sample.data.id')\n",
    "df2 = df[[identifier, 'lon', 'lat', 'study', 'attributes.accession', 'relationships.study.data.id', 'relationships.sample.data.id', 'relationships.assembly.data.id']].copy()\n",
    "df2 = df2.set_index(\"study\")\n",
    "df2 = df2.rename(columns={\"attributes.accession\": \"analysis_ID\", \n",
    "                          'relationships.study.data.id': \"study_ID\",\n",
    "                          'relationships.sample.data.id': \"sample_ID\", \n",
    "                          'relationships.assembly.data.id': \"assembly_ID\"\n",
    "                         })\n",
    "m = leafmap.Map(center=(0, 0), zoom=2)\n",
    "m.add_points_from_xy(df2, \n",
    "                     x='lon', \n",
    "                     y='lat', \n",
    "                     popup=[\"study_ID\", \"sample_ID\", \"assembly_ID\", \"analysis_ID\"],\n",
    "                    color_column=identifier, add_legend=False)\n",
    "m"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3985abb6-73a6-418f-8657-5fb383022d90",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
